### Use Cases and Applications: Research articles using the SG-NEx data

This site lists some examples how the SG-NEx data resource is used in research:

#### Transcript discovery/quantification

- Schulz, Laura, et al. "Direct long-read RNA sequencing identifies a subset of questionable exitrons likely arising from reverse transcription artifacts." _Genome Biology_ 22.1 (2021): 1-12. https://doi.org/10.1186/s13059-021-02411-1
- Annaldasula, Siddharth, Martyna Gajos, and Andreas Mayer. "IsoTV: processing and visualizing functional features of translated transcript isoforms." _Bioinformatics_ (2021). https://doi.org/10.1093/bioinformatics/btab103

#### RNA modifications

- Pratanwanich, Ploy N., et al. "Identification of differential RNA modifications from nanopore direct RNA sequencing with xPore." _Nature Biotechnology_ (2021): 1-9. https://doi.org/10.1038/s41587-021-00949-w
- Hendra, Christopher, et al. "Detection of m6A from direct RNA sequencing using a Multiple Instance Learning framework." _bioRxiv_ (2021). https://doi.org/10.1101/2021.09.20.461055
- Campos, Jo√£o H., et al. "Direct RNA sequencing reveals SARS-CoV-2 m6A sites and possible differential DRACH motif methylation among variants." _bioRxiv_ (2021). https://doi.org/10.1101/2021.08.24.457397

#### Fusion detection

- Davidson, Nadia M., et al. "JAFFAL: Detecting fusion genes with long read transcriptome sequencing." _bioRxiv_ (2021). https://doi.org/10.1101/2021.04.26.441398

#### Reviews and other use cases

- De Paoli-Iseppi, Ricardo, Josie Gleeson, and Michael B. Clark. "Isoform age-splice isoform profiling using long-read technologies." Frontiers in Molecular Biosciences 8 (2021). https://doi.org/10.3389/fmolb.2021.711733

Please feel free to add more examples by creating a pull request.
