---
title: "Figure 3"
output: html_notebook
---

# set-up
```{r 0-setup, include=FALSE}
knitr::opts_chunk$set(
  comment = '', fig.width = 6, fig.height = 6,
  results = "hide",
  message = FALSE
)
```

```{r 0-library-loads}
require(GenomicFeatures)
require(GenomicAlignments) ##readGAlignments
require(AnnotationDbi)#loadDb
require(data.table)#fast large dataset manipulation
require(readxl)
require(dplyr)


require(ggplot2)
require(RColorBrewer)
require(gridExtra)

library(gplots)
library(RColorBrewer)
library(limma)
library(ggpubr)

library(ComplexHeatmap)
library(circlize)
library(viridis)
gctorture(FALSE)
library(tximport) # 

library(cowplot) ## ggdraw
library(ggplotify)
source("utility_function.R")
```

# load general data
```{r}
cat('Setting working directory')
wkdir <- 'wkdir'
general_list <- readRDS("general_list2023-04-27.rds")
samples_wSpikein <- general_list$samples_wSpikein
cellLines <- general_list$cellLines
protocolCol <-  general_list$protocolCol
protocolVec <-  general_list$protocolVec
protocolLabel <- general_list$protocolLabel

 txvec <- fread(paste0("txList_matchingToGTF_wtChrIs.txt"), header = FALSE)

    txvec <- gsub("\\..*","",txvec$V1)
    ensemblAnnotations.transcripts <- copy(general_list$ensemblAnnotations.transcripts)
    setnames(ensemblAnnotations.transcripts, "ensembl_gene_id","gene_name")
    ensemblAnnotations.transcripts <- data.table(tx_name = txvec, status = TRUE)[ensemblAnnotations.transcripts, on = "tx_name"]
    ensemblAnnotations.transcripts[is.na(status), status := FALSE]
    ensemblAnnotations.transcripts[, all_in := all(status), by = gene_name]
    genevec <- unique(ensemblAnnotations.transcripts[which(all_in)]$gene_name)
samples <- general_list$samples
```

# load all data
```{r}
temp <- samples_wSpikein[,.(old_runname, runname)]
setnames(temp, "runname", "new_name")
```


```{r bambu-lr expression data using all methods}
comDataTranscript <- readRDS(paste0(wkdir, "combinedExpressionDataTranscript_19June2023.rds"))
comDataGene <- readRDS(paste0(wkdir, "combinedExpressionDataGene_19June2023.rds"))

comDataGene[, old_runname := runname]
comDataGene <- temp[comDataGene, on = "old_runname"]
comDataGene[!is.na(new_name), runname := new_name]
comDataGene[, `:=`(old_runname = NULL, new_name = NULL)]
comDataGene[runname == "GIS_HCT116_PacBio-SMRTcell_Rep5_Run1", runname := "SGNex_Hct116_PacBio-SMRTcell_replicate7_run1"]
comDataGene[, cellLine := gsub("HCT116","Hct116",cellLine)]

comDataTranscript[, old_runname := runname]
comDataTranscript <- temp[comDataTranscript, on = "old_runname"]
comDataTranscript[!is.na(new_name), runname := new_name]
comDataTranscript[, `:=`(old_runname = NULL, new_name = NULL)]
comDataTranscript[runname == "GIS_HCT116_PacBio-SMRTcell_Rep5_Run1", runname := "SGNex_Hct116_PacBio-SMRTcell_replicate7_run1"]
comDataTranscript[, cellLine := gsub("HCT116","Hct116",cellLine)]

```


```{r}
runnamevec <- unique(comDataTranscript[method %in%  c("bambu_lr","salmon_sr")&(ntotal>=400000)]$runname)
```


# load spike-in data
```{r}
com_dataList <- readRDS("combinedExpressionDataList_spikein_25May.rds")
com_data <- com_dataList[[1]]
com_data_gene <- com_dataList[[2]]
```

# normalize concentration

```{r}
samples_wSpikein <- general_list$samples_wSpikein
samples_wSpikein[grepl("PacBio", runname), RNAcontent := gsub("SIRV-1 E2 ","",gsub("sequin Mix A v1.0 ","",RNAcontent))]
```


```{r spike-in}
spike_in <- readRDS("spike_in.rds")
setnames(spike_in, "spike_in","spike_in_name")
spike_in[, spike_in_type := ifelse(spike_in_name == "sirv_e2","SIRV-1",
                                   ifelse(spike_in_name %in% c( "sequinMixAV1","sequinMixAV2"),spike_in_name, "SIRV-4"))]
spike_in[, spike_in_perc := ifelse(spike_in_type != "SIRV-4",1, 
                            ifelse(spike_in_name == "ercc", 51.8/102.2,
                                   ifelse(spike_in_name == "long_sirv",9.0/102.2,41.4/102.2)))]

# find out the total read depth for each spike-in and protocol type, then normalize accordingly 
spike_in_samples <- optiRum::CJ.dt(data.table(spike_in_type = unique(spike_in$spike_in_type), spike_in_attr = c("SIRV-1","SIRV-4","v1.0","V2")),
                       data.table(protocol = unique(com_data$protocol)))
spikein_samples_list <- lapply(seq_len(nrow(spike_in_samples)), function(t){
    return(c(samples_wSpikein[grepl(spike_in_samples[t]$spike_in_attr,RNAcontent)&(gsub("-SMRTcell","",protocol_type) == spike_in_samples[t]$protocol)]$runname,samples_wSpikein[grepl(spike_in_samples[t]$spike_in_attr,RNAcontent)&(gsub("-SMRTcell","",protocol_type) == spike_in_samples[t]$protocol)]$old_runname))
})

ntotalData <- unique(comDataTranscript[, .(runname,ntotal, protocol, method)])

# sanity check to check what is the degree of difference of read count between different methods 
ntotalCount <- unlist(lapply(seq_len(nrow(spike_in_samples)), function(t){
    method_type <- "bambu_lr"
    if(spike_in_samples[t]$protocol == "Illumina") method_type <- "salmon_sr"
    return(sum(ntotalData[runname %in% spikein_samples_list[[t]]&(method == method_type)]$ntotal))
}))
spike_in_samples[, ntotal := ntotalCount]
spike_in <- spike_in_samples[spike_in, on = "spike_in_type", allow.cartesian = TRUE]
spike_in[, sum_conc := sum(conc,na.rm = TRUE), by = list(spike_in_type, protocol, spike_in_name)]
spike_in <- spike_in[, norm_conc := conc/sum_conc*spike_in_perc*0.01*ntotal, by = list(spike_in_type, protocol)]

spike_in_gene <- unique(spike_in[,list(conc = sum(conc,na.rm = TRUE),
                                       norm_conc = sum(norm_conc,na.rm = TRUE)), by = list(spike_in_name, spike_in_type,gene_name, protocol)])
```

# prepare gene metrics 
```{r spike-in gene count}
si_data <- com_data_gene[grepl("^(R|ERCC|SIRV)",gene_name)]
si_data[, runname := gsub("v1.0_PacBio","V2_PacBio",runname)]
si_data <- si_data[!grepl("V-1_PacBio",runname)]
si_data[, spike_in_type := gsub("v1.0","V1",unlist(strsplit(runname, "_"))[2]), by = runname]
p_type <- c("directcDNA","cDNA","directRNA")
si_data[, runname := gsub("_sorted","", runname)]
ntotal_bambu <- unique(si_data[method %in% c("bambu_lr","salmon_sr")][,.(ntotal,runname)])


cj_si <- spike_in_gene[unique(si_data[,.(method, runname, spike_in_type,protocol)]), on = c("spike_in_type","protocol"),allow.cartesian = TRUE]
si_data <- si_data[cj_si, on = c("spike_in_type","gene_name","protocol","runname","method")]
si_data[is.na(estimates), estimates := 0]
si_data[is.na(normEst), normEst := 0]
si_data <- ntotal_bambu[si_data, on = "runname"]
si_data[, cpm_norm_conc := norm_conc/ntotal*10^6]
si_data <- si_data[!is.na(spike_in_name)]
si_data[, spike_in_general_name := ifelse(grepl("SIRV", spike_in_type),"SIRV","Sequin")]
si_data[, spike_in_version := ifelse(grepl("1", spike_in_type),"1","2")]
si_data[, spike_in_name := factor(spike_in_name, levels = c("sequinMixAV1","sequinMixAV2",
    "sirv_e2","ercc","sirv_e0","long_sirv"), labels = c("Sequin (v1)","Sequin (v2)", "SIRV (E2)",
    "ERCC","SIRV (E0)", "Long SIRV"))]
si_data[, spike_in_general_name_revised := ifelse(grepl("SIRV", spike_in_name),"SIRV",
                                ifelse(grepl("Sequin", spike_in_name),"Sequin",
                                ifelse(grepl("ERCC", spike_in_name),"ERCC",NA)))]
si_data[, `:=`(ae = abs(log2(cpm_norm_conc+1)-log2(normEst+1)),
               ard = abs(log2(cpm_norm_conc+1)-log2(normEst+1))/(log2(cpm_norm_conc+1)+log2(normEst+1))*2)]
si_data[is.na(cpm_norm_conc), cpm_norm_conc := 0]
si_data <- si_data[!is.na(conc)]

si_data_gene <- copy(si_data)

```

```{r spike-in gene metrics}
rsq <- function(x, y) summary(lm(y~x))$r.squared #https://stackoverflow.com/questions/40901445/function-to-calculate-r2-r-squared-in-r
si_data[, new_spike_in_name := ifelse(spike_in_name %in% c("SIRV (E0)","Long SIRV"), "SIRV (E0+Long)"]
Allmetrics <- unique(si_data[, list(corM = cor(log2(cpm_norm_conc+1), log2(normEst+1),  method = "spearman"),
                                    corP = cor(log2(cpm_norm_conc+1), log2(normEst+1)),
                                    mae = mean(abs(log2(cpm_norm_conc+1)-log2(normEst+1))),
                                    rmse = sqrt(mean((log2(cpm_norm_conc+1)-log2(normEst+1))^2)),
                                    mard = mean(abs(log2(cpm_norm_conc+1)-log2(normEst+1))/(log2(cpm_norm_conc+1)+log2(normEst+1))*2, na.rm = TRUE),
                                    mard_mod =  mean((log2(cpm_norm_conc+1)-log2(normEst+1))/(log2(cpm_norm_conc+1)+log2(normEst+1))*2, na.rm = TRUE),
                                    r2 = rsq(log2(cpm_norm_conc+1),log2(normEst+1))), by = list(method, protocol, spike_in_type, spike_in_name)])
library(hrbrthemes)
library(ggpubr)

Allmetrics_long <- melt(Allmetrics, id.vars = c("method","protocol","spike_in_type","spike_in_name"), measure.vars = c("corM","corP","mae","rmse","mard","mard_mod","r2"))

gene_metrics <- copy(Allmetrics_long)
```

# main figure 

## a spike-in gene scatter 
```{r}
plotdata_gene <- si_data_gene[method %in% c("salmon_lr","salmon_sr")&!(protocol %in% c("directRNA","PacBio"))]
plotdata_gene[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_gene_spikein_scatter <- ggplot(plotdata_gene, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = spike_in_version), size = 0.5, se = FALSE )+
                geom_point(aes(col = spike_in_version),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = spike_in_version,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(spike_in_general_name_revised~protocol)+
                theme_classic()
```



## b replicate gene correlation
```{r}
mat_cor <- readRDS(paste0("replicate_gene_expression_comparison_salmon_lr_salmon_sr_25May2023.rds"))
mat_cor[, within_cellLine := !grepl("_",cellLine)]

mat_cor[, pc_factor := factor(protocol_comparison, rev(unique(mat_cor$protocol_comparison)[c(1,4,2,5,3,6)]))]
mat_cor[, ct_factor := factor(common_type, unique(mat_cor$common_type))]
mat_cor[, gc_factor := factor(agg_gene_cluster, unique(mat_cor$agg_gene_cluster)[c(3,2,1,4)])]

mat_cor[, short_read := factor(!grepl("Illumina",pc_factor), labels = c("short read vs long read","long read vs long read"))]
p_replicate_gene_final_protein <-
    ggplot(mat_cor[gc_factor %in% c("protein_coding","lncRNA")[1]], aes(pc_factor , r, fill = within_cellLine))+
    geom_hline(yintercept = c(0.6,0.8), col = "grey", linetype = "dashed")+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 4)+
    scale_color_brewer(type = "qual", palette = 4)+
    facet_wrap(~short_read, ncol = 2, scales = "free")+
    ylab("Spearman correlation")+
    xlab("Protocol pairs")+
    theme_classic()
p_replicate_gene_final_lncRNA <-
    ggplot(mat_cor[gc_factor %in% c("protein_coding","lncRNA")[2]], aes(pc_factor , r, fill = within_cellLine))+
    geom_hline(yintercept = c(0.6,0.8), col = "grey", linetype = "dashed")+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 4)+
    scale_color_brewer(type = "qual", palette = 4)+
    facet_wrap(~short_read, ncol = 2, scales = "free")+
    ylab("Spearman correlation")+
    xlab("Protocol pairs")+
    theme_classic()
```

## c replicate gene scatter 
```{r}
library(GGally)
cellLines <- c("A549","K562","HepG2","Hct116","MCF7","H9","HEYA8")
k <- cellLines[1]
source(paste0('gene_cluster_code.R'))
    temp_data <- comDataGene[cellLine == k&(method %in% c("salmon_lr","salmon_sr"))] 
    ave_data <- unique(temp_data[, list(meanNormEst = mean(normEst)), by = list(gene_name, protocol_general)])
    ave_data <-  unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[ave_data, on = "gene_name"]
    ave_data[, gene_cluster := grepl("protein_coding", gene_biotype)]
    ave_data[, log2NormEst := log2(meanNormEst+1)]
     
    ave_data[, gene_cluster:=ifelse(gene_biotype %in% tr_gene_list,'TR gene',
                                    ifelse(gene_biotype %in% long_noncoding_rna_list, 'lncRNA',
                                           ifelse(gene_biotype %in% noncoding_rna_list, 'ncRNA',
                                                  ifelse(gene_biotype %in% pseudogene_list,'Pseudogene', ifelse(gene_biotype %in% ig_gene_list, 'IG gene',gene_biotype)))))]
    ave_data[, gene_cluster := ifelse(gene_cluster %in% c("IG gene","TR gene", "Mt_tRNA","Mt_rRNA","ribozyme"),"others", gene_cluster)]
    
   ave_data[, agg_gene_cluster := ifelse(gene_cluster == "processed_transcript", "lncRNA",
                                                 ifelse(gene_cluster %in% c("ncRNA","Pseudogene"), "others", gene_cluster))]

    plotdata1 <- dcast(ave_data[grepl("ENSG", gene_name)], gene_name + agg_gene_cluster + gene_cluster + gene_biotype ~ protocol_general, value.var = "log2NormEst")
    p_protein_coding <- ggplot(plotdata1[agg_gene_cluster == "protein_coding"], aes(x = directcDNA, y = Illumina))+
        geom_abline(intercept = 0, slope = 1)+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(breaks = c(10,100,1000),low = "steelblue1", high = "steelblue4",trans = "log10")+
        geom_density_2d()+
         xlim(c(0,14))+
         ylim(c(0,14))+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
        ggtitle("protein coding")+theme_classic()
     p_lncRNA <- ggplot(plotdata1[agg_gene_cluster == "lncRNA"], aes(x = directcDNA, y = Illumina))+
          geom_abline(intercept = 0, slope = 1)+
         geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(breaks = c(10,100,1000),low = "steelblue1", high = "steelblue4",trans = "log10")+
        ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
         xlim(c(0,14))+
         ylim(c(0,14))+
         ggtitle("lncRNA")+theme_classic()
```

## d gene heatmap with dendrograms 
```{r}
source("utility_function.R")
runnamevec <- unique(comDataTranscript[method %in%  c("bambu_lr","salmon_sr")&(ntotal>=400000)]$runname)
methodNamesList <- CJ(lr = c("bambu_lr","salmon_lr"),
                      sr = c("rsem_sr","salmon_sr"),
                      gene = c(TRUE, FALSE),
                      data_type = c("all","protocol_batch_removal","cellline_batch_removal"))
comDataGeneFiltered <- comDataGene[method %in% c("bambu_lr","salmon_lr","rsem_sr","salmon_sr")&(!grepl("PacBio",protocol_general))]


lr <- "salmon_lr"
sr <- "salmon_sr"
gene <- TRUE
data_type <- "all"
number_of_genes <- 10000

p_heatmap <- plot_heatmap(dt = comDataGeneFiltered, methodNames = c(lr, sr), gene, data_type,genevec,ensemblAnnotations.transcripts , samples, number_of_genes, runnamevec, plot_type = "heatmap")
```


# Fig. 3
arrange main figure 
```{r}
# as.ggplot to change non ggplot types to ggplot so that plots can be integrated 
## 
hm = draw(p_heatmap)

pdf("figure3_draft_4Sep2023.pdf", width = 18, height = 7)
ggdraw() +
     draw_plot(p_gene_spikein_scatter+ theme(legend.position="none"),0,1/5, 1/3, 4/5) +
  draw_plot(p_replicate_gene_final_protein + theme(legend.position="none"),1/3,3/5,1/6 , 2/5) +
    draw_plot(p_replicate_gene_final_lncRNA + theme(legend.position="none"),1/2,3/5,1/6 , 2/5) +
    draw_plot(p_protein_coding + theme(legend.position="none"),1/3,1/5,1/6 , 2/5) +
     draw_plot(p_lncRNA + theme(legend.position="none"),1/2,1/5,1/6 , 2/5) +
    draw_plot(as.ggplot(grid.grabExpr(draw(p_heatmap, show_annotation_legend = FALSE, show_heatmap_legend = FALSE)))+ theme(legend.position="none"),2/3,1/5,1/3 , 4/5) +
     draw_plot(as_ggplot(get_legend(p_gene_spikein_scatter)),0,0,1/6,1/5)+
     draw_plot(as_ggplot(get_legend(p_replicate_gene_final_protein)),1/6,0,1/6,1/5)+
     draw_plot(as_ggplot(get_legend(p_protein_coding)),1/3,0,1/6,1/5)+
     draw_plot(as_ggplot(get_legend(p_lncRNA)),1/2,0,1/6,1/5)+
   draw_plot(as.ggplot(grid.grabExpr(color_mapping_legend(hm@ht_list[[1]]@matrix_color_mapping, plot = TRUE))),2/3,0,1/6,1/5)
    
dev.off()
```


# Supplementary Fig. 4

## a spike-in gene scatter-remaining ones
```{r}
plotdata_gene_suppl <- si_data_gene[!(method %in% c("salmon_lr","salmon_sr"))|((method %in% c("salmon_lr","salmon_sr")) & protocol %in% c("directRNA","PacBio"))]
plotdata_gene_suppl[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_gene_spikein_scatter_suppl <- ggplot(plotdata_gene_suppl, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = method), size = 0.5, se = FALSE )+
                geom_point(aes(col = method),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = method,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(paste0(spike_in_general_name_revised,spike_in_version)~protocol)+
                theme_classic()
```

## b spike-in mean absolute errors
```{r}

p_gene_spikein_abe_suppl <- ggplot(si_data_gene, aes(x=ae, fill = method)) + 
    geom_histogram(aes(y=..density..), alpha = 0.5, position = "identity",binwidth = 0.25) +  
      scale_fill_brewer(type = "qual", palette = 3)+
                xlab('Absolute error')+
                facet_grid(paste0(spike_in_general_name_revised,spike_in_version)~protocol)+
                theme_classic()
```


## c spike-in gene metrics between expected and observed
```{r}
pList_gene_no_legend <- lapply(unique(gene_metrics$variable), function(x){
    plotdata_final_metrics <- gene_metrics[(variable == x)]
    plotdata_final_metrics[, na_status := all(is.na(value))|all(value == 0), by = list(spike_in_name)]
   plotdata_final_metrics <- plotdata_final_metrics[which(!na_status)]
   plotdata_final_metrics[, spike_in_name_factor := factor(spike_in_name, levels = c("ERCC",
            "Sequin (v1)","Sequin (v2)","SIRV (E2)","SIRV (E0)","Long SIRV"))]
    if(x %in%c("mae","rmse","mard")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "blue", high = "white") + # red means bad
    geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
             ggtitle(x)+
    xlab("")+
             ylab("")+
   theme_minimal()+
           theme(
               axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }else if(x %in% c("corM","r2")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "white", high = "blue") + # red means bad
    geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
   xlab("")+
             ylab("")+
    #facet_wrap(~variable, scales = "free")+
    theme_minimal()+
           theme( 
       axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }else{
         px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient2(low = "white",
                       mid = "blue",
                       high = "white") +
     geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
             xlab("")+
             ylab("")+
   theme_minimal()+
           theme(
       axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }
    return(px)
})


pList_gene_only_legend <- lapply(unique(gene_metrics$variable), function(x){
    plotdata_final_metrics <- gene_metrics[(variable == x)]
   plotdata_final_metrics[, na_status := all(is.na(value))|all(value == 0), by = list(spike_in_name)]
   plotdata_final_metrics <- plotdata_final_metrics[which(!na_status)]
   plotdata_final_metrics[, spike_in_name_factor := factor(spike_in_name, levels = c("ERCC",
            "Sequin (v1)","Sequin (v2)","SIRV (E2)","SIRV (E0)","Long SIRV"))]
    if(x %in%c("mae","rmse","mard")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "blue", high = "white") + # red means bad
   geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
             ggtitle(x)+
   theme_minimal()
    }else if(x %in% c("corM","r2")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "white", high = "blue") + # red means bad
    geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
    theme_minimal()
    }else{
         px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
 scale_fill_gradient2(low = "white",
                       mid = "blue",
                       high = "white") +
     geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
   theme_minimal()
    }
    return(as_ggplot(get_legend(px)))
})


protocol_annotation <- ggplot(gene_metrics[(variable == "corM")&(spike_in_name == "Sequin (v2)")], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= protocol)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())

method_annotation <- ggplot(Allmetrics_long[(variable == "corM")&(spike_in_name == "Sequin (v2)")], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= method)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())

metric_annotation <- ggplot(gene_metrics[(variable == "corM")&(protocol == "cDNA")&(method == "bambu_lr")&(!is.na(value))], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= spike_in_name)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())


library(gridExtra)


p_gene_spikein_metrics_summarised <- ggarrange( plotlist = pList_gene_no_legend,nrow = 1, ncol =6, align = "hv")
```



## d correlation between long read and short read for spike-in genes and transcripts 
```{r}
gene_mat <- dcast(si_data_gene[method %in% c("salmon_lr","salmon_sr")], spike_in_type + spike_in_name + gene_name ~ protocol, value.var = "normEst")
snames <- unique(gene_mat$spike_in_name)
geneMetric <- do.call("rbind",lapply(seq_along(snames), function(k){
    sname <- snames[k]
    geneCorMetric <- cor(log2(gene_mat[spike_in_name == sname,-c(1:3), with=FALSE]+1), method = "spearman")
    geneCorMetric_melt <- melt(data.table(geneCorMetric,keep.rownames = TRUE), id.vars = "rn", 
                           measure.vars = colnames(geneCorMetric))
    geneCorMetric_melt[, protocol_comparison := paste(sort(c(rn, as.character(variable))), collapse = ""), by = list(rn, variable)]
    geneCorMetric_melt[, duplicated_status := duplicated(protocol_comparison)]
    geneCorMetric_melt <- geneCorMetric_melt[rn != variable &(!duplicated_status)]
    geneCorMetric_melt[, spike_in_name := sname]
    return(geneCorMetric_melt)
}))

# >cor(log2(gene_mat[,-c(1:3), with=FALSE]+1), method = "spearman")
#             Illumina PacBio      cDNA directRNA directcDNA
# Illumina   1.0000000     NA 0.9566994 0.8129958  0.9440563
# PacBio            NA      1        NA        NA         NA
# cDNA       0.9566994     NA 1.0000000 0.8330946  0.9744028
# directRNA  0.8129958     NA 0.8330946 1.0000000  0.8545719
# directcDNA 0.9440563     NA 0.9744028 0.8545719  1.0000000


  


mat_cor_gene <- readRDS(paste0("replicate_gene_expression_comparison_salmon_lr_salmon_sr_25May2023.rds"))

mat_cor_gene[, protocol_comparison := gsub(" vs ","", protocol_comparison)]
geneMetric[, variable := NULL]
setnames(geneMetric, c("spike_in_name","value"),c("variable","spr"))
geneMetric[, spike_in := TRUE]
mat_cor_gene[, spike_in := FALSE]

setnames(mat_cor_gene, "cellLine","variable")
setnames(mat_cor_gene, "r","spr")
geneMetric[, gene := TRUE]
mat_cor_gene[, gene := TRUE]
merge.names <- c("spr","protocol_comparison","variable","spike_in","gene")
corMetricsAll <- do.call("rbind", list(geneMetric[,merge.names, with = FALSE],
                      mat_cor_gene[agg_gene_cluster == "all"&(!grepl("_HepG2",variable)),merge.names, with = FALSE])
                      )


plotdata <- corMetricsAll[grepl("Illumina",protocol_comparison)&(!grepl("PacBio",protocol_comparison))]

plotdata[, spike_in := factor(spike_in, levels = c(TRUE,FALSE), labels = c("Spike-in","Cell lines"))]
p_correlation_lr_sr_spike_in <- ggplot(plotdata[spike_in == "Spike-in"], aes(x = variable, y = spr, fill = variable))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 2)+
    xlab("")+
    ylab("Spr")+
    ggtitle("Spike-in")+
   theme_classic()+
    theme( 
    legend.position = "top")
p_correlation_lr_sr_cellline <- ggplot(plotdata[spike_in == "Cell lines"], aes(x = variable, y = spr, fill = variable))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 2)+
    xlab("")+
    ylab("Spr")+
    ggtitle("Cell lines")+
    theme_classic()+
    theme( 
    legend.position = "top")
p_gene_spikein_correlation_lr_sr_arranged <- ggarrange(p_correlation_lr_sr_spike_in + theme(legend.position = "none"), p_correlation_lr_sr_cellline + theme(legend.position = "none"),
          nrow = 1, align = "hv")

p_gene_spikein_correlation_lr_sr_arranged_legend <- ggarrange(as_ggplot(get_legend(p_correlation_lr_sr_spike_in)), as_ggplot(get_legend(p_correlation_lr_sr_cellline)),
          nrow = 1, align = "hv")
```

## e replicate gene scatter --remaining ones
```{r}
library(GGally)
cellLines <- c("A549","K562","HepG2","Hct116","MCF7","H9","HEYA8")
source(paste0('gene_cluster_code.R'))

temp_data <- comDataGene[cellLine %in% cellLines&(method %in% c("salmon_lr","salmon_sr"))] 
ave_data <- unique(temp_data[, list(meanNormEst = mean(normEst)), by = list(gene_name, protocol_general, cellLine)])
ave_data <-  unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[ave_data, on = "gene_name"]
ave_data[, log2NormEst := log2(meanNormEst+1)]
ave_data[, gene_cluster:=ifelse(gene_biotype %in% tr_gene_list,'TR gene',
                                    ifelse(gene_biotype %in% long_noncoding_rna_list, 'lncRNA',
                                           ifelse(gene_biotype %in% noncoding_rna_list, 'ncRNA',
                                                  ifelse(gene_biotype %in% pseudogene_list,'Pseudogene', ifelse(gene_biotype %in% ig_gene_list, 'IG gene',gene_biotype)))))]
ave_data[, gene_cluster := ifelse(gene_cluster %in% c("IG gene","TR gene", "Mt_tRNA","Mt_rRNA","ribozyme"),"others", gene_cluster)]
    
ave_data[, agg_gene_cluster := ifelse(gene_cluster == "processed_transcript", "lncRNA",
                                                 ifelse(gene_cluster %in% c("ncRNA","Pseudogene"), "others", gene_cluster))]
ave_data[, short_read := grepl("Illumina", protocol_general)]
plotdata1 <- dcast(ave_data[grepl("ENSG", gene_name)], gene_name + agg_gene_cluster + gene_cluster + gene_biotype+ cellLine ~ protocol_general, value.var = "log2NormEst")
plotdata2 <- melt(plotdata1, id.vars = c("gene_name","agg_gene_cluster","gene_biotype","cellLine",
                                         "Illumina"), measure.vars = c("PacBio-SMRTcell","cDNA",
                                         "directcDNA","directRNA"))
plotdata2[, cellLine := factor(cellLine, levels = c("A549","Hct116","HepG2","K562","MCF7","HEYA8","H9"))]
plotdata2[, variable := factor(variable, levels = c("cDNA","directcDNA","directRNA","PacBio-SMRTcell")
                               )]
p_gene_cellLine_scatter_suppl1 <- ggplot(plotdata2[agg_gene_cluster=="protein_coding"], aes(x = value, y = Illumina))+
        geom_abline(intercept = 0, slope = 1)+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(breaks = c(10,100,1000),low = "steelblue1", high = "steelblue4",trans = "log10")+
        geom_density_2d()+
         xlim(c(0,14))+
         ylim(c(0,14))+
         ggpubr::stat_cor(aes(label = after_stat(r.label)),method = "spearman", cor.coef.name = "Sp.R")+
        facet_grid(cellLine~variable)+
        ggtitle("protein coding")+
    theme_classic()

p_gene_cellLine_scatter_suppl2 <- ggplot(plotdata2[agg_gene_cluster=="lncRNA"], aes(x = value, y = Illumina))+
        geom_abline(intercept = 0, slope = 1)+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(breaks = c(10,100,1000),low = "steelblue1", high = "steelblue4",trans = "log10")+
         geom_density_2d()+
         xlim(c(0,14))+
         ylim(c(0,14))+
         ggpubr::stat_cor(aes(label = after_stat(r.label)),method = "spearman", cor.coef.name = "Sp.R")+
        facet_grid(cellLine~variable)+
        ggtitle("lncRNA")+
    theme_classic()
p_gene_cellLine_scatter_suppl <- ggarrange(p_gene_cellLine_scatter_suppl1+ theme(legend.position="none"),
p_gene_cellLine_scatter_suppl2+ theme(legend.position="none"), align = "hv", nrow = 1, ncol =2 )
```

## f replicate gene other metrics other methods (to check and finalize) 
```{r}
methodNamesList <- CJ(lr = c("bambu_lr","salmon_lr"),
                      sr = c("rsem_sr","salmon_sr"),
                      gene = c("gene", "transcript"),
                      metric = c("","mae_","mard_","mard_mod_","rmse_"))
geneCellLineMetrics <- methodNamesList[gene == "gene"]
matCorData <- do.call("rbind",lapply(seq_len(nrow(geneCellLineMetrics)), function(k){
    print(k)
    metric <- geneCellLineMetrics[k]$metric
    saved_date <- ifelse(metric == "", "25May2023","1Aug2023")
    lr <- geneCellLineMetrics[k]$lr
    sr <- geneCellLineMetrics[k]$sr
    mat_cor <- readRDS(paste0("replicateResults/replicate_gene_expression_comparison_",lr,"_",sr,"_",metric,saved_date,".rds"))
mat_cor[, within_cellLine := !grepl("_|all",cellLine)]

mat_cor[, pc_factor := factor(protocol_comparison, rev(unique(mat_cor$protocol_comparison)[c(1,4,2,5,3,6)]))]


if(metric == ""){
    mat_cor[, gc_factor := factor(agg_gene_cluster, c("protein_coding","lncRNA","others","all"))]
    setnames(mat_cor, "r","metricValue")
}else{
    mat_cor[, gc_factor := factor(gene_cluster, c("protein_coding","lncRNA","others","all"))]
    setnames(mat_cor, "mae","metricValue")
}
outData <- mat_cor[!is.na(pc_factor),.(metricValue, gc_factor, pc_factor, within_cellLine,
                                       match_status)]
outData[, method_pair := paste0(lr,sr)]
outData[, metric_type := ifelse(metric == "", "cor",metric)]
return(outData)
}))
saveRDS(matCorData, file = "replicateResults/matCorData_7Aug2023.rds")

matCorData <- readRDS("replicateResults/matCorData_7Aug2023.rds")
methodPairVec <- unique(matCorData$method_pair)
metricVec <- unique(matCorData$metric_type)
matCorData[, short_read := grepl("Illumina", pc_factor)]
matCorData[, dcDNA := grepl("directcDNA",pc_factor)]
matCorData[, dRNA := grepl("directRNA",pc_factor)]
matCorData[, cDNA := grepl("vs cDNA",pc_factor)|grepl("^cDNA",pc_factor)]


matCorData[, new_pc_factor := paste0(c("","Illumina")[short_read+1],
c("","cDNA")[cDNA+1],
c("","dcDNA")[dcDNA+1],
c("","dRNA")[dRNA+1])]

matCorData_gene <- copy(matCorData)
hlineVec <- list(c(0.6,0.8),
                 NULL,
                 NULL,
                NULL,
                NULL
                 )
pList <- lapply(1:5, function(kk){
    p_replicate_gene_metric <-
    ggplot(matCorData[!(method_pair=="salmon_lrsalmon_sr"&(metric_type == "cor"))&(metric_type == metricVec[kk])], aes(new_pc_factor , metricValue, fill = within_cellLine))+
   geom_hline(yintercept = hlineVec[[kk]], col = "grey", linetype = "dashed")+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 4)+
    scale_color_brewer(type = "qual", palette = 4)+
    facet_grid(method_pair~gc_factor, scales = "free")+
    ylab(c("Spearman correlation",
           "MAE",
           "MARD",
           "MRD",
           "RMSE")[kk])+
    xlab("Protocol pairs")+
    theme_classic()
    return(p_replicate_gene_metric)
})


p_metrics_suppl <- ggarrange(plotlist = pList, nrow = 5, ncol = 1, align = "hv")
```




## g gene pca plot with dendrograms using other methods 
```{r}
lrsrList <- CJ(lr = c("bambu_lr","salmon_lr"),
                      sr = c("rsem_sr","salmon_sr"))
comDataGeneFiltered <- comDataGene[method %in% c("bambu_lr","salmon_lr","rsem_sr","salmon_sr")&(!grepl("PacBio",protocol_general))]
p_pca_list <- lapply(seq_len(nrow(lrsrList)), function(k){
    lr <- lrsrList[k]$lr
sr <-  lrsrList[k]$sr
gene <- TRUE
data_type <- "all"
number_of_genes <- 10000
p_pca <- plot_heatmap(dt = comDataGeneFiltered, methodNames = c(lr, sr), gene, data_type,genevec,ensemblAnnotations.transcripts , samples, number_of_genes, runnamevec, plot_type = "pca")
})

```


## arrange supplementary figure
```{r}
p_pca_list_mod <- lapply(seq_along(p_pca_list), function(x) p_pca_list[[x]] + ggtitle(apply(lrsrList,1,paste, collapse = "")[x])+theme(legend.position = "none"))
pdf("suppl_figure2_draft_4Sep2023.pdf", width = 14, height = 20)
ggdraw() +
     draw_plot(p_gene_spikein_scatter_suppl+ theme(legend.position="none"),0,11/16, 1/3, 5/16) +
    draw_plot(p_gene_spikein_abe_suppl+ theme(legend.position="none"),1/3,11/16, 1/3, 5/16) +
  draw_plot(p_gene_spikein_metrics_summarised + theme(legend.position="none"),2/3,13.5/16,1/3,2.5/16) +
   draw_plot(p_gene_spikein_correlation_lr_sr_arranged ,2/3,11/16, 1/3, 2.5/16) +
    draw_plot(p_gene_cellLine_scatter_suppl + theme(legend.position="none"),0,6/16,1 , 5/16) +
    draw_plot(ggarrange(pList[[1]]+theme(legend.position = "none"),
                        pList[[3]]+theme(legend.position = "none"),nrow = 2, ncol =1, align = "hv") + theme(legend.position="none"),0,1/16,1/2 , 5/16) +
     draw_plot(ggarrange(plotlist = p_pca_list_mod, nrow = 2, ncol = 2, align = "hv"),1/2,1/16,1/2, 5/16) +
    draw_plot(as_ggplot(get_legend(p_gene_spikein_scatter_suppl)),0,0,1/8,1/16)+
    draw_plot(as_ggplot(get_legend(p_pca_list[[1]])),1/8,0,1/8,1/16)+
     draw_plot(protocol_annotation,2/8,0,1/8,1/16)+
    draw_plot(method_annotation,3/8,0,1/8,1/16)+
     draw_plot(as_ggplot(get_legend(p_gene_cellLine_scatter_suppl1)),1/2,0,1/8,1/16)+
    draw_plot(as_ggplot(get_legend(p_gene_cellLine_scatter_suppl2)),5/8,0,1/8,1/16)+
    draw_plot(ggarrange( plotlist = pList_gene_only_legend,nrow = 2, ncol =3, align = "hv"),3/4,0,1/8,1/16)+
     draw_plot(p_gene_spikein_correlation_lr_sr_arranged_legend,7/8,0,1/8,1/16)
    
dev.off()

pdf("pca_gene_expression_list.pdf", width = 10, height = 8)
ggarrange(plotlist = p_pca_list, nrow = 2, ncol = 2)
dev.off()
```

## Supplementary Table 
```{r}
gene_metrics_wide <- dcast(gene_metrics, method + protocol + spike_in_type + spike_in_name ~ variable, value.var = "value")
tx_metrics_wide <- dcast(tx_metrics, method + protocol + spike_in_type + spike_in_name ~ variable, value.var = "value")
write.table(gene_metrics_wide, file = "suppl_table3_1.txt",
             col.names = TRUE, row.names = FALSE, sep = ",", quote = FALSE)
write.table(tx_metrics_wide, file = "suppl_table3_2.txt",
             col.names = TRUE, row.names = FALSE, sep = ",", quote = FALSE)
```


# r2.1 all methods spearman correlation
```{r}
p_sprcor_gene <- ggplot(gene_metrics[variable == "corM"&(protocol != "directRNA")&(protocol!= "PacBio")], aes(x = factor(method, levels = c("salmon_lr","bambu_lr","NanoCount_lr","salmon_sr","rsem_sr")) , y = value))+
    geom_boxplot()+geom_point(aes(col = spike_in_name), position = "jitter", size = 2, shape = 16)+
     scale_color_brewer(type = "qual", palette = 3)+
    theme_classic()
p_lrsr_gene <- ggplot(matCorData_gene[metric_type == "cor"&(within_cellLine==TRUE)], aes(x = factor(gc_factor, levels = c("protein_coding","lncRNA","others","all")), y = metricValue, fill = method_pair))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 3)+
    theme_classic()


p_sprcor_tx <- ggplot(tx_metrics[variable == "corM"&(protocol != "directRNA")&(protocol!= "PacBio")], aes(x = factor(method, levels = c("salmon_lr","bambu_lr","NanoCount_lr","salmon_sr","rsem_sr")) , y = value))+
    geom_boxplot()+geom_point(aes(col = spike_in_name), position = "jitter", size = 2, shape = 16)+
     scale_color_brewer(type = "qual", palette = 3)+
    theme_classic()
p_lrsr_tx <- ggplot(matCorData_tx[metric_type == "cor"&(gc_factor == "protein_coding")], aes(x = ct_factor, y = metricValue, fill = method_pair))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 3)+
    theme_classic()
```



```{r}
pdf("FigureR2.pdf",width = 7, height = 5)
ggdraw() +
     draw_plot(p_sprcor_gene+ theme(legend.position="none"),0,3/5, 1/3, 2/5) +
    draw_plot(p_lrsr_gene+ theme(legend.position="none"),1/3,3/5, 2/3, 2/5) +
      draw_plot(p_sprcor_tx+ theme(legend.position="none"),0,1/5, 1/3, 2/5) +
    draw_plot(p_lrsr_tx+ theme(legend.position="none"),1/3,1/5, 2/3, 2/5) +
     draw_plot(as_ggplot(get_legend(p_sprcor_gene)),0,0,1/4,1/5)+
     draw_plot(as_ggplot(get_legend(p_lrsr_gene)),1/4,0,1/4,1/5)+
    draw_plot(as_ggplot(get_legend(p_sprcor_gene)),1/2,0,1/4,1/5)+
     draw_plot(as_ggplot(get_legend(p_lrsr_gene)),3/4,0,1/4,1/5)
    
dev.off()
```

# Supplementary Text Fig. 14
### r7 replicate correlation 
#### all genes
```{r}
temp <- comDataGeneTrimmed[method %in% c("salmon_lr","salmon_sr","trim_lr_sim")&(runname %in% runnamevec)]
temp[, runname := gsub("GIS_K562_Illumina_Rep5-Run1","GIS_k562_Illumina_Rep5-Run1", runname)]
setnames(temp, "runname","rn")
temp[grepl("GIS", rn ),rn := samples[match(rn, old_runname)]$runname]
dt_wide <- dcast(temp, gene_name ~ paste0(rn,method), value.var = "normEst")
    corMatrix <- cor(log2(dt_wide[,-1,with = FALSE]), method = "spearman", use = "pairwise.complete.obs")
    
    corMatrix_melt <- melt(data.table(corMatrix,keep.rownames = TRUE), id.vars = "rn", 
                           measure.vars = colnames(corMatrix))
    corMatrix_melt[, rn_var := paste(sort(c(rn, as.character(variable))), collapse = "_"), by = list(rn, variable)]
    corMatrix_melt[, duplicated_status := duplicated(rn_var)]
    corMatrix_melt <- corMatrix_melt[rn != variable &(!duplicated_status)]
    
    corMatrix_melt[, `:=`(cellLine_1 = unlist(strsplit(rn, "\\_"))[2],
                          protocol_1 = gsub("Stranded","",unlist(strsplit(rn, "\\_"))[3]),
                          bio_rep1 = gsub("replicate","",unlist(strsplit(rn, "\\_"))[4]),
                          tech_rep1 = gsub("run|salmon|trim","",unlist(strsplit(rn, "\\_"))[5]),
                          method1 = ifelse(grepl("salmon_sr",rn),"salmon_sr",
                                           ifelse(grepl("salmon_lr",rn),"salmon_lr","lr_sim"))), by = rn]
    corMatrix_melt[, `:=`(cellLine_2 = unlist(strsplit(as.character(variable), "\\_"))[2],
                          protocol_2 =  gsub("Stranded","",unlist(strsplit(as.character(variable), "\\_"))[3]),
                          bio_rep2 = gsub("replicate","",unlist(strsplit(as.character(variable), "\\_"))[4]),
                          tech_rep2 = gsub("run|salmon|trim","",unlist(strsplit(as.character(variable), "\\_"))[5]),
                          method2 = ifelse(grepl("salmon_sr",variable),"salmon_sr",
                                           ifelse(grepl("salmon_lr",variable),"salmon_lr","lr_sim"))), by = variable]
    

  plotdata <- corMatrix_melt[cellLine_1==cellLine_2&(protocol_1 == protocol_2)&(method1==method2)]
  plotdata[, match_status := (bio_rep1 == bio_rep2)]
  plotdata[, new_protocol_variable := ifelse(grepl("sim",method1),"lr_sim",protocol_1)]
  
  geneCorAll <- copy(plotdata)
  geneCorAll[, feature := "gene"]
  geneCorAll[, subset := FALSE]
```

#### gene protein coding
```{r}
temp <- comDataGeneTrimmed[method %in% c("salmon_lr","salmon_sr","trim_lr_sim")&(runname %in% runnamevec)&(gene_name %in% unique(ensemblAnnotations.transcripts[gene_biotype == "protein_coding"]$gene_name))]
temp[, runname := gsub("GIS_K562_Illumina_Rep5-Run1","GIS_k562_Illumina_Rep5-Run1", runname)]
setnames(temp, "runname","rn")
temp[grepl("GIS", rn ),rn := samples[match(rn, old_runname)]$runname]
dt_wide <- dcast(temp, gene_name ~ paste0(rn,method), value.var = "normEst")
corMatrix <- cor(log2(dt_wide[,-1,with = FALSE]), method = "spearman", use = "pairwise.complete.obs")
    corMatrix_melt <- melt(data.table(corMatrix,keep.rownames = TRUE), id.vars = "rn", 
                           measure.vars = colnames(corMatrix))
    corMatrix_melt[, rn_var := paste(sort(c(rn, as.character(variable))), collapse = "_"), by = list(rn, variable)]
    corMatrix_melt[, duplicated_status := duplicated(rn_var)]
    corMatrix_melt <- corMatrix_melt[rn != variable &(!duplicated_status)]
    
    corMatrix_melt[, `:=`(cellLine_1 = unlist(strsplit(rn, "\\_"))[2],
                          protocol_1 = gsub("Stranded","",unlist(strsplit(rn, "\\_"))[3]),
                          bio_rep1 = gsub("replicate","",unlist(strsplit(rn, "\\_"))[4]),
                          tech_rep1 = gsub("run|salmon|trim","",unlist(strsplit(rn, "\\_"))[5]),
                          method1 = ifelse(grepl("salmon_sr",rn),"salmon_sr",
                                           ifelse(grepl("salmon_lr",rn),"salmon_lr","lr_sim"))), by = rn]
    corMatrix_melt[, `:=`(cellLine_2 = unlist(strsplit(as.character(variable), "\\_"))[2],
                          protocol_2 =  gsub("Stranded","",unlist(strsplit(as.character(variable), "\\_"))[3]),
                          bio_rep2 = gsub("replicate","",unlist(strsplit(as.character(variable), "\\_"))[4]),
                          tech_rep2 = gsub("run|salmon|trim","",unlist(strsplit(as.character(variable), "\\_"))[5]),
                          method2 = ifelse(grepl("salmon_sr",variable),"salmon_sr",
                                           ifelse(grepl("salmon_lr",variable),"salmon_lr","lr_sim"))), by = variable]
  plotdata <- corMatrix_melt[cellLine_1==cellLine_2&(protocol_1 == protocol_2)&(method1==method2)]
  plotdata[, match_status := (bio_rep1 == bio_rep2)]
  plotdata[, new_protocol_variable := ifelse(grepl("sim",method1),"lr_sim",protocol_1)]
  
  geneCorProtein <- copy(plotdata)
  geneCorProtein[, feature := "gene"]
  geneCorProtein[, subset := TRUE]
```


#### all tx
```{r}
temp <- comDataTranscriptTrimmed[method %in% c("salmon_lr","salmon_sr","lr_sim")&(runname %in% runnamevec)&(gene_name %in% unique(ensemblAnnotations.transcripts[gene_biotype == "protein_coding"]$gene_name))]
temp[, runname := gsub("GIS_K562_Illumina_Rep5-Run1","GIS_k562_Illumina_Rep5-Run1", runname)]
setnames(temp, "runname","rn")
temp[grepl("GIS", rn ),rn := samples[match(rn, old_runname)]$runname]
dt_wide <- dcast(temp, gene_name + tx_name ~ paste0(rn, method), value.var = "normEst")
    corMatrix <- cor(log2(dt_wide[,-c(1,2),with = FALSE]), method = "spearman", use = "pairwise.complete.obs")
    
corMatrix_melt <- melt(data.table(corMatrix,keep.rownames = TRUE), id.vars = "rn", 
                           measure.vars = colnames(corMatrix))
corMatrix_melt[, rn_var := paste(sort(c(rn, as.character(variable))), collapse = "_"), by = list(rn, variable)]
corMatrix_melt[, duplicated_status := duplicated(rn_var)]
corMatrix_melt <- corMatrix_melt[rn != variable &(!duplicated_status)]
    
corMatrix_melt[, `:=`(cellLine_1 = unlist(strsplit(rn, "\\_"))[2],
                          protocol_1 = gsub("Stranded","",unlist(strsplit(rn, "\\_"))[3]),
                          bio_rep1 = gsub("replicate","",unlist(strsplit(rn, "\\_"))[4]),
                          tech_rep1 = gsub("run|salmon|lr","",unlist(strsplit(rn, "\\_"))[5]),
                          method1 = ifelse(grepl("salmon_sr",rn),"salmon_sr",
                                           ifelse(grepl("salmon_lr",rn),"salmon_lr","lr_sim"))), by = rn]
    corMatrix_melt[, `:=`(cellLine_2 = unlist(strsplit(as.character(variable), "\\_"))[2],
                          protocol_2 =  gsub("Stranded","",unlist(strsplit(as.character(variable), "\\_"))[3]),
                          bio_rep2 = gsub("replicate","",unlist(strsplit(as.character(variable), "\\_"))[4]),
                          tech_rep2 = gsub("run|salmon|lr","",unlist(strsplit(as.character(variable), "\\_"))[5]),
                          method2 = ifelse(grepl("salmon_sr",variable),"salmon_sr",
                                           ifelse(grepl("salmon_lr",variable),"salmon_lr","lr_sim"))), by = variable]
  plotdata <- corMatrix_melt[cellLine_1==cellLine_2&(protocol_1 == protocol_2)&(method1==method2)]
  plotdata[, match_status := (bio_rep1 == bio_rep2)]
  plotdata[, new_protocol_variable := ifelse(grepl("sim",method1),"lr_sim",protocol_1)]
  txCorAll <- copy(plotdata)
  txCorAll[, feature := "transcript"]
  txCorAll[, subset := FALSE]
```


#### major tx 
```{r}
filtered_data <- comDataTranscript[method %in% c("salmon_lr","salmon_sr")&(gene_name %in% unique(ensemblAnnotations.transcripts[gene_biotype == "protein_coding"]$gene_name))]

dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")
dominant_typeData[, short_read := as.numeric(as.logical(short_read))]
filtered_data <- dominant_typeData[filtered_data, on = c("tx_name","gene_name","cellLine","short_read")]
filtered_data <- filtered_data[runname %in% runnamevec]
# by expression level
corMatrix_melt <- do.call("rbind",lapply(3, function(xx){
    if(xx == 1){
        dt_wide <- dcast(filtered_data[which(majorLongRead)], gene_name + tx_name ~ runname, value.var = "normEst")
    }else if(xx == 2){
        dt_wide <- dcast(filtered_data[which(majorShortRead)], gene_name + tx_name ~ runname, value.var = "normEst")
    }else if(xx == 3){
        dt_wide <- dcast(filtered_data[which(majorBoth)], gene_name + tx_name ~ runname, value.var = "normEst")
    }else if(xx == 4){
        
        cols <- colnames(dt_wide)[-c(1:2)]
        dt_wide[ , (cols) := lapply(.SD, function(kkk) ifelse(kkk<30,NA,kkk)), .SDcols =cols]
    }else{
        dt_wide <- dcast(filtered_data, gene_name + tx_name ~ runname, value.var = "normEst")
    }
    
# expression filtering directly
corMatrix <- cor(log2(dt_wide[,-c(1,2),with = FALSE]), method = "spearman", use = "pairwise.complete.obs")
    
    corMatrix_melt <- melt(data.table(corMatrix,keep.rownames = TRUE), id.vars = "rn", 
                           measure.vars = colnames(corMatrix))
    corMatrix_melt[grepl("GIS", rn), rn := samples[match(gsub("GIS_K562_Illumina_Rep5-Run1","GIS_k562_Illumina_Rep5-Run1",rn), old_runname)]$runname]
    corMatrix_melt[grepl("GIS", variable), variable := samples[match(gsub("GIS_K562_Illumina_Rep5-Run1","GIS_k562_Illumina_Rep5-Run1",variable), old_runname)]$runname]
    corMatrix_melt[, rn_var := paste(sort(c(rn, as.character(variable))), collapse = "_"), by = list(rn, variable)]
    corMatrix_melt[, duplicated_status := duplicated(rn_var)]
    corMatrix_melt <- corMatrix_melt[rn != variable &(!duplicated_status)]
    
    corMatrix_melt[, `:=`(cellLine_1 = unlist(strsplit(rn, "\\_"))[2],
                          protocol_1 = gsub("Stranded","",unlist(strsplit(rn, "\\_"))[3]),
                          bio_rep1 = gsub("replicate","",unlist(strsplit(rn, "\\_"))[4]),
                          tech_rep1 = gsub("run","",unlist(strsplit(rn, "\\_"))[5]),
                          method1 = ifelse(grepl("Illumina",rn),"salmon_sr","salmon_lr")), by = rn]
    corMatrix_melt[, `:=`(cellLine_2 = unlist(strsplit(as.character(variable), "\\_"))[2],
                          protocol_2 =  gsub("Stranded","",unlist(strsplit(as.character(variable), "\\_"))[3]),
                          bio_rep2 = gsub("replicate","",unlist(strsplit(as.character(variable), "\\_"))[4]),
                          tech_rep2 = gsub("run","",unlist(strsplit(as.character(variable), "\\_"))[5]),
                          method2 = ifelse(grepl("Illumina",variable),"salmon_sr","salmon_lr")), by = variable]
    corMatrix_melt[, major := c("majorLongRead","majorShortRead","majorBoth","by_expression","no_filtering")[xx]]
    return(corMatrix_melt)
}))

 plotdata <- corMatrix_melt[cellLine_1==cellLine_2&(protocol_1 == protocol_2)&(major == "majorBoth")]
 plotdata[, match_status := (bio_rep1 == bio_rep2)]
 plotdata[, new_protocol_variable := ifelse(grepl("sim",method1),"lr_sim",protocol_1)]
 plotdata[, major := NULL]
 txCorMajor <- copy(plotdata)
  txCorMajor[, feature := "transcript"]
  txCorMajor[, subset := TRUE]
```

# Supplementary Text Fig. 3
### % of reads with both primers 
```{r}
stats.pychopper <- fread("adapter_analysis/pychopper_results/cdna/statistics.tsv", header = TRUE)

stats.pychopper[Category == "Hits"]

stats.pychopper[Category == "Hits"&(grepl("VNP,-SSP",Name)&!grepl("-VNP|(,SSP)|(^SSP)|(-SSP,VNP)",Name))] # 57240016
stats.pychopper[Category == "Hits"&(grepl("SSP,-VNP",Name)&!grepl("-SSP|(,VNP)|(^VNP)|(-VNP,SSP)",Name))] #49174479


stats.pychopper[Category == "Hits"&(grepl("VNP,-SSP",Name)|grepl("SSP,-VNP",Name))&(!grepl("VNP,-SSP,SSP,-VNP",Name))&(!grepl("VNP,-SSP,SSP,-VNP",Name)))]

stats.pychopper[Category == "Hits"&((grepl("VNP,-SSP",Name)&!grepl("SSP,-VNP",Name))|(!grepl("VNP,-SSP",Name)&grepl("SSP,-VNP",Name))|(grepl("^SSP",Name)&(grepl("-VNP$",Name)))|(grepl("-SSP$",Name)&(grepl("^VNP",Name))))]
```

### r11 all metrics heatmap + histograms 
```{r}
pdf("figure_r3.11_draft_12Oct2023.pdf", width = 14, height = 20)
ggdraw() +
     draw_plot(p_gene_spikein_metrics_summarised,0,4/5,1,1/5) +
     draw_plot(p_tx_spikein_metrics_summarised,0,3/5,1,1/5) +
    draw_plot(p_gene_spikein_abe_suppl+ theme(legend.position="none"),0,2/5, 1, 1/5) +
     draw_plot(p_tx_spikein_abe_suppl+ theme(legend.position="none"),0,1/5,1, 1/5) +
    draw_plot(as_ggplot(get_legend(p_gene_spikein_abe_suppl)),2/3,1/10,1/3,1/10)+
     draw_plot(protocol_annotation,0,1/10,1/3,1/10)+
    draw_plot(method_annotation,1/3,1/10,1/3,1/10)+
    draw_plot(ggarrange( plotlist = pList_gene_only_legend,nrow = 1, ncol =6, align = "hv"),0,0,1/2,1/10)+
     draw_plot(ggarrange( plotlist = pList_tx_only_legend,nrow = 1, ncol =6, align = "hv"),1/2,0,1/2,1/10)
dev.off()
```

## r3.15 metrics plots showing human cell line gene and transcript level results similar for all method pairs 
```{r}
matCorData <- readRDS("replicateResults/matCorData_7Aug2023.rds")
methodPairVec <- unique(matCorData$method_pair)
metricVec <- unique(matCorData$metric_type)
matCorData[, short_read := grepl("Illumina", pc_factor)]
matCorData[, dcDNA := grepl("directcDNA",pc_factor)]
matCorData[, dRNA := grepl("directRNA",pc_factor)]
matCorData[, cDNA := grepl("vs cDNA",pc_factor)|grepl("^cDNA",pc_factor)]


matCorData[, new_pc_factor := paste0(c("","Illumina")[short_read+1],
c("","cDNA")[cDNA+1],
c("","dcDNA")[dcDNA+1],
c("","dRNA")[dRNA+1])]

matCorData_gene <- copy(matCorData)
hlineVec <- list(c(0.6,0.8),
                 NULL,
                 NULL,
                NULL,
                NULL
                 )
pList <- lapply(1:5, function(kk){
    p_replicate_gene_metric <-
    ggplot(matCorData[(metric_type == metricVec[kk])&(within_cellLine == TRUE)], aes(new_pc_factor , metricValue))+
    geom_hline(yintercept = hlineVec[[kk]], col = "grey", linetype = "dashed")+
    geom_boxplot()+
    facet_grid(method_pair ~ gc_factor, scales = "free")+
    ylab(c("Spearman correlation",
           "MAE",
           "MARD",
           "MRD",
           "RMSE")[kk])+
    xlab("Protocol pairs")+
    theme_classic()
    return(p_replicate_gene_metric)
})

p_metrics_suppl <- ggarrange(plotlist = pList, nrow = 5, ncol = 1, align = "hv")
```


```{r}
matCorData <- readRDS("matCorDataTx_7Aug2023.rds")
methodPairVec <- unique(matCorData$method_pair)
metricVec <- unique(matCorData$metric_type)
matCorData[, short_read := grepl("Illumina", pc_factor)]
matCorData[, dcDNA := grepl("directcDNA",pc_factor)]
matCorData[, dRNA := grepl("directRNA",pc_factor)]
matCorData[, cDNA := grepl("vs cDNA",pc_factor)|grepl("^cDNA",pc_factor)]


matCorData[, new_pc_factor := paste0(c("","Illumina")[short_read+1],
c("","cDNA")[cDNA+1],
c("","dcDNA")[dcDNA+1],
c("","dRNA")[dRNA+1])]

matCorData_tx <- copy(matCorData)

methodPairVec <- unique(matCorData$method_pair)
library(ggpubr)

## new split by metric
p_metric_list <- lapply(1:5, function(k){
    
    my_comparisons <- list(c("majorBoth","majorLongReadOnly"),
                       c("majorLongReadOnly","majorShortReadOnly"),
                       c("majorShortReadOnly","others"))
labeldata <- unique(matCorData[, list(med_value = median(metricValue)),
                                   by = list(ct_factor,short_read,method_pair, metric_type)])
labeldata[, metricValue := ifelse(metric_type=="cor",0.93, 
                         ifelse(metric_type == "mae_",1.8,
                                            ifelse(metric_type == "mard_",1.8,
                                                   ifelse(metric_type == "mard_mod_",1.2,
                                                          ifelse(metric_type == "rmse_",1.8,NA)))))]
plotdata <- matCorData[(metric_type %in% metricVec[k])]
plotdata[, short_read := factor(short_read, labels = rev(c("short read vs long read","long read vs long read")))]
plotdata[, labelY := ifelse(metric_type=="cor",0.93, 
                         ifelse(metric_type == "mae_",1.8,
                                            ifelse(metric_type == "mard_",1.8,
                                                   ifelse(metric_type == "mard_mod_",1.2,
                                                          ifelse(metric_type == "rmse_",1.8,NA)))))]
labelData <- labeldata[(metric_type %in% metricVec[k])]
labelData[, short_read := factor(short_read, labels = c("short read vs long read","long read vs long read"))]
p_replicate_tx_final_overall <- ggplot(plotdata, aes(x = ct_factor, y = metricValue))+
    geom_boxplot(outlier.size = 1, outlier.color = "grey", outlier.shape = 16)+
    xlab("Protocol comparison")+
    stat_compare_means(comparisons = my_comparisons,
                       paired = FALSE,
                       #label.y = c(1,0.95,0.9,0.85,0.8),
                      aes(label.y = labelY))+
   
    facet_grid(method_pair~short_read , scales = "free_y")+ 
     geom_text(data=labelData, aes(label = round(med_value,2)))+
    ggtitle(c("Spearman correlation","MAE",
           "MARD","MRD",
           "RMSE")[k])+
    theme_classic()+
    theme(legend.position = "top")
    return(p_replicate_tx_final_overall)
})


p_metric_list_mod <- lapply(seq_along(p_metric_list), function(x) p_metric_list[[x]] + theme(legend.position = "none"))
```


```{r}
pdf("r3.15.pdf", width = 14, height = 14)
ggarrange(pList[[1]]+theme(legend.position = "none"),
                        pList[[3]]+theme(legend.position = "none"),
                        p_metric_list[[1]]+theme(legend.position = "none"),
                        p_metric_list[[3]]+theme(legend.position = "none"),nrow = 2, ncol = 2, align = "hv")
dev.off()
```


# Supplementary Text Fig. 19
## spike in split by CPM < 2.5 and above 2.5
```{r}
plotdata_gene_suppl <- si_data_gene[cpm_norm_conc < 2.5]
plotdata_gene_suppl[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_gene_spikein_scatter_suppl_cpm0_2.5 <- ggplot(plotdata_gene_suppl, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
               geom_point(aes(col = method),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = method,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(paste0(spike_in_general_name_revised,spike_in_version)~protocol)+
               theme_classic()
pdf("gene_metrics_scatterplots_cpm0_2.5_16Feb2024.pdf", width = 17, height = 9)
p_gene_spikein_scatter_suppl_cpm0_2.5
dev.off()

plotdata_gene_suppl <- si_data_gene[cpm_norm_conc >= 2.5]
plotdata_gene_suppl[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_gene_spikein_scatter_suppl_cpm2.5 <- ggplot(plotdata_gene_suppl, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = method), size = 0.5, se = FALSE )+
                geom_point(aes(col = method),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = method,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(paste0(spike_in_general_name_revised,spike_in_version)~protocol)+
               theme_classic()



plotdata_gene <- si_data_gene[method %in% c("salmon_lr","salmon_sr")&!(protocol %in% c("directRNA","PacBio"))&(cpm_norm_conc<2.5)]
plotdata_gene[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_gene_spikein_scatter_cpm0_2.5 <- ggplot(plotdata_gene, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = spike_in_version), size = 0.5, se = FALSE )+
                geom_point(aes(col = spike_in_version),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = spike_in_version,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(spike_in_general_name_revised~protocol)+
                 theme_classic()

plotdata_gene <- si_data_gene[method %in% c("salmon_lr","salmon_sr")&!(protocol %in% c("directRNA","PacBio"))&(cpm_norm_conc<2.5)]
plotdata_gene[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_gene_spikein_scatter_cpm0_2.5 <- ggplot(plotdata_gene, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = spike_in_version), size = 0.5, se = FALSE )+
                geom_point(aes(col = spike_in_version),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = spike_in_version,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(spike_in_general_name_revised~protocol)+
                # geom_smooth(method = "lm")+
                theme_classic()
```

```{r}
rsq <- function(x, y) summary(lm(y~x))$r.squared #https://stackoverflow.com/questions/40901445/function-to-calculate-r2-r-squared-in-r
si_data[, new_spike_in_name := ifelse(spike_in_name %in% c("SIRV (E0)","Long SIRV"), "SIRV (E0+Long)",spike_in_name)]
Allmetrics_cpm0_2.5 <- unique(si_data[cpm_norm_conc >= 2.5, list(corM = cor(log2(cpm_norm_conc+1), log2(normEst+1),  method = "spearman"),
                                    corP = cor(log2(cpm_norm_conc+1), log2(normEst+1)),
                                    mae = mean(abs(log2(cpm_norm_conc+1)-log2(normEst+1))),
                                    rmse = sqrt(mean((log2(cpm_norm_conc+1)-log2(normEst+1))^2)),
                                    mard = mean(abs(log2(cpm_norm_conc+1)-log2(normEst+1))/(log2(cpm_norm_conc+1)+log2(normEst+1))*2, na.rm = TRUE),
                                    mard_mod =  mean((log2(cpm_norm_conc+1)-log2(normEst+1))/(log2(cpm_norm_conc+1)+log2(normEst+1))*2, na.rm = TRUE),
                                    r2 = rsq(log2(cpm_norm_conc+1),log2(normEst+1))), by = list(method, protocol, spike_in_type, spike_in_name)])
library(hrbrthemes)
library(ggpubr)

Allmetrics_long_cpm0_2.5 <- melt(Allmetrics_cpm0_2.5, id.vars = c("method","protocol","spike_in_type","spike_in_name"), measure.vars = c("corM","corP","mae","rmse","mard","mard_mod","r2"))

gene_metrics_cpm0_2.5 <- copy(Allmetrics_long_cpm0_2.5)
```

```{r}
pList_gene_no_legend_cpm0_2.5 <- lapply(unique(gene_metrics_cpm0_2.5$variable), function(x){
    plotdata_final_metrics <- gene_metrics_cpm0_2.5[(variable == x)]
    plotdata_final_metrics[, na_status := all(is.na(value))|all(value == 0), by = list(spike_in_name)]
   plotdata_final_metrics <- plotdata_final_metrics[which(!na_status)]
   plotdata_final_metrics[, spike_in_name_factor := factor(spike_in_name, levels = c("ERCC",
            "Sequin (v1)","Sequin (v2)","SIRV (E2)","SIRV (E0)","Long SIRV"))]
    if(x %in%c("mae","rmse","mard")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "blue", high = "white") + # red means bad
    geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
             ggtitle(x)+
    xlab("")+
             ylab("")+
   theme_minimal()+
           theme( 
        axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }else if(x %in% c("corM","r2")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "white", high = "blue") + # red means bad
    geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
   xlab("")+
             ylab("")+
   theme_minimal()+
           theme( 
     axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }else{
         px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
 scale_fill_gradient2(low = "white",
                       mid = "blue",
                       high = "white") +
     geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
             xlab("")+
             ylab("")+
    theme_minimal()+
           theme(
      axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }
    return(px)
})


pList_gene_only_legend_cpm0_2.5 <- lapply(unique(gene_metrics_cpm0_2.5$variable), function(x){
    plotdata_final_metrics <- gene_metrics_cpm0_2.5[(variable == x)]
   plotdata_final_metrics[, na_status := all(is.na(value))|all(value == 0), by = list(spike_in_name)]
   plotdata_final_metrics <- plotdata_final_metrics[which(!na_status)]
   plotdata_final_metrics[, spike_in_name_factor := factor(spike_in_name, levels = c("ERCC",
            "Sequin (v1)","Sequin (v2)","SIRV (E2)","SIRV (E0)","Long SIRV"))]
    if(x %in%c("mae","rmse","mard")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "blue", high = "white") + # red means bad
   geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
             ggtitle(x)+
    theme_minimal()
    }else if(x %in% c("corM","r2")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "white", high = "blue") + # red means bad
   geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
    theme_minimal()
    }else{
         px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
   scale_fill_gradient2(low = "white",
                       mid = "blue",
                       high = "white") +
     geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
    theme_minimal()
    }
    return(as_ggplot(get_legend(px)))
})


protocol_annotation <- ggplot(gene_metrics_cpm0_2.5[(variable == "corM")&(spike_in_name == "Sequin (v2)")], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= protocol)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())

method_annotation <- ggplot(Allmetrics_long_cpm0_2.5[(variable == "corM")&(spike_in_name == "Sequin (v2)")], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= method)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())

metric_annotation <- ggplot(gene_metrics_cpm0_2.5[(variable == "corM")&(protocol == "cDNA")&(method == "bambu_lr")&(!is.na(value))], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= spike_in_name)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())

library(gridExtra)

pdf("gene_metrics_heatmap_cpm0_2.5_16Feb2024.pdf", width = 18, height = 9)
p_gene_spikein_metrics_summarised_cpm0_2.5 <- ggarrange( plotlist = pList_gene_no_legend_cpm0_2.5,nrow = 1, ncol =7, align = "hv")
p_gene_spikein_metrics_summarised_cpm0_2.5
dev.off()

pdf("gene_metrics_heatmap_cpm0_2.5_16Feb2024_legend.pdf")
pList_gene_only_legend_cpm0_2.5
dev.off()
```


# Supplementary Fig. 4 and 5 
```{r}
gene_mat <- dcast(si_data_gene[method %in% c("salmon_lr","salmon_sr")], spike_in_type + spike_in_name + gene_name ~ protocol, value.var = "normEst")
snames <- unique(gene_mat$spike_in_name)
geneMetric <- do.call("rbind",lapply(seq_along(snames), function(k){
    sname <- snames[k]
    geneCorMetric <- cor(log2(gene_mat[spike_in_name == sname,-c(1:3), with=FALSE]+1), method = "spearman")
    geneCorMetric_melt <- melt(data.table(geneCorMetric,keep.rownames = TRUE), id.vars = "rn", 
                           measure.vars = colnames(geneCorMetric))
    geneCorMetric_melt[, protocol_comparison := paste(sort(c(rn, as.character(variable))), collapse = ""), by = list(rn, variable)]
    geneCorMetric_melt[, duplicated_status := duplicated(protocol_comparison)]
    geneCorMetric_melt <- geneCorMetric_melt[rn != variable &(!duplicated_status)]
    geneCorMetric_melt[, spike_in_name := sname]
    return(geneCorMetric_melt)
}))
transcript_mat <- dcast(si_data_transcript[method %in% c("salmon_lr","salmon_sr")], spike_in_type + spike_in_name + gene_name + tx_name ~ protocol, value.var = "normEst")
snames <- unique(transcript_mat$spike_in_name)
txMetric <- do.call("rbind",lapply(seq_along(snames), function(k){
    sname <- snames[k]
    geneCorMetric <- cor(log2(transcript_mat[spike_in_name == sname,-c(1:4), with=FALSE]+1), method = "spearman")
    geneCorMetric_melt <- melt(data.table(geneCorMetric,keep.rownames = TRUE), id.vars = "rn",
                           measure.vars = colnames(geneCorMetric))
    geneCorMetric_melt[, protocol_comparison := paste(sort(c(rn, as.character(variable))), collapse = ""), by = list(rn, variable)]
    geneCorMetric_melt[, duplicated_status := duplicated(protocol_comparison)]
    geneCorMetric_melt <- geneCorMetric_melt[rn != variable &(!duplicated_status)]
    geneCorMetric_melt[, spike_in_name := sname]
    return(geneCorMetric_melt)
}))
# >cor(log2(gene_mat[,-c(1:3), with=FALSE]+1), method = "spearman")
#             Illumina PacBio      cDNA directRNA directcDNA
# Illumina   1.0000000     NA 0.9566994 0.8129958  0.9440563
# PacBio            NA      1        NA        NA         NA
# cDNA       0.9566994     NA 1.0000000 0.8330946  0.9744028
# directRNA  0.8129958     NA 0.8330946 1.0000000  0.8545719
# directcDNA 0.9440563     NA 0.9744028 0.8545719  1.0000000


  


mat_cor_gene <- readRDS(paste0("replicate_gene_expression_comparison_salmon_lr_salmon_sr_25May2023.rds"))

mat_cor_gene[, protocol_comparison := gsub(" vs ","", protocol_comparison)]
geneMetric[, variable := NULL]
setnames(geneMetric, c("spike_in_name","value"),c("variable","spr"))
geneMetric[, spike_in := TRUE]
mat_cor_gene[, spike_in := FALSE]

setnames(mat_cor_gene, "cellLine","variable")
setnames(mat_cor_gene, "r","spr")
geneMetric[, gene := TRUE]
mat_cor_gene[, gene := TRUE]


mat_cor_tx <- readRDS(paste0("replicate_transcript_expression_comparison_salmon_lr_salmon_sr_25May2023.rds"))
mat_cor_tx[, protocol_comparison := gsub(" vs ","", protocol_comparison)]

txMetric[, variable := NULL]

setnames(txMetric, c("spike_in_name","value"),c("variable","spr"))
txMetric[, spike_in := TRUE]
mat_cor_tx[, spike_in := FALSE]

setnames(mat_cor_tx, "cellLine","variable")
txMetric[, gene := FALSE]
mat_cor_tx[, gene := FALSE]

merge.names <- c("spr","protocol_comparison","variable","spike_in","gene")
corMetricsAll <- do.call("rbind", list(geneMetric[,merge.names, with = FALSE],
                      mat_cor_gene[agg_gene_cluster == "all"&(!grepl("_HepG2",variable)),merge.names, with = FALSE],
                      txMetric[,merge.names, with = FALSE],
                      mat_cor_tx[agg_gene_cluster == "all"&(common_type == "all")&(!grepl("_HepG2",variable)),merge.names, with = FALSE]))


plotdata <- corMetricsAll[grepl("Illumina",protocol_comparison)&(!grepl("PacBio",protocol_comparison))]
plotdata[, gene := factor(gene, levels = c(TRUE,FALSE), labels = c("Gene","Transcript"))]
plotdata[, spike_in := factor(spike_in, levels = c(TRUE,FALSE), labels = c("Spike-in","Cell lines"))]
p_correlation_lr_sr_spike_in <- ggplot(plotdata[spike_in == "Spike-in"], aes(x = gene, y = spr, fill = variable))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 3)+
    xlab("")+
    ylab("Spr")+
    ggtitle("Spike-in")+
    theme_classic()+
    theme( 
    legend.position = "top")
p_correlation_lr_sr_cellline <- ggplot(plotdata[spike_in == "Cell lines"], aes(x = gene, y = spr, fill = variable))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 3)+
    xlab("")+
    ylab("Spr")+
    ggtitle("Cell lines")+
    theme_classic()+
    theme( 
        legend.position = "top")
p_correlation_lr_sr_arranged <- ggarrange(p_correlation_lr_sr_spike_in + theme(legend.position = "none"), p_correlation_lr_sr_cellline + theme(legend.position = "none"),
          nrow = 1, align = "hv")

p_correlation_lr_sr_arranged_legend <- ggarrange(as_ggplot(get_legend(p_correlation_lr_sr_spike_in)), as_ggplot(get_legend(p_correlation_lr_sr_cellline)),
          nrow = 1, align = "hv")


median(corMetricsAll[(protocol_comparison == "cDNAIllumina")&gene&grepl("ERCC|Sequin|SIRV",variable)]$spr)
median(corMetricsAll[(protocol_comparison == "directcDNAIllumina")&gene&grepl("ERCC|Sequin|SIRV",variable)]$spr)

median(corMetricsAll[(protocol_comparison == "cDNAIllumina")&gene&(variable %in% cellLines)]$spr)
median(corMetricsAll[(protocol_comparison == "directcDNAIllumina")&gene&(variable %in% cellLines)]$spr)


median(corMetricsAll[(protocol_comparison == "cDNAIllumina")&(!gene)&grepl("ERCC|Sequin|SIRV",variable)]$spr)
median(corMetricsAll[(protocol_comparison == "directcDNAIllumina")&(!gene)&grepl("ERCC|Sequin|SIRV",variable)]$spr)

median(corMetricsAll[(protocol_comparison == "cDNAIllumina")&(!gene)&(variable %in% cellLines)]$spr)
median(corMetricsAll[(protocol_comparison == "directcDNAIllumina")&(!gene)&(variable %in% cellLines)]$spr)



median(na.omit(corMetricsAll[grepl("IlluminaPacBio", protocol_comparison)&gene]$spr))
median(na.omit(corMetricsAll[grepl("PacBio", protocol_comparison)&(!grepl("IlluminaPacBio", protocol_comparison))&gene]$spr))
```

```{r}
pdf("r3.16.pdf", width = 7, height = 3)
ggdraw() +
     draw_plot(p_transcript_diversity_by_geneLength+ theme(legend.position="none"),0,1/5, 1/3, 4/5) +
     draw_plot(p_correlation_lr_sr_arranged+ theme(legend.position="none"),1/3,1/5, 2/3, 4/5) +
    draw_plot(as_ggplot(get_legend(p_transcript_diversity_by_geneLength)),0,0,1/3,1/5)+
    draw_plot(p_correlation_lr_sr_arranged_legend,1/3,0,2/3,1/5)
dev.off()
```

# Supplementary Text Fig. 6-8
gene level heatmap/pca
```{r 11-gene-expression-2}
source("utility_function.R")
runnamevec <- unique(comDataTranscript[method %in%  c("bambu_lr","salmon_sr")&(ntotal>=400000)]$runname)
methodNamesList <- CJ(lr = c("bambu_lr","salmon_lr"),
                      sr = c("rsem_sr","salmon_sr"),
                      gene = c(TRUE, FALSE),
                      data_type = c("all","protocol_batch_removal","cellline_batch_removal"))
comDataGene <- comDataGene[method %in% c("bambu_lr","salmon_lr","rsem_sr","salmon_sr")&(!grepl("PacBio",protocol_general))]

comDataTranscript <- comDataTranscript[method %in% c("bambu_lr","salmon_lr","rsem_sr","salmon_sr")&(!grepl("PacBio",protocol_general))]
noprint <- lapply(seq_len(nrow(methodNamesList)), function(k){
    lr <- methodNamesList[k]$lr
    sr <- methodNamesList[k]$sr
    gene <- methodNamesList[k]$gene
    data_type <- methodNamesList[k]$data_type
    number_of_genes <- 10000
    if(k%%3==1){
        pdf(paste0(wkdir,"geneHeatmaps/",lr,"_",sr,"_", c("transcript","gene")[gene+1],"_",number_of_genes,"_25May.pdf"), width = 8, height = 5)
    }
    if(gene){
        
        plot_heatmap(dt = comDataGene, methodNames = c(lr, sr), gene, data_type,genevec,ensemblAnnotations.transcripts , samples, number_of_genes, runnamevec)
    }else{
        plot_heatmap(dt = comDataTranscript, methodNames = c(lr, sr), gene, data_type,genevec,ensemblAnnotations.transcripts , samples, number_of_genes, runnamevec)
    }
    if(k%%3==0){
        dev.off()
    }
})
```




