---
title: "R Notebook"
output: html_notebook
---

```{r 0-setup, include=FALSE}
knitr::opts_chunk$set(
  comment = '', fig.width = 6, fig.height = 6,
  results = "hide",
  message = FALSE
)
```

# library set-up
```{r 0-library-loads}
require(GenomicFeatures)
require(GenomicAlignments) ##readGAlignments
require(AnnotationDbi)#loadDb
require(data.table)#fast large dataset manipulation
require(readxl)
require(dplyr)


require(ggplot2)
require(RColorBrewer)
require(gridExtra)

library(gplots)
library(RColorBrewer)
library(limma)
library(ggpubr)

library(ComplexHeatmap)
library(circlize)
library(viridis)
gctorture(FALSE)
library(tximport) 

library(cowplot)
library(ggplotify)# 
source("utility_function_revised.R")
```

# load general data
```{r}
cat('Setting working directory')
wkdir <- 'wkdir'
general_list <- readRDS("general_list2023-04-27.rds")
samples_wSpikein <- general_list$samples_wSpikein
cellLines <- general_list$cellLines
protocolCol <-  general_list$protocolCol
protocolVec <-  general_list$protocolVec
protocolLabel <- general_list$protocolLabel

txvec <- fread(paste0("txList_matchingToGTF_wtChrIs.txt"), header = FALSE)
txvec <- gsub("\\..*","",txvec$V1)
ensemblAnnotations.transcripts <- copy(general_list$ensemblAnnotations.transcripts)
setnames(ensemblAnnotations.transcripts, "ensembl_gene_id","gene_name")
ensemblAnnotations.transcripts <- data.table(tx_name = txvec, status = TRUE)[ensemblAnnotations.transcripts, on = "tx_name"]
ensemblAnnotations.transcripts[is.na(status), status := FALSE]
ensemblAnnotations.transcripts[, all_in := all(status), by = gene_name]
genevec <- unique(ensemblAnnotations.transcripts[which(all_in)]$gene_name)
samples <- general_list$samples
```


# prepare count data
```{r}
temp <- samples_wSpikein[,.(old_runname, runname)]
setnames(temp, "runname", "new_name")
```


```{r bambu-lr expression data using all methods}
comDataTranscript <- readRDS(paste0(wkdir, "combinedExpressionDataTranscript_19June2023.rds"))
comDataGene <- readRDS(paste0(wkdir, "combinedExpressionDataGene_19June2023.rds"))

comDataGene[, old_runname := runname]
comDataGene <- temp[comDataGene, on = "old_runname"]
comDataGene[!is.na(new_name), runname := new_name]
comDataGene[, `:=`(old_runname = NULL, new_name = NULL)]
comDataGene[runname == "GIS_HCT116_PacBio-SMRTcell_Rep5_Run1", runname := "SGNex_Hct116_PacBio-SMRTcell_replicate7_run1"]
comDataGene[, cellLine := gsub("HCT116","Hct116",cellLine)]

comDataTranscript[, old_runname := runname]
comDataTranscript <- temp[comDataTranscript, on = "old_runname"]
comDataTranscript[!is.na(new_name), runname := new_name]
comDataTranscript[, `:=`(old_runname = NULL, new_name = NULL)]
comDataTranscript[runname == "GIS_HCT116_PacBio-SMRTcell_Rep5_Run1", runname := "SGNex_Hct116_PacBio-SMRTcell_replicate7_run1"]
comDataTranscript[, cellLine := gsub("HCT116","Hct116",cellLine)]

```

## prepare trimmed data
```{r}
comDataTranscriptTrimmed <- readRDS(paste0(wkdir, "combinedExpressionDataTranscript_trimreads_20May2024.rds"))
comDataGeneTrimmed <- readRDS(paste0(wkdir, "combinedExpressionDataGene_trimreads_20May2024.rds"))

comDataGeneTrimmed[, old_runname := runname]
comDataGeneTrimmed <- temp[comDataGeneTrimmed, on = "old_runname"]
comDataGeneTrimmed[!is.na(new_name), runname := new_name]
comDataGeneTrimmed[, `:=`(old_runname = NULL, new_name = NULL)]


comDataTranscriptTrimmed[, old_runname := runname]
comDataTranscriptTrimmed <- temp[comDataTranscriptTrimmed, on = "old_runname"]
comDataTranscriptTrimmed[!is.na(new_name), runname := new_name]
comDataTranscriptTrimmed[, `:=`(old_runname = NULL, new_name = NULL)]
```


```{r}
runnamevec <- unique(comDataTranscript[method %in%  c("bambu_lr","salmon_sr")&(ntotal>=400000)]$runname)
```


# prepare spike-in data
```{r}
com_dataList <- readRDS("combinedExpressionDataList_spikein_25May.rds")
com_data <- com_dataList[[1]]
com_data_gene <- com_dataList[[2]]
```


```{r}
samples_wSpikein <- general_list$samples_wSpikein
samples_wSpikein[grepl("PacBio", runname), RNAcontent := gsub("SIRV-1 E2 ","",gsub("sequin Mix A v1.0 ","",RNAcontent))]
```


```{r spike-in}
spike_in <- readRDS("spike_in.rds")
setnames(spike_in, "spike_in","spike_in_name")
spike_in[, spike_in_type := ifelse(spike_in_name == "sirv_e2","SIRV-1",
                                   ifelse(spike_in_name %in% c( "sequinMixAV1","sequinMixAV2"),spike_in_name, "SIRV-4"))]
spike_in[, spike_in_perc := ifelse(spike_in_type != "SIRV-4",1, 
                            ifelse(spike_in_name == "ercc", 51.8/102.2,
                                   ifelse(spike_in_name == "long_sirv",9.0/102.2,41.4/102.2)))]

# spike_in info table 
tt <- unique(spike_in[,.(spike_in_name, conc, tx_name, gene_name)])
tableOutput <- unique(tt[, list(nV = length(unique(na.omit(conc))),
          medV = round(median(unique(na.omit(conc))),2),
          minV = round(min(na.omit(conc)),3),
          maxV = round(max(na.omit(conc)),2)), by = list(spike_in_name)])
tableOutput[, reportV := paste0(medV, " (",minV, ",",maxV,")")]
#   spike_in_name    nV   medV   minV     maxV              reportV
#           <char> <int>  <num>  <num>    <num>               <char>
# 1:       sirv_e2     4 625.00 31.250  4000.00     625 (31.25,4000)
# 2:       sirv_e0     1  60.00 60.000    60.00           60 (60,60)
# 3:     long_sirv     1  60.00 60.000    60.00           60 (60,60)
# 4:          ercc    22   1.10  0.001  1500.00     1.1 (0.001,1500)
# 5:  sequinMixAV1    81  13.43  0.004 30000.00  13.43 (0.004,30000)
# 6:  sequinMixAV2   117   4.23  0.004  1691.89 4.23 (0.004,1691.89)



# find out the total read depth for each spike-in and protocol type, then normalize accordingly 
## optiRum not able to be installed from cran, installed from github instead
#devtools::install_github("jumpingrivers/optiRum")
spike_in_samples <- optiRum::CJ.dt(data.table(spike_in_type = unique(spike_in$spike_in_type), spike_in_attr = c("SIRV-1","SIRV-4","v1.0","V2")),
                       data.table(protocol = unique(com_data$protocol)))
spikein_samples_list <- lapply(seq_len(nrow(spike_in_samples)), function(t){
    return(c(samples_wSpikein[grepl(spike_in_samples[t]$spike_in_attr,RNAcontent)&(gsub("-SMRTcell","",protocol_type) == spike_in_samples[t]$protocol)]$runname,samples_wSpikein[grepl(spike_in_samples[t]$spike_in_attr,RNAcontent)&(gsub("-SMRTcell","",protocol_type) == spike_in_samples[t]$protocol)]$old_runname))
})

ntotalData <- unique(comDataTranscript[, .(runname,ntotal, protocol, method)])

# sanity check to check what is the degree of difference of read count between different methods 
ntotalCount <- unlist(lapply(seq_len(nrow(spike_in_samples)), function(t){
    method_type <- "bambu_lr"
    if(spike_in_samples[t]$protocol == "Illumina") method_type <- "salmon_sr"
    return(sum(ntotalData[runname %in% spikein_samples_list[[t]]&(method == method_type)]$ntotal))
}))
spike_in_samples[, ntotal := ntotalCount]
spike_in <- spike_in_samples[spike_in, on = "spike_in_type", allow.cartesian = TRUE]
spike_in[, sum_conc := sum(conc,na.rm = TRUE), by = list(spike_in_type, protocol, spike_in_name)]
spike_in <- spike_in[, norm_conc := conc/sum_conc*spike_in_perc*0.01*ntotal, by = list(spike_in_type, protocol)]

spike_in_gene <- unique(spike_in[,list(conc = sum(conc,na.rm = TRUE),
                                       norm_conc = sum(norm_conc,na.rm = TRUE)), by = list(spike_in_name, spike_in_type,gene_name, protocol)])
```



# prepare spikein transcript metrics 
```{r spike-in transcript count}
si_data <- com_data[grepl("^(R|ERCC|SIRV)",gene_name)]
si_data[, runname := gsub("v1.0_PacBio","V2_PacBio",runname)]
si_data <- si_data[!grepl("V-1_PacBio",runname)]
si_data[, spike_in_type := gsub("v1.0","V1",unlist(strsplit(runname, "_"))[2]), by = runname]
p_type <- c("directcDNA","cDNA","directRNA")
si_data[, runname := gsub("_sorted","", runname)]
ntotal_bambu <- unique(si_data[method %in% c("bambu_lr","salmon_sr")][,.(ntotal,runname)]) # to get the read count for each sample 
cj_si <- spike_in[unique(si_data[,.(method, runname, spike_in_type,protocol)]), on = c("spike_in_type","protocol"),allow.cartesian = TRUE]
si_data <- si_data[cj_si, on = c("spike_in_type","tx_name","gene_name","protocol","runname","method")]
si_data[is.na(estimates), estimates := 0]
si_data[is.na(normEst), normEst := 0]
si_data <- ntotal_bambu[si_data, on = "runname"]
si_data[, cpm_norm_conc := norm_conc/ntotal*10^6]
si_data <- si_data[!is.na(spike_in_name)]
si_data[method == "bambu_lr"&(is.na(i.ntotal)), normEst:=estimates/ntotal*10^6]
si_data[, spike_in_general_name := ifelse(grepl("SIRV", spike_in_type),"SIRV","Sequin")]
si_data[, spike_in_version := ifelse(grepl("1", spike_in_type),"1","2")]
si_data[, spike_in_name := factor(spike_in_name, levels = c("sequinMixAV1","sequinMixAV2",
    "sirv_e2","ercc","sirv_e0","long_sirv"), labels = c("Sequin (v1)","Sequin (v2)", "SIRV (E2)",
    "ERCC","SIRV (E0)", "Long SIRV"))]
si_data[, spike_in_general_name_revised := ifelse(grepl("SIRV", spike_in_name),"SIRV",
                                ifelse(grepl("Sequin", spike_in_name),"Sequin",
                                ifelse(grepl("ERCC", spike_in_name),"ERCC",NA)))]
si_data[, `:=`(ae = abs(log2(cpm_norm_conc+1)-log2(normEst+1)),
               ard = abs(log2(cpm_norm_conc+1)-log2(normEst+1))/(log2(cpm_norm_conc+1)+log2(normEst+1))*2)]
si_data[is.na(cpm_norm_conc), cpm_norm_conc := 0]

si_data <- si_data[!is.na(conc)]
si_data_transcript <- copy(si_data)
```

```{r spike-in transcript metrics}
rsq <- function(x, y) summary(lm(y~x))$r.squared #https://stackoverflow.com/questions/40901445/function-to-calculate-r2-r-squared-in-r
Allmetrics <- unique(si_data[, list(corM = cor(log2(cpm_norm_conc+1), log2(normEst+1),  method = "spearman"),
                                    corP = cor(log2(cpm_norm_conc+1), log2(normEst+1)),
                                    mae = mean(abs(log2(cpm_norm_conc+1)-log2(normEst+1))),
                                    rmse = sqrt(mean((log2(cpm_norm_conc+1)-log2(normEst+1))^2)),
                                    mard = mean(abs(log2(cpm_norm_conc+1)-log2(normEst+1))/(log2(cpm_norm_conc+1)+log2(normEst+1))*2, na.rm = TRUE),
                                    mard_mod =  mean((log2(cpm_norm_conc+1)-log2(normEst+1))/(log2(cpm_norm_conc+1)+log2(normEst+1))*2, na.rm = TRUE),
                                    r2 = rsq(log2(cpm_norm_conc+1),log2(normEst+1))), by = list(method, protocol, spike_in_type, spike_in_name)])
library(hrbrthemes)
library(ggpubr)

Allmetrics_long <- melt(Allmetrics, id.vars = c("method","protocol","spike_in_type","spike_in_name"), measure.vars = c("corM","corP","mae","rmse","mard","mard_mod","r2"))


tx_metrics <- copy(Allmetrics_long)
```

# Fig. 4 and Fig. 5b-d

## a spike-in scatter
```{r}
plotdata_tx <- si_data_transcript[method %in% c("salmon_lr","salmon_sr")&!(protocol %in% c("directRNA","PacBio"))]
plotdata_tx[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_tx_spikein_scatter <- ggplot(plotdata_tx, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = spike_in_version), size = 0.5, se = FALSE )+
                geom_point(aes(col = spike_in_version),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = spike_in_version,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(spike_in_general_name_revised~protocol)+
                theme_classic()
```

## b tx replicate correlation
```{r}
mat_cor <- readRDS(paste0("replicate_transcript_expression_comparison_salmon_lr_salmon_sr_25May2023.rds"))
mat_cor[, pc_factor := factor(protocol_comparison, unique(mat_cor$protocol_comparison)[c(2,4,1,6,3,5)])]
mat_cor[, ct_factor := factor(common_type, unique(mat_cor$common_type))]
mat_cor[, gc_factor := factor(agg_gene_cluster, unique(mat_cor$agg_gene_cluster)[c(4,3,5,1,2,6,7)])]

plotdata <- mat_cor[(gc_factor == "protein_coding")&(ct_factor %in% c("majorBoth",
"majorLongReadOnly","majorShortReadOnly","others"))&(!is.na(pc_factor))&(!grepl("_HepG2",cellLine))] 
plotdata[, short_read := factor(!grepl("Illumina",pc_factor), labels = c("short read vs long read","long read vs long read"))]
 my_comparisons <- list(c("majorBoth","majorLongReadOnly"),
                       c("majorLongReadOnly","majorShortReadOnly"),
                       c("majorShortReadOnly","others"))

p_replicate_tx_final <- ggplot(plotdata, aes(x = pc_factor, y = r, fill = ct_factor))+
    geom_hline(yintercept = c(0.6,0.8), col = "grey", linetype = "dashed")+
    geom_boxplot(outlier.size = 1, outlier.color = "grey", outlier.shape = 16)+
    scale_fill_brewer(type = "qual", palette = 1)+
    xlab("Protocol comparison")+
    facet_wrap(~ct_factor, nrow = 1)+
    theme_classic()+
    theme(legend.position = "top")
```


## c cell line transcript scatter plot
```{r}
dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")
 cellLines <- c("A549","K562","HepG2","Hct116","MCF7","H9","HEYA8")
k <- cellLines[1]
ddData <- unique(dominant_typeData[cellLine == k,
                .(tx_name, gene_name,majorBoth, majorEither, majorEitherOnly, majorSecBoth, majorLongRead, majorShortRead)])
temp_data <- comDataTranscript[cellLine == k&(method %in% c("salmon_lr","salmon_sr"))&grepl("^ENSG",gene_name)&(gene_name %in% genevec)&(runname %in% runnamevec)]
temp_data <- unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[temp_data, on = "gene_name"]
ave_data <- unique(temp_data[gene_biotype == "protein_coding", list(meanNormEst = mean(normEst)), by = list(gene_name, tx_name, protocol_general)])
ave_data <-  ddData[ave_data, on = c("tx_name","gene_name")]
  
ave_data[, log2NormEst := log2(meanNormEst+1)]

plotdata1 <- dcast(ave_data, tx_name + gene_name + majorBoth + majorEitherOnly + majorLongRead + majorShortRead ~ protocol_general, value.var = "log2NormEst")
  p_major_both <- ggplot(plotdata1[which(majorBoth)], aes(x = directcDNA, y = Illumina))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
        ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+
      xlim(c(0,15))+
      ylim(c(0,15))+
        ggtitle("majorBoth")+theme_classic()
  p_majorLongReadOnly <- ggplot(plotdata1[which(majorEitherOnly&majorLongRead)], aes(x = directcDNA, y = Illumina))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
            breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
        xlim(c(0,15))+
      ylim(c(0,15))+
        ggtitle("majorLongReadOnly")+theme_classic()
  p_majorShortReadOnly <- ggplot(plotdata1[which(majorEitherOnly&majorShortRead)], aes(x = directcDNA, y = Illumina))+
         geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
            breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
       xlim(c(0,15))+
      ylim(c(0,15))+
        ggtitle("majorShortReadOnly")+theme_classic()
 p_minor <- ggplot(plotdata1[which(!(majorBoth|majorEitherOnly))], aes(x = directcDNA, y = Illumina))+
         geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
     xlim(c(0,15))+
      ylim(c(0,15))+
        ggtitle("others")+theme_classic()
```

## d dominant isoform annotations comparison
```{r dominant-data-type-comparison}
dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")
ddData <- unique(dominant_typeData[,
                .(cellLine, tx_name, gene_name,majorBoth, majorEither, majorEitherOnly, majorSecBoth, majorLongRead, majorShortRead)])
tt <- dcast(ddData[which(majorEitherOnly)], gene_name + cellLine ~ majorLongRead, value.var = "tx_name")
setnames(tt, 3:4, c("sr","lr"))

library(bambu)
seOutput <- readRDS("bambuOutput_May25.rds")
annotations <- rowRanges(seOutput)
ct <- compareTranscripts(annotations[tt$sr],annotations[tt$lr])
generate_metrics <- function(ct){
    cols <- 4:17
    metrics <- do.call("rbind",lapply(cols, function(cc){
        m <- table(ct[,cc])
        return(data.table(variable = names(m),
                   value = as.numeric(m),
                   type = colnames(ct)[cc]))
    }))
    return(metrics)
}

# background distribution   sr as reference
set.seed(1234)
background_metrics <- do.call("rbind",lapply(unique(tt$cellLine), function(cl){
    tt_temp <- tt[cellLine == cl]
    background_metrics <- do.call("rbind", lapply(1:3, function(t){
           
    if(t == 1){
        dd_rest <- ddData[!(tx_name %in% tt_temp$sr)&(cellLine == cl)]
    }else if(t==2){
        dd_rest <- ddData[!(tx_name %in% tt_temp$lr)&(cellLine == cl)]
    }else{
        dd_rest <- ddData[which(!majorBoth)&(cellLine == cl)]
    }
   dd_rest[, dummy := 1]
   dd_rest[, ir := cumsum(dummy), by = gene_name]
   xxx_ <- splitAsList(dd_rest$ir, dd_rest$gene_name)
   simIds <- replicate(20, sapply(xxx_, sample, 1))
   simIds <- data.table(simIds, keep.rownames = TRUE)
   background_metrics <- do.call("rbind",lapply(seq_len(ncol(simIds)-1), function(x){
    simDt <- simIds[,c("rn", colnames(simIds)[x+1]),with = FALSE]
    setnames(simDt, 1:2, c("gene_name","ir"))
    dd_selected <- dd_rest[,.(tx_name, gene_name, ir)][simDt, on = c("gene_name","ir")]
    if(t==1){
        ct <- compareTranscripts(annotations[tt_temp$sr],annotations[dd_selected[match(tt_temp$gene_name, gene_name)]$tx_name])
    }else if(t==2){
         ct <- compareTranscripts(annotations[dd_selected[match(tt_temp$gene_name, gene_name)]$tx_name],annotations[tt_temp$lr])
    }else{
    geneNameVec <- intersect( unique(ddData[which(!majorBoth&(cellLine == cl))]$gene_name), unique(ddData[which(majorBoth&(cellLine == cl))]$gene_name))
    ct <- compareTranscripts(annotations[ddData[which(majorBoth&(cellLine == cl))][match(geneNameVec, gene_name)]$tx_name],annotations[dd_selected[match(geneNameVec, gene_name)]$tx_name])
    }
    
    metrics <- generate_metrics(ct)
    metrics[, sim := x]
    metrics[, ref := c("sr","lr","both")[t]]
    metrics[, cellLine := cl]
    return(metrics)
}))
   return(background_metrics)
}))
    return(background_metrics)
}))


obs_metrics <- do.call("rbind",lapply(unique(tt$cellLine), function(cl){
    print(cl)
    ct_temp <- ct[which(tt$cellLine == cl),]
    obs_metrics <- generate_metrics(ct_temp)
    obs_metrics[, cellLine := cl]
    return(obs_metrics)
}))




binary.vars <- c("alternativeFirstExon",
                 "internalFirstExon.query",
                 "internalFirstExon.subject",
                 "alternativeLastExon",
                 "internalLastExon.query",
                 "internalLastExon.subject")

cat.vars <- c("intronRetention.query",
              "intronRetention.subject",
                 "exonSkipping.query",
                 "exonSkipping.subject",
                 "exon5Prime",
                 "exon3Prime")
con.vars <- c("alternativeTSS",
                 "alternativeTES")


binary.sim <- do.call("rbind",lapply(binary.vars, function(v){
    temp <- unique(background_metrics[type == v, list(prop = value/sum(value),
                                                      variable = variable), by = list(sim, type, ref, cellLine)])[variable == TRUE]
    return(unique(temp[,list(me = mean(prop),
                      sd = sd(prop)), by = list(type, ref, cellLine)]))
}))



cat.sim <- do.call("rbind",lapply(cat.vars, function(v){
    
    
        temp <- background_metrics[type == v]
         temp[, bin_variable := ifelse(as.numeric(variable)>=3, ">=3",variable)]
         temp[, bin_value := sum(value), by = list(sim,ref,bin_variable, cellLine)]
         temp <- unique(temp[,.(bin_variable, bin_value, sim,ref, cellLine)])
         temp[,prop := bin_value/sum(bin_value), by = list(sim, ref, cellLine)]
         metrics <- unique(temp[,list(me = mean(prop),
                      sd = sd(prop)), by = list(ref, bin_variable, cellLine)])
         metrics[, type:=v]
    return(metrics)
}))


con.sim <- do.call("rbind", lapply(con.vars, function(v){
    temp <- background_metrics[type == v&(as.numeric(value)!=0)]
    temp <- unique(temp[,list(me_value = mean(rep(as.numeric(variable), value)),
                              sd_value = sd(rep(as.numeric(variable), value)),
                              med_value = median(rep(as.numeric(variable), value))), 
        by = list(sim, type, ref, cellLine)])
    return(unique(temp[,list(me = mean(me_value),
                      sd = sd(me_value)), by = list(type, ref, cellLine)]))
}))



obs_metrics[type %in% binary.vars, prop := value/sum(value), by = list(type, cellLine)]
binary.obs <- unique(obs_metrics[type %in% binary.vars&(variable == TRUE),c("prop","type","cellLine"),with=FALSE])

obs_metrics[type %in% cat.vars, prop := value/sum(value), by = list(type, cellLine)]
cat.obs <- unique(obs_metrics[type %in% cat.vars,c("prop","type","variable", "cellLine"),with=FALSE])
cat.obs[, bin_variable := ifelse(as.numeric(variable)>=3, ">=3",variable)]
cat.obs[, bin_prop := sum(prop), by = list(type,bin_variable, cellLine)]
cat.obs <- unique(cat.obs[,.(bin_prop, bin_variable,type, cellLine)])


con.obs <- unique(obs_metrics[type %in% con.vars,c("variable","type","value","cellLine"),with=FALSE])
con.obs <- con.obs[, list(exact_value = rep(as.integer(variable), value)), by = list(type, cellLine)]

binary.obs[, subject := grepl("subject", type)]
binary.obs[, query := grepl("query", type)]
binary.obs[, query_factor := factor(query, rev(c("TRUE","FALSE")), rev(c("sr","lr")))]
```


binary data process
```{r}
plotdata1 <- binary.sim[ref == "sr"][binary.obs, on = c("type","cellLine")]
plotdata2 <- binary.sim[ref == "lr"][binary.obs, on = c("type","cellLine")]
plotdata <- do.call("rbind", list(plotdata1,plotdata2))
plotdata[, ref_factor := factor(ref, levels = c("lr","sr","both"), labels = c("random sr isoform keep lr isoform",
                                                                       "random lr isoform keep sr isoform","random subject keep both as query"))]

binary.obs[, `:=`(type.adj = gsub("\\.(query|subject)","", type))]

plotdata[, `:=`(type.adj = gsub("\\.(query|subject)","", type))]


plotdata <- plotdata[(query+subject)<1 |(ref == "sr" & query)|(ref == "lr" & subject)|(ref == "both" &query)]
binary.plotdata <- plotdata[!((query&(ref == "lr"))|(subject&(ref == "sr")))]
binary.plotdata[, `:=`(type.adj = gsub("\\.(query|subject)","", type))]
binary.obs[, `:=`(type.adj = gsub("\\.(query|subject)","", type))]
binary.obs_alt <- binary.obs[grepl("alternative",type.adj)]
binary.obs_alt[, query_factor := "sr"]
binary.obs <- rbindlist(list(binary.obs, binary.obs_alt))
binary.plotdata[grepl("alt",type.adj), query_factor := ref]
```

process cat data
```{r}
cat.obs[, subject := grepl("subject", type)]
cat.obs[, query := grepl("query", type)]
cat.obs[, query_factor := factor(query, rev(c("TRUE","FALSE")), rev(c("sr","lr")))]
plotdata1 <- cat.sim[ref == "sr"][cat.obs, on = c("type","bin_variable","cellLine")]
plotdata2 <- cat.sim[ref == "lr"][cat.obs, on = c("type","bin_variable","cellLine")]
plotdata <- do.call("rbind", list(plotdata1,plotdata2))
plotdata[, ref_factor := factor(ref, levels = c("lr","sr"), labels = c("random sr isoform keep lr isoform",
                                                                       "random lr isoform keep sr isoform"))]
plotdata[, bin_variable_factor := factor(bin_variable, levels = rev(c("0","1","2",">=3")))]
cat.obs[, bin_variable_factor := factor(bin_variable, levels = rev(c("0","1","2",">=3")))]
cat.plotdata <- plotdata[!((query&(ref == "lr"))|(subject&(ref == "sr")))]

cat.plotdata[, `:=`(type.adj = gsub("\\.(query|subject)","", type))]
cat.obs[, `:=`(type.adj = gsub("\\.(query|subject)","", type))]
setnames(cat.plotdata,"bin_prop","prop")
setnames(cat.obs,"bin_prop","prop")
cat.obs_prime <- cat.obs[grepl("Prime",type.adj)]
cat.obs_prime[, query_factor := "sr"]
cat.obs <- rbindlist(list(cat.obs, cat.obs_prime))
cat.plotdata[grepl("Prime",type.adj), query_factor := ref]
```




combine all data
```{r}
obsData <- do.call("rbind", list(binary.obs[,.(prop, cellLine, query_factor, type.adj)],
                       cat.obs[bin_variable_factor == 1,.(prop, cellLine, query_factor, type.adj) ]))
simData <- do.call("rbind", list(binary.plotdata[,.(type.adj, prop,cellLine,me, sd, query_factor)],
                                 cat.plotdata[bin_variable_factor == 1,.(type.adj, prop,cellLine,me, sd, query_factor)]))

saveRDS(list(obsData, simData), file = paste0(wkdir, "obs_sim_data_list.rds"))
```


```{r}
obsSimData <- readRDS(paste0(wkdir, "obs_sim_data_list.rds"))
cl <- "A549"
obsData <- obsSimData[[1]][cellLine == cl&(!grepl("alt",type.adj))]
simData <- obsSimData[[2]][cellLine == cl&(!grepl("alt",type.adj))]
simData[, z_stat := abs(prop-me)/sd]
simData[, z_pvalue := pnorm(z_stat, lower.tail = FALSE)*2]
simData[, z_pvalue_adjusted := pmin(1,z_pvalue*7)]


p_combined <- ggplot(obsData, aes(x = type.adj, y = prop))+
    geom_col(aes(fill = query_factor), alpha = 0.5)+
    geom_point(data = simData, aes(y = me, col= query_factor))+
    geom_errorbar(data = simData, aes(ymin = me-2*sd, ymax = me+2*sd, group = query_factor, col= query_factor), width = 0.2)+
     geom_text(data = simData, aes( y = me+3*sd, label = round(z_pvalue_adjusted,4)))+
    facet_wrap(~query_factor, ncol = 2)+
    xlab("sr (query) vs lr (subject)")+
     theme_classic() +
    theme(legend.position = "top")


# sr: internalFirstExon
# pvalData <- simData[type.adj == "internalFirstExon"][]
# pvalData[, z_stat := (prop-me)/sd]
# pvalData[, z_pvalue := pnorm(z_stat, lower.tail = FALSE)*2]
# pvalData[, z_pvalue_adjusted := pmin(1,z_pvalue*14)]
# pt((0.13588953-0.22669956)/0.002660127,18)
# 
# > pvalData
#              type.adj       prop cellLine         me          sd query_factor
#  1: internalFirstExon 0.23339405     K562 0.14388585 0.002884383           sr
#  2: internalFirstExon 0.21003614       H9 0.13175563 0.002164416           sr
#  3: internalFirstExon 0.22669956     A549 0.13588953 0.002660127           sr
#  4: internalFirstExon 0.20753853    HEYA8 0.12429422 0.002684779           sr
#  5: internalFirstExon 0.21870224    HepG2 0.13240239 0.001637166           sr
#  6: internalFirstExon 0.25927372     MCF7 0.15354679 0.002329515           sr
#  7: internalFirstExon 0.25173725   Hct116 0.15033434 0.002779508           sr
#  8: internalFirstExon 0.06727383     K562 0.05994536 0.001786984           lr
#  9: internalFirstExon 0.06171810       H9 0.05531693 0.001172121           lr
# 10: internalFirstExon 0.05030154     A549 0.04608690 0.001522928           lr
# 11: internalFirstExon 0.04929040    HEYA8 0.04728369 0.001691549           lr
# 12: internalFirstExon 0.05655134    HepG2 0.05147978 0.001539015           lr
# 13: internalFirstExon 0.05492646     MCF7 0.05404790 0.001373487           lr
# 14: internalFirstExon 0.05664088   Hct116 0.05171103 0.001896865           lr
#         z_stat      z_pvalue z_pvalue_adjusted
#  1: 31.0320074 1.995444e-211     2.793621e-210
#  2: 36.1670423 2.008040e-286     2.811256e-285
#  3: 34.1374807 2.051071e-255     2.871499e-254
#  4: 31.0060159 4.472434e-211     6.261408e-210
#  5: 52.7129593  0.000000e+00      0.000000e+00
#  6: 45.3858156  0.000000e+00      0.000000e+00
#  7: 36.4823283 2.114361e-291     2.960106e-290
#  8:  4.1010296  4.113160e-05      5.758423e-04
#  9:  5.4611838  4.729699e-08      6.621579e-07
# 10:  2.7674581  5.649530e-03      7.909342e-02
# 11:  1.1863176  2.354969e-01      1.000000e+00
# 12:  3.2953271  9.830718e-04      1.376301e-02
# 13:  0.6396588  5.223944e-01      1.000000e+00
# 14:  2.5989491  9.350964e-03      1.309135e-01
```


## e number of junctions

```{r}
dt <- readRDS("junc_dt_26June2023.rds")
dt[, njunc_cat_new := cut(njunc, c(-0.01,0,1,6,10,50,120,187))]
plotData <- unique(dt[, list(count=sum(nread)), by = list(njunc_cat_new,runname,protocol_type,cellLine)])
plotData[,total_count:= sum(count), by = list(protocol_type,runname)]
plotData[, frac_count:=count/total_count]
plotData[, protocol_type_factor:=factor(protocol_type, levels = protocolVec )]

p_njunc <- ggplot(plotData[!grepl("allSpikin",runname)], 
                  aes(x = njunc_cat_new,
                           y = count/total_count,
                           fill = protocol_type_factor,
                           col = protocol_type_factor))+
     geom_boxplot(aes(fill = protocol_type_factor,col = protocol_type_factor), width = 0.5, position=position_dodge(0.75), outlier.shape = NA)+
      scale_fill_manual(values = protocolCol,
                      name = "Protocol",
                      limits = protocolVec,
                      labels = protocolLabel)+
    scale_color_manual(values = protocolCol,
                       name = "Protocol",
                       limits = protocolVec,
                       labels = protocolLabel)+
    stat_summary(
        fun = median,
        geom = 'line',
        aes(group = protocol_type_factor, colour = protocol_type_factor),
        position = position_dodge(width = 0.9) #this has to be added
    )+
    labs(title = 'All cell lines',
         x = 'Number of junctions',
         y = 'Percentage over all reads')+
    theme_classic()

p_njunc
```




## f number of transcripts compatible per read 
```{r}
ntx_dt_sum <- readRDS("ntx_dt_sum_26June2023.rds")
ntx_dt_sum[trimmed == TRUE, protocol_type_factor := "lr_sim"]
p_number_of_transcripts_per_read_trimmed <- ggplot(ntx_dt_sum[!grepl("all",runname)], aes(x = ntx_cat, y = nperc,  fill = protocol_type_factor, col =  protocol_type_factor))+ geom_boxplot( width = 0.5, position=position_dodge(0.75), outlier.shape = NA)+ 
     stat_summary(
        fun = median,
        geom = 'line',
        aes(group = protocol_type_factor, colour = protocol_type_factor),
        position = position_dodge(width = 0.75) #this has to be added
    )+
    scale_fill_brewer(type = "qual", palette = 3)+
    scale_color_brewer(type = "qual", palette = 3)+
    scale_x_discrete(limits = c(0:10,">10"))+
    ylab("Relative proportion of reads")+
    xlab("Number of transcripts per read aligned to")+
    theme_classic()
```

## g dominant isoform expression support
```{r}
dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")
ddData <- unique(dominant_typeData[,
                .(cellLine, tx_name, gene_name,majorBoth, majorEither, majorEitherOnly, majorSecBoth, majorLongRead, majorShortRead)])
ddData[, major_status := ifelse(majorEitherOnly&majorLongRead, "majorLongReadOnly",ifelse(majorEitherOnly&majorShortRead, "majorShortReadOnly", ifelse(majorBoth, "majorBoth","others")))]
ddData <- ddData[,.(tx_name, gene_name, cellLine, major_status), with = TRUE]
countData <- rbind(comDataTranscript[method %in% c("salmon_lr","salmon_sr"),c("tx_name", 
"gene_name", "cellLine", "protocol_general", "runname", "normEst","method"), with = FALSE],
comDataTranscriptTrimmed[method %in% c("trim_lr_sim"),c("tx_name", 
"gene_name", "cellLine", "protocol_general", "runname", "normEst","method"), with = FALSE])

ddData_wCount <- countData[ddData, on = c("tx_name","gene_name","cellLine")]

plotdata <- unique(ddData_wCount[,list(n = sum(normEst>=1),
                                       ntotal = .N), by = list(cellLine, protocol_general,major_status ,runname, method)])
plotdata[grepl("sim", method), protocol_general := "lr_sim"]


p_major_expression <- ggplot(plotdata, aes(x = major_status, y = n/ntotal, fill = protocol_general))+
    geom_boxplot()+
    xlab("")+
    ylab("Percentage of transcripts expressed with \n >= 1 CPM in all samples")+
    scale_fill_brewer(type = "qual", palette = 3)+
    theme_classic()
```


## h trimmed vs original long read correlation

```{r}
methodVec_used <- c("salmon_lr","lr_300bp_1ts","lr_150bp_1ts","lr_sim","salmon_sr")
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
rep_dt <- unique(plotdata[,.(runname, method, cellLineRep)])
rep_dt[,nc := length(unique(method)), by = cellLineRep]
matched_names <- unique(rep_dt[nc>4&(method == "salmon_lr")]$runname)
cor_data_across2 <- do.call("rbind",lapply(matched_names, function(x){
    
    ynames <- unique(plotdata[method %in% "salmon_sr"&(cellLineRep == unique(plotdata[runname == x]$cellLineRep))]$runname)
    cor_data_across <- do.call("rbind",lapply(ynames, function(k){
        x_data <- plotdata[(gene_biotype == "protein_coding")&(runname %in% c(x, k))&(method %in% methodVec_used)]
         print(paste(x,k))
    plotdata_wide <- dcast(x_data, gene_name + tx_name + gene_biotype ~ method, value.var = "normEst")
    plotdata_wide <- unique(dominant_typeData[cellLine == unique(x_data$cellLine),.(tx_name, gene_name, geneExpressedInBoth, majorBoth, majorEither, majorLongRead, majorShortRead, majorEitherOnly)])[plotdata_wide, on = c("tx_name","gene_name")]
  
    colIds <- setdiff(match(rev(methodVec_used), colnames(plotdata_wide)),NA)
    cor_data_across <- data.table(spCor = c(cor(log2(plotdata_wide[majorBoth&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,length(colIds)],
                     cor(log2(plotdata_wide[majorLongRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,length(colIds)],
cor(log2(plotdata_wide[majorShortRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,length(colIds)],
cor(log2(plotdata_wide[!((majorBoth|majorEitherOnly)&geneExpressedInBoth),colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,length(colIds)],
cor(log2(plotdata_wide[,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,length(colIds)]),
           methodNames = rep(colnames(plotdata_wide)[colIds],5),
           data_type = rep(c("majorBoth","majorLongread","majorShortRead","others","all"), each = length(colIds)))
    cor_data_across[, lr_name := x]
    cor_data_across[, sr_name := k]
    return(cor_data_across)
    
    }))
    return(cor_data_across)
}))

saveRDS(cor_data_across2, file = paste0(wkdir, "cor_data_across2.rds"))
```

```{r}
cor_data_across2 <- readRDS(paste0(wkdir, "cor_data_across2.rds"))
plotdata_final <- na.omit(cor_data_across2)
plotdata_final[,nc := .N, by = list(lr_name, sr_name, data_type)]
p_sim_lr_cor <-ggplot(plotdata_final[methodNames == "lr_sim"&(data_type != "all")], aes(x = data_type, y = spCor))+
    geom_hline(yintercept = c(0.6,0.8, 0.9), linetype = "dashed", col = "grey")+
     geom_line(aes(group = paste(lr_name,sr_name)), col = "grey", linewidth = 0.4)+
    geom_boxplot(col = "black",width = 0.5, outlier.shape = NA)+
    geom_point(alpha = 0.2, shape = 16, col = "grey")+
    ylab("Spearman correlation \n (trimmed vs original)")+
    xlab("Transcript type")+
    theme_classic()
```


## i trimmed vs original long read mrd
```{r}
methodVec_used <- c("sr_150bp","salmon_lr","lr_300bp_1ts","lr_150bp_1ts","lr_sim","salmon_sr")
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
rep_dt <- unique(plotdata[method %in% methodVec_used,.(runname, method, cellLineRep)])
rep_dt[,nc := length(unique(method)), by = cellLineRep]
length(unique(rep_dt[nc==6&(method == "salmon_lr")]$runname))
[1] 67

matched_names <- unique(rep_dt[nc>4&(method == "salmon_lr")]$runname)
mrd_data_across <- do.call("rbind",lapply(matched_names, function(x){
    
    ynames <- unique(plotdata[method %in% "salmon_sr"&(cellLineRep == unique(plotdata[runname == x]$cellLineRep))]$runname)
    cor_data_across <- do.call("rbind",lapply(ynames, function(k){
        x_data <- plotdata[(gene_biotype == "protein_coding")&(runname %in% c(x, k))&(method %in% methodVec_used)]
         print(paste(x,k))
    plotdata_wide <- dcast(x_data, gene_name + tx_name + gene_biotype ~ method, value.var = "normEst")
    plotdata_wide <- unique(dominant_typeData[cellLine == unique(x_data$cellLine),.(tx_name, gene_name, geneExpressedInBoth, majorBoth, majorEither, majorLongRead, majorShortRead, majorEitherOnly)])[plotdata_wide, on = c("tx_name","gene_name")]
  
    colIds <- setdiff(match(rev(methodVec_used), colnames(plotdata_wide)),NA)
   diff_pairs <- combn(c(5,2,1,6),2)[,c(1,2,3)] # salmon lr vs lr sim, salmon sr, sr_150bp
   
cor_data_across <- data.table(mrd = c(pair_mrd(log2(plotdata_wide[majorBoth&geneExpressedInBoth,colIds, with = FALSE]+1), diff_pairs),
pair_mrd(log2(plotdata_wide[majorLongRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), diff_pairs),
pair_mrd(log2(plotdata_wide[majorShortRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), diff_pairs),
pair_mrd(log2(plotdata_wide[!((majorBoth|majorEitherOnly)&geneExpressedInBoth),colIds, with = FALSE]+1), diff_pairs),
pair_mrd(log2(plotdata_wide[,colIds, with = FALSE]+1), diff_pairs)),
           methodNames = rep(colnames(plotdata_wide)[colIds][diff_pairs[2,]],5),
           data_type = rep(c("majorBoth","majorLongread","majorShortRead","others","all"), each = ncol(diff_pairs)))
    cor_data_across[, lr_name := x]
    cor_data_across[, sr_name := k]
    return(cor_data_across)
    
    }))
    return(cor_data_across)
}))
saveRDS(mrd_data_across, file = paste0(wkdir, "mrd_data_across.rds"))
```

```{r}
mrd_data_across <- readRDS(paste0(wkdir, "mrd_data_across.rds"))
plotdata_final <- na.omit(mrd_data_across)
plotdata_final[,nc := .N, by = list(lr_name, sr_name, data_type)]
p_sim_lr_mrd <-ggplot(plotdata_final[methodNames == "lr_sim"&(data_type != "all")], aes(x = data_type, y = mrd))+
    geom_hline(yintercept = c(0), linetype = "dashed", col = "grey")+
      geom_line(aes(group = paste(lr_name,sr_name)), col = "grey", linewidth = 0.4)+
     geom_boxplot(col = "black",width = 0.5, outlier.shape = NA)+
  
    geom_point(alpha = 0.2, shape = 16, col = "grey")+
     ylab("Mean relative absolute \n difference  (trimmed vs original)")+
    xlab("Transcript type")+
    theme_classic()
```


## j trimmed and long read vs short read correlation 
```{r}
methodVec_used <- c("salmon_lr","lr_300bp_1ts","lr_150bp_1ts","lr_sim","salmon_sr")
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
rep_dt <- unique(plotdata[,.(runname, method, cellLineRep)])
rep_dt[,nc := length(unique(method)), by = cellLineRep]
matched_names <- unique(rep_dt[nc>4&(method == "salmon_lr")]$runname)
cor_data_across <- do.call("rbind",lapply(matched_names, function(x){
    
    ynames <- unique(plotdata[method %in% "salmon_sr"&(cellLineRep == unique(plotdata[runname == x]$cellLineRep))]$runname)
    cor_data_across <- do.call("rbind",lapply(ynames, function(k){
        x_data <- plotdata[(gene_biotype == "protein_coding")&(runname %in% c(x, k))&(method %in% methodVec_used)]
         print(paste(x,k))
    plotdata_wide <- dcast(x_data, gene_name + tx_name + gene_biotype ~ method, value.var = "normEst")
    plotdata_wide <- unique(dominant_typeData[cellLine == unique(x_data$cellLine),.(tx_name, gene_name, geneExpressedInBoth, majorBoth, majorEither, majorLongRead, majorShortRead, majorEitherOnly)])[plotdata_wide, on = c("tx_name","gene_name")]
  
    colIds <- setdiff(match(rev(methodVec_used), colnames(plotdata_wide)),NA)
    cor_data_across <- data.table(spCor = c(cor(log2(plotdata_wide[majorBoth&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
                     cor(log2(plotdata_wide[majorLongRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[majorShortRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[!((majorBoth|majorEitherOnly)&geneExpressedInBoth),colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1]),
           methodNames = rep(colnames(plotdata_wide)[colIds],5),
           data_type = rep(c("majorBoth","majorLongread","majorShortRead","others","all"), each = length(colIds)))
    cor_data_across[, lr_name := x]
    cor_data_across[, sr_name := k]
    return(cor_data_across)
    
    }))
    return(cor_data_across)
}))

saveRDS(cor_data_across, file = paste0(wkdir, "cor_data_across.rds"))
```


#### the correlation between kallisto_sr_public or rsem_sr_public 
```{r}
methodVec_used <- c("lr_sim","salmon_lr","kallisto_sr_public")
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
## create name match CJ
rname_pair <- do.call("rbind",lapply(cellLines[1:5], function(k){
    dt <- CJ(sr_runname = unique(plotdata[method == "kallisto_sr_public"&(i.cellLine == k)]$runname),
       lr_runname = unique(plotdata[method == "lr_sim"&(cellLine == k)]$runname))
    dt[, cellLine := k]
    return(dt)
}))
cor_data_across <- do.call("rbind",lapply(seq_len(nrow(rname_pair)), function(x){
        temp_rname <- rname_pair[x]
        x_data <- plotdata[(gene_biotype == "protein_coding")&(runname %in% c(temp_rname$sr_runname, temp_rname$lr_runname))&(method %in% methodVec_used)]
    plotdata_wide <- dcast(x_data, gene_name + tx_name + gene_biotype ~ method, value.var = "normEst")
    plotdata_wide <- unique(dominant_typeData[cellLine == unique(temp_rname$cellLine),.(tx_name, gene_name, geneExpressedInBoth, majorBoth, majorEither, majorLongRead, majorShortRead, majorEitherOnly)])[plotdata_wide, on = c("tx_name","gene_name")]
  
    colIds <- setdiff(match(rev(methodVec_used), colnames(plotdata_wide)),NA)
    cor_data_across <- data.table(spCor = c(cor(log2(plotdata_wide[majorBoth&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
                     cor(log2(plotdata_wide[majorLongRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[majorShortRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[!((majorBoth|majorEitherOnly)&geneExpressedInBoth),colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1]),
           methodNames = rep(colnames(plotdata_wide)[colIds],5),
           data_type = rep(c("majorBoth","majorLongread","majorShortRead","others","all"), each = length(colIds)))
    cor_data_across[, lr_name := temp_rname$lr_runname]
    cor_data_across[, sr_name := temp_rname$sr_runname]
    return(cor_data_across)
    
    }))
saveRDS(cor_data_across, file = paste0(wkdir, "cor_data_across_encode_kallisto.rds"))
```

```{r}
methodVec_used <- c("lr_sim","salmon_lr","rsem_sr_public")
cor_data_across <- do.call("rbind",lapply(seq_len(nrow(rname_pair)), function(x){
        temp_rname <- rname_pair[x]
        x_data <- plotdata[(gene_biotype == "protein_coding")&(runname %in% c(temp_rname$sr_runname, temp_rname$lr_runname))&(method %in% methodVec_used)]
    plotdata_wide <- dcast(x_data, gene_name + tx_name + gene_biotype ~ method, value.var = "normEst")
    plotdata_wide <- unique(dominant_typeData[cellLine == unique(temp_rname$cellLine),.(tx_name, gene_name, geneExpressedInBoth, majorBoth, majorEither, majorLongRead, majorShortRead, majorEitherOnly)])[plotdata_wide, on = c("tx_name","gene_name")]
  
    colIds <- setdiff(match(rev(methodVec_used), colnames(plotdata_wide)),NA)
    cor_data_across <- data.table(spCor = c(cor(log2(plotdata_wide[majorBoth&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
                     cor(log2(plotdata_wide[majorLongRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[majorShortRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[!((majorBoth|majorEitherOnly)&geneExpressedInBoth),colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1]),
           methodNames = rep(colnames(plotdata_wide)[colIds],5),
           data_type = rep(c("majorBoth","majorLongread","majorShortRead","others","all"), each = length(colIds)))
    cor_data_across[, lr_name := temp_rname$lr_runname]
    cor_data_across[, sr_name := temp_rname$sr_runname]
    return(cor_data_across)
    
    }))
saveRDS(cor_data_across, file = paste0(wkdir, "cor_data_across_encode_rsem.rds"))
```

```{r}
cor_data_across <- readRDS(paste0(wkdir, "cor_data_across.rds"))
plotdata_final <- na.omit(cor_data_across)
plotdata_final[,nc := .N, by = list(lr_name, sr_name, data_type)]
methodVec_used <- c("salmon_lr","lr_sim")
labeldata <- unique(plotdata_final[, list(med_value = median(spCor)),
                                   by = list(methodNames,data_type)])
labeldata <- labeldata[methodNames %in% c("lr_sim","salmon_lr")&(data_type != "all")]
labeldata <- dcast(labeldata, data_type ~ methodNames, value.var = "med_value")
labeldata[, label := paste0(round(salmon_lr,2), " vs ", round(lr_sim, 2))]
labeldata[, `:=`(methodNames = "salmon_lr",
                 spCor = 0.95)]
p_sim_lr_vs_sr_cor <-ggplot(plotdata_final[methodNames %in% c("lr_sim","salmon_lr")&(data_type != "all")], aes(x = factor(methodNames,methodVec_used), y = spCor))+
     geom_hline(yintercept = c(0.6,0.8, 0.9), linetype = "dashed", col = "grey")+
     geom_line(aes(group = paste(lr_name,sr_name)), col = "grey", size = 0.4)+
    geom_boxplot(col = "black",width = 0.5, outlier.shape = NA)+
   
   
    geom_point(aes(col = factor(methodNames,methodVec_used)), alpha = 0.5)+
    stat_compare_means(paired = TRUE, method = "t.test", comparisons = list(c("lr_sim","salmon_lr")),
                        label.y = c(1, 0.9, 0.8),
                        symnum.args <- list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, Inf), symbols = c("****", "***", "**", "*", "ns")))+
   geom_text(data = labeldata, aes(label = label))+
    scale_color_brewer(type = "qual", palette = 2)+
    scale_fill_brewer(type = "qual", palette = 2)+
    facet_wrap(~data_type, ncol = 5)+
     ylab("Spearman correlation")+
    xlab("Long reads: trimmed to mimic short read")+
    theme_classic()
```

```{r}
labeldata <- unique(plotdata_final[, list(med_value = median(spCor)),
                                   by = list(methodNames,data_type)])
labeldata <- labeldata[methodNames %in% c("lr_sim","salmon_lr")&(data_type == "all")]
labeldata <- dcast(labeldata, data_type ~ methodNames, value.var = "med_value")
labeldata[, label := paste0(round(salmon_lr,2), " vs ", round(lr_sim, 2))]
labeldata[, `:=`(methodNames = "salmon_lr",
                 spCor = 0.95)]
p_sim_lr_vs_sr_cor_all <-ggplot(plotdata_final[methodNames %in% c("lr_sim","salmon_lr")&(data_type == "all")], aes(x = factor(methodNames,methodVec_used), y = spCor))+
     geom_hline(yintercept = c(0.6,0.8, 0.9), linetype = "dashed", col = "grey")+
     geom_line(aes(group = paste(lr_name,sr_name)), col = "grey", size = 0.4)+
    geom_boxplot(col = "black",width = 0.5, outlier.shape = NA)+
   
   
    geom_point(aes(col = factor(methodNames,methodVec_used)), alpha = 0.5)+
    stat_compare_means(paired = TRUE, method = "t.test", comparisons = list(c("lr_sim","salmon_lr")),
                        label.y = c(1, 0.9, 0.8),
                        symnum.args <- list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, Inf), symbols = c("****", "***", "**", "*", "ns")))+
   geom_text(data = labeldata, aes(label = label))+
    scale_color_brewer(type = "qual", palette = 2)+
    scale_fill_brewer(type = "qual", palette = 2)+
    facet_wrap(~data_type, ncol = 5)+
     ylab("Spearman correlation")+
    xlab("Long reads: trimmed to mimic short read")+
    theme_classic()
```


```{r}
cor_data_across <- readRDS(paste0(wkdir, "cor_data_across_encode_kallisto.rds"))
plotdata_final <- na.omit(cor_data_across)
plotdata_final[,nc := .N, by = list(lr_name, sr_name, data_type)]
methodVec_used <- c("salmon_lr","lr_sim")
labeldata <- unique(plotdata_final[, list(med_value = median(spCor)),
                                   by = list(methodNames,data_type)])
labeldata <- labeldata[methodNames %in% c("lr_sim","salmon_lr")&(data_type != "all")]
labeldata <- dcast(labeldata, data_type ~ methodNames, value.var = "med_value")
labeldata[, label := paste0(round(salmon_lr,2), " vs ", round(lr_sim, 2))]
labeldata[, `:=`(methodNames = "salmon_lr",
                 spCor = 0.95)]
p_sim_lr_vs_sr_cor_kallisto <-ggplot(plotdata_final[methodNames %in% c("lr_sim","salmon_lr")&(data_type != "all")], aes(x = factor(methodNames,methodVec_used), y = spCor))+
     geom_hline(yintercept = c(0.6,0.8, 0.9), linetype = "dashed", col = "grey")+
     geom_line(aes(group = paste(lr_name,sr_name)), col = "grey", size = 0.4)+
    geom_boxplot(col = "black",width = 0.5, outlier.shape = NA)+
   
   
    geom_point(aes(col = factor(methodNames,methodVec_used)), alpha = 0.5)+
    stat_compare_means(paired = TRUE, method = "t.test", comparisons = list(c("lr_sim","salmon_lr")))+
   geom_text(data = labeldata, aes(label = label))+
    scale_color_brewer(type = "qual", palette = 2)+
    scale_fill_brewer(type = "qual", palette = 2)+
    facet_wrap(~data_type, ncol = 5)+
     ylab("Spearman correlation")+
    xlab("Long reads: trimmed to mimic short read")+
    theme_classic()
```

```{r}
labeldata <- unique(plotdata_final[, list(med_value = median(spCor)),
                                   by = list(methodNames,data_type)])
labeldata <- labeldata[methodNames %in% c("lr_sim","salmon_lr")&(data_type == "all")]
labeldata <- dcast(labeldata, data_type ~ methodNames, value.var = "med_value")
labeldata[, label := paste0(round(salmon_lr,2), " vs ", round(lr_sim, 2))]
labeldata[, `:=`(methodNames = "salmon_lr",
                 spCor = 0.95)]
p_sim_lr_vs_sr_cor_all_kallisto <-ggplot(plotdata_final[methodNames %in% c("lr_sim","salmon_lr")&(data_type == "all")], aes(x = factor(methodNames,methodVec_used), y = spCor))+
     geom_hline(yintercept = c(0.6,0.8, 0.9), linetype = "dashed", col = "grey")+
     geom_line(aes(group = paste(lr_name,sr_name)), col = "grey", size = 0.4)+
    geom_boxplot(col = "black",width = 0.5, outlier.shape = NA)+
   
   
    geom_point(aes(col = factor(methodNames,methodVec_used)), alpha = 0.5)+
    stat_compare_means(paired = TRUE, method = "t.test", comparisons = list(c("lr_sim","salmon_lr")),
                        label.y = c(1, 0.9, 0.8,0.7),
                        symnum.args <- list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, Inf), symbols = c("****", "***", "**", "*", "ns")))+
   geom_text(data = labeldata, aes(label = label))+
    scale_color_brewer(type = "qual", palette = 2)+
    scale_fill_brewer(type = "qual", palette = 2)+
    facet_wrap(~data_type, ncol = 5)+
     ylab("Spearman correlation")+
    xlab("Long reads: trimmed to mimic short read")+
    theme_classic()
```



```{r}
cor_data_across <- readRDS(paste0(wkdir, "cor_data_across_encode_rsem.rds"))
plotdata_final <- na.omit(cor_data_across)
plotdata_final[,nc := .N, by = list(lr_name, sr_name, data_type)]
methodVec_used <- c("salmon_lr","lr_sim")
labeldata <- unique(plotdata_final[, list(med_value = median(spCor)),
                                   by = list(methodNames,data_type)])
labeldata <- labeldata[methodNames %in% c("lr_sim","salmon_lr")&(data_type != "all")]
labeldata <- dcast(labeldata, data_type ~ methodNames, value.var = "med_value")
labeldata[, label := paste0(round(salmon_lr,2), " vs ", round(lr_sim, 2))]
labeldata[, `:=`(methodNames = "salmon_lr",
                 spCor = 0.95)]
p_sim_lr_vs_sr_cor_rsem <-ggplot(plotdata_final[methodNames %in% c("lr_sim","salmon_lr")&(data_type != "all")], aes(x = factor(methodNames,methodVec_used), y = spCor))+
     geom_hline(yintercept = c(0.6,0.8, 0.9), linetype = "dashed", col = "grey")+
     geom_line(aes(group = paste(lr_name,sr_name)), col = "grey", size = 0.4)+
    geom_boxplot(col = "black",width = 0.5, outlier.shape = NA)+
   
   
    geom_point(aes(col = factor(methodNames,methodVec_used)), alpha = 0.5)+
    stat_compare_means(paired = TRUE, method = "t.test", comparisons = list(c("lr_sim","salmon_lr")))+
   geom_text(data = labeldata, aes(label = label))+
    scale_color_brewer(type = "qual", palette = 2)+
    scale_fill_brewer(type = "qual", palette = 2)+
    facet_wrap(~data_type, ncol = 5)+
     ylab("Spearman correlation")+
    xlab("Long reads: trimmed to mimic short read")+
    theme_classic()
```

```{r}
labeldata <- unique(plotdata_final[, list(med_value = median(spCor)),
                                   by = list(methodNames,data_type)])
labeldata <- labeldata[methodNames %in% c("lr_sim","salmon_lr")&(data_type == "all")]
labeldata <- dcast(labeldata, data_type ~ methodNames, value.var = "med_value")
labeldata[, label := paste0(round(salmon_lr,2), " vs ", round(lr_sim, 2))]
labeldata[, `:=`(methodNames = "salmon_lr",
                 spCor = 0.95)]
p_sim_lr_vs_sr_cor_all_rsem <-ggplot(plotdata_final[methodNames %in% c("lr_sim","salmon_lr")&(data_type == "all")], aes(x = factor(methodNames,methodVec_used), y = spCor))+
     geom_hline(yintercept = c(0.6,0.8, 0.9), linetype = "dashed", col = "grey")+
     geom_line(aes(group = paste(lr_name,sr_name)), col = "grey", size = 0.4)+
    geom_boxplot(col = "black",width = 0.5, outlier.shape = NA)+
   
   
    geom_point(aes(col = factor(methodNames,methodVec_used)), alpha = 0.5)+
    stat_compare_means(paired = TRUE, method = "t.test", comparisons = list(c("lr_sim","salmon_lr")))+
   geom_text(data = labeldata, aes(label = label))+
    scale_color_brewer(type = "qual", palette = 2)+
    scale_fill_brewer(type = "qual", palette = 2)+
    facet_wrap(~data_type, ncol = 5)+
     ylab("Spearman correlation")+
    xlab("Long reads: trimmed to mimic short read")+
    theme_classic()
```
```{r}
library(cowplot)
library(ggplotify)
pdf("lr_vs_sr_correlation_encode.pdf", width = 14, height = 12)
ggarrange(p_sim_lr_vs_sr_cor,p_sim_lr_vs_sr_cor_all,
          p_sim_lr_vs_sr_cor_kallisto,p_sim_lr_vs_sr_cor_all_kallisto,
          p_sim_lr_vs_sr_cor_rsem,p_sim_lr_vs_sr_cor_all_rsem,widths = c(4.5,1.5), heights = c(1,1,1), nrow = 3, ncol =2, common.legend = TRUE, align = 'hv')
dev.off()
```




## arrange main figure
```{r}
library(cowplot)
library(ggplotify)
pdf("figure4_draft_addspaceforschematic_4Sep2023.pdf", width = 14, height = 16)
ggdraw() +
 draw_plot(p_tx_spikein_scatter+ theme(legend.position="none"),0,5/9, 3/7, 4/9) +
  draw_plot(p_replicate_tx_final + theme(legend.position="none"),3/7,23/27,4/7 , 4/27) +
   
 draw_plot(p_combined + theme(legend.position="none"),3/7,5/9,4/7 , 4/27) +
    draw_plot(p_major_both + theme(legend.position="none"),3/7,19/27,1/7 , 4/27)+
     draw_plot(p_majorLongReadOnly + theme(legend.position="none"),4/7,19/27,1/7 , 4/27)+
    draw_plot(p_majorShortReadOnly + theme(legend.position="none"),5/7,19/27,1/7 ,4/27)+
    draw_plot(p_minor + theme(legend.position="none"),6/7,19/27,1/7 , 4/27)+
      draw_plot(p_major_expression + theme(legend.position="none"),0,3/9,1/3 , 2/9)+
    draw_plot(p_njunc + theme(legend.position="none"),1/3,3/9,1/3 , 2/9)+
    draw_plot(p_number_of_transcripts_per_read_trimmed + theme(legend.position="none"),2/3,3/9,1/3 , 2/9)+
   
    draw_plot(p_sim_lr_cor + ggtitle("any")+ theme(legend.position="none"),1.5/7,1/9,1.5/7 , 2/9)+
     draw_plot(p_sim_lr_mrd + ggtitle("any")+ theme(legend.position="none"),3/7,1/9,1.5/7 , 2/9)+
    draw_plot(p_sim_lr_vs_sr_cor + theme(legend.position="none"),4.5/7,1/9,2.5/7 , 2/9)+
      draw_plot(as_ggplot(get_legend(p_replicate_tx_final)),0,0,1/7,1/9)+
    draw_plot(as_ggplot(get_legend(p_njunc)),1/7,0,1/7,1/9)+
    draw_plot(as_ggplot(get_legend(p_number_of_transcripts_per_read_trimmed)),2/7,0,1/7,1/9)+
     draw_plot(as_ggplot(get_legend(p_major_expression)),3/7,0,1/7,1/9)+
    draw_plot(as_ggplot(get_legend(p_sim_lr_vs_sr_cor)),4/7,0,1/7,1/9)+
      draw_plot(as_ggplot(get_legend(p_majorLongReadOnly)),5/7,0,1/7,1/9)
dev.off()
```


# Supplementary Fig. 5
## a spike-in tx scatter-remaining ones
```{r}
plotdata_tx_suppl <- si_data_transcript[!(method %in% c("salmon_lr","salmon_sr"))|((method %in% c("salmon_lr","salmon_sr")) & (protocol %in% c("PacBio","direct RNA")))]
plotdata_tx_suppl[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_tx_spikein_scatter_suppl <- ggplot(plotdata_tx_suppl, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = method), size = 0.5, se = FALSE )+
                geom_point(aes(col = method),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = method,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(paste0(spike_in_general_name_revised,spike_in_version)~protocol)+
               theme_classic()
```
## b spike-in mean absolute errors
```{r}
p_transcript_spikein_abe_suppl <- ggplot(si_data_transcript, aes(x=ae, fill = method)) +
    geom_histogram(aes(y=..density..), alpha = 0.5, position = "identity",binwidth = 0.25) +  
      scale_fill_brewer(type = "qual", palette = 3)+
                xlab('Absolute error')+
                facet_grid(paste0(spike_in_general_name_revised,spike_in_version)~protocol)+
               theme_classic()

pdf("tx_abe_suppl.pdf", width = 12, height = 12)
p_transcript_spikein_abe_suppl
dev.off()
```


## c spike-in tx metrics between expected and observed
```{r}
pList_tx_no_legend <- lapply(unique(tx_metrics$variable), function(x){
    plotdata_final_metrics <- tx_metrics[(variable == x)]
    plotdata_final_metrics[, na_status := all(is.na(value))|all(value == 0), by = list(spike_in_name)]
   plotdata_final_metrics <- plotdata_final_metrics[which(!na_status)]
   plotdata_final_metrics[, spike_in_name_factor := factor(spike_in_name, levels = c("ERCC",
            "Sequin (v1)","Sequin (v2)","SIRV (E2)","SIRV (E0)","Long SIRV"))]
    if(x %in%c("mae","rmse","mard")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "blue", high = "white") + # red means bad
   geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+ggtitle(x)+
    xlab("")+ylab("")+
   theme_minimal()+
           theme( 
        axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }else if(x %in% c("corM","r2")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "white", high = "blue") + # red means bad
    geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
   xlab("")+
             ylab("")+
    theme_minimal()+
           theme( 
        axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }else{
         px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
    scale_fill_gradient2(low = "white",
                       mid = "blue",
                       high = "white") +
     geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
             xlab("")+
             ylab("")+
   theme_minimal()+
           theme( 
       axis.text.y=element_blank(),  #remove y axis labels
        axis.ticks.y=element_blank() , #remove y axis ticks
                  legend.position = "none")
    }
    return(px)
})


pList_tx_only_legend <- lapply(unique(tx_metrics$variable), function(x){
    plotdata_final_metrics <- tx_metrics[(variable == x)]
   plotdata_final_metrics[, na_status := all(is.na(value))|all(value == 0), by = list(spike_in_name)]
   plotdata_final_metrics <- plotdata_final_metrics[which(!na_status)]
   plotdata_final_metrics[, spike_in_name_factor := factor(spike_in_name, levels = c("ERCC",
            "Sequin (v1)","Sequin (v2)","SIRV (E2)","SIRV (E0)","Long SIRV"))]
    if(x %in%c("mae","rmse","mard")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "blue", high = "white") + # red means bad
    geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
             ggtitle(x)+
    theme_minimal()
    }else if(x %in% c("corM","r2")){
        px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
   geom_tile(color = "black") +
  scale_fill_gradient(low = "white", high = "blue") + # red means bad
    geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
    theme_minimal()
    }else{
         px <- ggplot( plotdata_final_metrics, aes(spike_in_name_factor,paste0(protocol,method), 
                                                                 fill= value)) + 
    geom_tile(color = "black") +
    scale_fill_gradient2(low = "white",
                       mid = "blue",
                       high = "white") +
     geom_text(aes(label = round(value,2)), color = "black", size = 4) +
  coord_fixed()+
            ggtitle(x)+
     theme_minimal()
    }
    return(as_ggplot(get_legend(px)))
})


protocol_annotation_tx <- ggplot(tx_metrics[(variable == "corM")&(spike_in_name == "Sequin (v2)")], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= protocol)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())

method_annotation_tx <- ggplot(tx_metrics[(variable == "corM")&(spike_in_name == "Sequin (v2)")], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= method)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())

metric_annotation_tx <- ggplot(tx_metrics[(variable == "corM")&(protocol == "cDNA")&(method == "bambu_lr")&(!is.na(value))], aes(spike_in_name,paste0(protocol,method), 
                                                                 fill= spike_in_name)) + 
   geom_tile(color = "white") +
  coord_equal()+
  theme(axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank())
p_tx_spikein_metrics_summarised <- ggarrange( plotlist = pList_tx_no_legend,nrow = 1, ncol =6, align = "hv")
p_tx_spikein_metrics_summarised_legend <- ggarrange( plotlist = pList_tx_only_legend,nrow = 1, ncol =6, align = "hv")
```




## d correlation between long read and short read for spike-in genes and transcripts 
```{r}
transcript_mat <- dcast(si_data_transcript[method %in% c("salmon_lr","salmon_sr")], spike_in_type + spike_in_name + gene_name + tx_name ~ protocol, value.var = "normEst")
snames <- unique(transcript_mat$spike_in_name)
txMetric <- do.call("rbind",lapply(seq_along(snames), function(k){
    sname <- snames[k]
    geneCorMetric <- cor(log2(transcript_mat[spike_in_name == sname,-c(1:4), with=FALSE]+1), method = "spearman")
    geneCorMetric_melt <- melt(data.table(geneCorMetric,keep.rownames = TRUE), id.vars = "rn",
                           measure.vars = colnames(geneCorMetric))
    geneCorMetric_melt[, protocol_comparison := paste(sort(c(rn, as.character(variable))), collapse = ""), by = list(rn, variable)]
    geneCorMetric_melt[, duplicated_status := duplicated(protocol_comparison)]
    geneCorMetric_melt <- geneCorMetric_melt[rn != variable &(!duplicated_status)]
    geneCorMetric_melt[, spike_in_name := sname]
    return(geneCorMetric_melt)
}))
#cor(log2(transcript_mat[,-c(1:4), with=FALSE]+1), method = "spearman")
# > cor(log2(transcript_mat[,-c(1:4), with=FALSE]+1), method = "spearman")
#             Illumina PacBio      cDNA directRNA directcDNA
# Illumina   1.0000000     NA 0.8965047 0.7963132  0.8645917
# PacBio            NA      1        NA        NA         NA
# cDNA       0.8965047     NA 1.0000000 0.8111494  0.9586989
# directRNA  0.7963132     NA 0.8111494 1.0000000  0.8352759
# directcDNA 0.8645917     NA 0.9586989 0.8352759  1.0000000




mat_cor_tx <- readRDS(paste0("replicate_transcript_expression_comparison_salmon_lr_salmon_sr_25May2023.rds"))
mat_cor_tx[, protocol_comparison := gsub(" vs ","", protocol_comparison)]

txMetric[, variable := NULL]

setnames(txMetric, c("spike_in_name","value"),c("variable","spr"))
txMetric[, spike_in := TRUE]
mat_cor_tx[, spike_in := FALSE]

setnames(mat_cor_tx, "cellLine","variable")
txMetric[, gene := FALSE]
mat_cor_tx[, gene := FALSE]
merge.names <- c("spr","protocol_comparison","variable","spike_in","gene")
corMetricsAll <- do.call("rbind", list(txMetric[,merge.names, with = FALSE],
                      mat_cor_tx[agg_gene_cluster == "all"&(common_type == "all")&(!grepl("_HepG2",variable)),merge.names, with = FALSE])
                      )


plotdata <- corMetricsAll[grepl("Illumina",protocol_comparison)&(!grepl("PacBio",protocol_comparison))]
plotdata[, spike_in := factor(spike_in, levels = c(TRUE,FALSE), labels = c("Spike-in","Cell lines"))]
p_correlation_lr_sr_spike_in <- ggplot(plotdata[spike_in == "Spike-in"], aes(x = variable, y = spr, fill = variable))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 2)+
    xlab("")+
    ylab("Spearman correlation")+
    ggtitle("Spike-in")+
    theme_classic()+
    theme(legend.position = "top")
p_correlation_lr_sr_cellline <- ggplot(plotdata[spike_in == "Cell lines"], aes(x = variable, y = spr, fill = variable))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 2)+
    xlab("")+
    ylab("Spearman correlation")+
    ggtitle("Cell lines")+
    theme_classic()+
    theme(legend.position = "top")
p_tx_spikein_correlation_lr_sr_arranged <- ggarrange(p_correlation_lr_sr_spike_in, p_correlation_lr_sr_cellline,
          nrow = 1, ncol = 2, align = "hv", labels = "auto")
```


## e cell line transcript scatter plot - remaining ones
```{r}
dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")
cellLines <- c("A549","K562","HepG2","Hct116","MCF7","H9","HEYA8")
ddData <- unique(dominant_typeData[,
                .(tx_name, gene_name,cellLine, majorBoth, majorEither, majorEitherOnly, majorSecBoth, majorLongRead, majorShortRead)])
temp_data <- comDataTranscript[method %in% c("salmon_lr","salmon_sr")&grepl("^ENSG",gene_name)&(gene_name %in% genevec)&(runname %in% runnamevec)]
temp_data <- unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[temp_data, on = "gene_name"]
ave_data <- unique(temp_data[gene_biotype == "protein_coding", list(meanNormEst = mean(normEst)), by = list(gene_name, tx_name, protocol_general, cellLine)])
ave_data <-  ddData[ave_data, on = c("tx_name","gene_name","cellLine")]
  
ave_data[, log2NormEst := log2(meanNormEst+1)]

plotdata1 <- dcast(ave_data, tx_name + gene_name + cellLine + majorBoth + majorEitherOnly + majorLongRead + majorShortRead ~ protocol_general, value.var = "log2NormEst")
plotdata1[, major_status := ifelse(majorEitherOnly&majorLongRead, "majorLongReadOnly",ifelse(majorEitherOnly&majorShortRead, "majorShortReadOnly", ifelse(majorBoth, "majorBoth","others")))]
plotdata2 <- melt(plotdata1, id.vars = c("gene_name","tx_name","major_status","cellLine",
                                         "Illumina"), measure.vars = c("PacBio-SMRTcell","cDNA",
                                         "directcDNA","directRNA"))
p_tx_cellline_scatter_suppl_majorBoth <- ggplot(plotdata2[cellLine %in% cellLines&(major_status == "majorBoth")], aes(x = value, y = Illumina))+
        geom_hex(binwidth = c(0.1, 0.1))+ #col = "steelblue", size = 0.5, alpha = 0.1
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
          facet_grid(cellLine~variable)+
    theme_classic()
p_tx_cellline_scatter_suppl_majorLongReadOnly <- ggplot(plotdata2[cellLine %in% cellLines&(major_status == "majorLongReadOnly")], aes(x = value, y = Illumina))+
         geom_hex(binwidth = c(0.1, 0.1))+ #col = "steelblue", size = 0.5, alpha = 0.1
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
          facet_grid(cellLine~variable)+
    theme_classic()
p_tx_cellline_scatter_suppl_majorShortReadOnly <- ggplot(plotdata2[cellLine %in% cellLines&(major_status == "majorShortReadOnly")], aes(x = value, y = Illumina))+
      geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
          facet_grid(cellLine~variable)+
    theme_classic()
p_tx_cellline_scatter_suppl_minor <- ggplot(plotdata2[cellLine %in% cellLines&(major_status == "others")], aes(x = value, y = Illumina))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
          facet_grid(cellLine~variable)+
    theme_classic()
```
## f simulation vs long read scatter plot
```{r}
ensemblAnnotations.transcripts <- general_list$ensemblAnnotations.transcripts

plotdata <- comDataTranscriptTrimmed
plotdata[, method := gsub("_salmon","",gsub("trim_","",method))]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
setnames(ensemblAnnotations.transcripts, "gene_id","gene_name")
plotdata <- unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[plotdata, on = c("gene_name")]

samples <- general_list$samples
plotdata <- samples[plotdata, on = "runname"]
library(GGally)    
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
rep_dt <- unique(plotdata[,.(runname, method, cellLineRep)])
rep_dt[,nc := length(unique(method)), by = cellLineRep]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
repVec <- unique(rep_dt[nc==10]$cellLineRep)

rv <- repVec[1]
temp_data <- unique(plotdata[cellLineRep == rv&(gene_biotype %in% pro_types), 
    list(normEst = mean(normEst)), by = list(gene_name, tx_name, gene_biotype, method)])

temp_data_wide <- dcast(temp_data, gene_name + tx_name + gene_biotype  ~ method, value.var = "normEst")
    temp_data_wide <- dominant_typeData[cellLine == gsub("rep.*","", rv)][temp_data_wide, on = c("tx_name","gene_name")]
colIds <- c(21,24,23,22,26,25,19,17,18,20)

temp_data_wide[, majorType := ifelse(majorBoth, "majorBoth", 
    ifelse(majorEitherOnly&majorLongRead, "majorLongReadOnly",
           ifelse(majorEitherOnly&majorShortRead, "majorShortReadOnly","minor")))]

p_sim_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(lr_sim+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("fragmentation simulated sr vs lr")+theme_classic()
    
p_sr_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(salmon_sr+1)))+
       geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("original sr vs lr")+theme_classic()

p_sr_vs_sim <- ggplot(temp_data_wide, aes(x = log2(lr_sim + 1), y = log2(salmon_sr+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("original sr vs fragmentation simulated")+theme_classic()
```




## g replicate correlation other metrics 
```{r}
methodNamesList <- CJ(lr = c("bambu_lr","salmon_lr"),
                      sr = c("rsem_sr","salmon_sr"),
                      gene = c("gene", "transcript"),
                      metric = c("","mae_","mard_","mard_mod_","rmse_"))
txCellLineMetrics <- methodNamesList[gene == "transcript"]
matCorData <- do.call("rbind",lapply(seq_len(nrow(txCellLineMetrics)), function(k){
    print(k)
    metric <- txCellLineMetrics[k]$metric
    saved_date <- ifelse(metric == "", "25May2023","1Aug2023")
    lr <- txCellLineMetrics[k]$lr
    sr <- txCellLineMetrics[k]$sr
    mat_cor <- readRDS(paste0("replicate_transcript_expression_comparison_",lr,"_",sr,"_",metric,saved_date,".rds"))
mat_cor[, within_cellLine := !grepl("_|all",cellLine)]

mat_cor[, pc_factor := factor(protocol_comparison, rev(unique(mat_cor$protocol_comparison)[c(1,4,2,5,3,6)]))]


mat_cor[, ct_factor := factor(common_type, unique(mat_cor$common_type))]



if(metric == ""){
    mat_cor[, gc_factor := factor(agg_gene_cluster, c("protein_coding","lncRNA","others","all"))]
    setnames(mat_cor, "spr","metricValue")
}else{
    mat_cor[, gc_factor := factor(gene_cluster, c("protein_coding","lncRNA","others","all"))]
    setnames(mat_cor, "mae","metricValue")
}
outData <- mat_cor[(gc_factor == "protein_coding")&(ct_factor %in% c("majorBoth",
"majorLongReadOnly","majorShortReadOnly","others"))&(!is.na(pc_factor))&(within_cellLine)] #&match_status
outData <- outData[,.(metricValue, cellLine, pc_factor, ct_factor, gc_factor)]
outData[, short_read := factor(!grepl("Illumina",pc_factor), labels = c("short read vs long read","long read vs long read"))]
outData[, method_pair := paste0(lr,sr)]
outData[, metric_type := ifelse(metric == "", "cor",metric)]
return(outData)
}))
saveRDS(matCorData, file = "matCorDataTx_7Aug2023.rds")

matCorData <- readRDS("matCorDataTx_7Aug2023.rds")
methodPairVec <- unique(matCorData$method_pair)
metricVec <- unique(matCorData$metric_type)
matCorData[, short_read := grepl("Illumina", pc_factor)]
matCorData[, dcDNA := grepl("directcDNA",pc_factor)]
matCorData[, dRNA := grepl("directRNA",pc_factor)]
matCorData[, cDNA := grepl("vs cDNA",pc_factor)|grepl("^cDNA",pc_factor)]


matCorData[, new_pc_factor := paste0(c("","Illumina")[short_read+1],
c("","cDNA")[cDNA+1],
c("","dcDNA")[dcDNA+1],
c("","dRNA")[dRNA+1])]

matCorData_tx <- copy(matCorData)

methodPairVec <- unique(matCorData$method_pair)
library(ggpubr)
p_metric_list <- lapply(1:4, function(k){
    
    my_comparisons <- list(c("majorBoth","majorLongReadOnly"),
                       c("majorLongReadOnly","majorShortReadOnly"),
                       c("majorShortReadOnly","others"))
labeldata <- unique(matCorData[, list(med_value = median(metricValue)),
                                   by = list(ct_factor,short_read,method_pair, metric_type)])
labeldata[, y := ifelse(metric_type=="cor",0.93, 
                                            ifelse(metric_type == "mard_",1.8,
                                                   ifelse(metric_type == "mard_mod_",1.2,NA)))]
plotdata <- matCorData[(metric_type %in% c("cor","mard_","mard_mod_"))&(method_pair == methodPairVec[k])]
plotdata[, short_read := factor(short_read, labels = c("short read vs long read","long read vs long read"))]
plotdata[, labelY := ifelse(metric_type=="cor",0.93, 
                                            ifelse(metric_type == "mard_",1.8,
                                                   ifelse(metric_type == "mard_mod_",1.2,NA)))]

p_replicate_tx_final_overall <- ggplot(plotdata, aes(x = ct_factor, y = metricValue, fill = ct_factor))+
    geom_boxplot(outlier.size = 1, outlier.color = "grey", outlier.shape = 16)+
    scale_fill_brewer(type = "qual", palette = 1)+
    xlab("Protocol comparison")+
    stat_compare_means(comparisons = my_comparisons,
                       paired = FALSE,
                       aes(label.y = labelY))+
    facet_grid(factor(metric_type, labels = c("Spearman correlation",
           "MARD","MRD"))~short_read, scales = "free_y")+
    ggtitle(methodPairVec[k])+
    theme_classic()+
    theme(legend.position = "top")
    return(p_replicate_tx_final_overall)
})

p_metric_list_mod <- lapply(seq_along(p_metric_list), function(x) p_metric_list[[x]] + theme(legend.position = "none"))
```


## h complexity plot for the salmon-lr vs salmon-sr comparison
```{r}
mat_cor <- readRDS(paste0("replicate_transcript_expression_comparison_salmon_lr_salmon_sr_by_complexity_13June2023.rds"))
mat_cor[, pc_factor := factor(protocol_comparison, unique(mat_cor$protocol_comparison)[c(2,4,1,6,3,5)])]
mat_cor[, ct_factor := factor(common_type, unique(mat_cor$common_type))]
mat_cor[, gc_factor := factor(agg_gene_cluster, unique(mat_cor$agg_gene_cluster)[c(4,3,5,1,2,6,7)])]



p_replicate_tx_complexity <- ggplot(mat_cor[!grepl("_HepG2",cellLine)&(gc_factor == "protein_coding")&(ct_factor != "all")], aes(x = pc_factor, 
                                            y = r,
                                            fill = ct_factor, col = ct_factor))+
    geom_hline(yintercept = c(0.6,0.8), col = "grey", linetype = "dashed")+
    geom_boxplot(alpha = 0.5)+
    scale_fill_brewer(type = "qual", palette = 2)+
    scale_color_brewer(type = "qual", palette = 2)+
    facet_wrap(~ct_factor, nrow = 1)+
    xlab("transcript type by complexity (number of isoforms)")+
    theme_classic()
```



## m check the full-length read support for those transcript dominant in long read vs dominant in short read as fraction of the gene 
```{r check-fl-agree-lrvssr}

ct$cellLine <- tt$cellLine
ct$gene_name <- tt$gene_name
ct <- data.table(ct)

geneVec_agree <- unique(ddData[which(majorBoth), .(gene_name, cellLine)])
geneVec_disagree <- unique(ddData[which(majorEitherOnly), .(gene_name, cellLine)])

geneVec_agree[, status := "agree"]
geneVec_disagree

ddData[, `:=`(statusBoth = any(majorBoth),
              statusEither = any(majorEitherOnly)), by = list(gene_name, cellLine)]
agreeDt <- ddData[which(statusBoth), list(type = majorBoth), by = list(gene_name, tx_name,cellLine)]
disagreeDt <- ddData[which(statusEither), list(type = ifelse(majorLongRead, 1, ifelse(majorShortRead, 2,3))), by = list(gene_name, tx_name, cellLine)]


# full-length
flDt <- data.table(assays(seOutput)$fullLengthCounts, keep.rownames = TRUE)
flDt <- melt(flDt, id.vars = "rn", measure.vars = colnames(flDt)[-1])
setnames(flDt, "rn","tx_name")
flDt[, cellLine := strsplit(as.character(variable),"\\_")[[1]][2], by = variable]
# unique read 
agreeDt <- flDt[agreeDt, on = c("tx_name","cellLine")]
agreeDt[, valueProp := value/sum(value,na.rm = TRUE), by = list(gene_name, variable, cellLine)]
agreeDt[, valueSum := sum(value,na.rm = TRUE), by = list(gene_name, variable, cellLine)]

agreeDtAve <- unique(agreeDt[, list(aveProp = mean(valueProp,na.rm = TRUE),
                             aveSum = mean(valueSum,na.rm = TRUE)), by = list(tx_name, gene_name, type, cellLine)])


disagreeDt <- flDt[disagreeDt, on = c("tx_name","cellLine")]
disagreeDt[, valueProp := value/sum(value,na.rm = TRUE), by = list(gene_name, variable)]
disagreeDt[, valueSum := sum(value,na.rm = TRUE), by = list(gene_name, variable)]

disagreeDtAve <- unique(disagreeDt[, list(aveProp = mean(valueProp,na.rm = TRUE),
                             aveSum = mean(valueSum,na.rm = TRUE)), by = list(tx_name, gene_name, type, cellLine)])




ct1 <- copy(ct[,c("queryId", "internalFirstExon.query", "internalLastExon.query", "gene_name","cellLine"), with = FALSE])
ct1[, type := 2]
setnames(ct1, c("queryId","internalFirstExon.query","internalLastExon.query"),c("tx_name","internalFirstExon","internalLastExon"))


ct2 <- copy(ct[,c("subjectId", "internalFirstExon.subject", "internalLastExon.subject", "gene_name","cellLine"), with = FALSE])
ct2[, type := 1]
setnames(ct2,  c("subjectId","internalFirstExon.subject","internalLastExon.subject"),c("tx_name","internalFirstExon","internalLastExon"))

         
disagreeDtAve <- do.call("rbind",list(ct1,ct2))[disagreeDtAve, on = c("gene_name","tx_name","type","cellLine")]


disagreeDtAve[, internal := ifelse(internalFirstExon&internalLastExon, "internalFirstLastExon",
  ifelse(internalFirstExon, "internalFirstExon",
  ifelse(internalLastExon, "internalLastExon","others")))]
disagreeDtAve[is.na(internal), internal := "others"]
disagreeDtAve[, internal_factor := factor(internal, 
    levels = c("internalFirstLastExon","internalFirstExon",
               "internalLastExon","others"))]
p_disagree <- ggplot(disagreeDtAve[aveSum>5], aes(x = as.factor(type), y = aveProp, fill = internal_factor))+
    geom_boxplot()+
    scale_fill_brewer(type = "qual", palette = 3)+
    scale_x_discrete(breaks =c("1","2","3"),
                 limits =c("1","2","3"),
                 labels = c("Long Read Only","Short Read Only","Others"))+
    xlab("transcript type")+
    facet_wrap(~cellLine)+
    ylab("full-length read support of isoform/gene")+
    theme_classic()
pdf(paste0(wkdir, "lrvssr_disagree_boxplot_fl_support.pdf"), width = 12, height = 10)
p_disagree
dev.off()
```



## h annotation comparison ---remaining ones 
```{r}
obsSimData <- readRDS(paste0(wkdir, "obs_sim_data_list.rds"))
obsData <- obsSimData[[1]][cellLine != "A549"][(!grepl("alt",type.adj))]
simData <- obsSimData[[2]][cellLine != "A549"][(!grepl("alt",type.adj))]

simData[, z_stat := abs(prop-me)/sd]
simData[, z_pvalue := pnorm(z_stat, lower.tail = FALSE)*2]
simData[, z_pvalue_adjusted := pmin(1,z_pvalue*7)]

p_combined_suppl <- ggplot(obsData, aes(x = type.adj, y = prop))+
    geom_col(aes(fill = query_factor), alpha = 0.5)+
    # geom_bar(stat = "identity", fill = "lightblue")+
    geom_point(data = simData, aes(y = me, col= query_factor))+
    geom_errorbar(data = simData, aes(ymin = me-2*sd, ymax = me+2*sd, group = query_factor, col= query_factor), width = 0.2)+
    geom_text(data = simData, aes( y = me+3*sd, label = round(z_pvalue_adjusted,4)))+
    facet_grid(cellLine~query_factor, scales = "free")+
    xlab("sr (query) vs lr (subject)")+
     theme_classic() +
    theme(legend.position = "top")
```

## i short read simulation
```{r}
ensemblAnnotations.transcripts <- general_list$ensemblAnnotations.transcripts

plotdata <- comDataTranscriptTrimmed
plotdata[, method := gsub("_salmon","",gsub("trim_","",method))]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
setnames(ensemblAnnotations.transcripts, "gene_id","gene_name")
plotdata <- unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[plotdata, on = c("gene_name")]

samples <- general_list$samples
plotdata <- samples[plotdata, on = "runname"]
sr_names <- unique(plotdata[protocol_general == "Illumina"]$runname)
cor_data_sr <- do.call("rbind", lapply(sr_names, function(x){
    x_data <- plotdata[(gene_biotype == "protein_coding")&(runname == x)]
    print(x)
    plotdata_wide <- dcast(x_data, gene_name + tx_name + gene_biotype ~ method, value.var = "normEst")
    plotdata_wide <- dominant_typeData[cellLine == unique(x_data$cellLine)][plotdata_wide, on = c("tx_name","gene_name")]
    
    colIds <- grep("sr", colnames(plotdata_wide))

cor_data_sr <- data.table(spCor = c(cor(log2(plotdata_wide[majorBoth&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
                     
                     cor(log2(plotdata_wide[majorLongRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[majorShortRead&majorEitherOnly&geneExpressedInBoth,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[!((majorBoth|majorEitherOnly)&geneExpressedInBoth),colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1],
cor(log2(plotdata_wide[,colIds, with = FALSE]+1), method = "spearman", use = "pairwise.complete.obs")[,1]),
           methodNames = rep(colnames(plotdata_wide)[colIds],5),
           data_type = rep(c("majorBoth","majorLongread","majorShortRead","others","all"), each = length(colIds)))
cor_data_sr[, runname := x]
return(cor_data_sr)
}
))
saveRDS(cor_data_sr, file = paste0(wkdir, "cor_data_sr.rds"))
```

```{r}
cor_data_sr <- readRDS(paste0(wkdir, "cor_data_sr.rds"))
plotdata_final <- na.omit(cor_data_sr)
plotdata_final[,nc := .N, by = list(runname, data_type)]
plotdata_final <- plotdata_final[nc>5]
labeldata <- unique(plotdata_final[(methodNames %in% c("sr_150bp", "sr_125bp","sr_100bp","sr_75bp","sr_50bp"))&(data_type != "all"), list(med_value = median(spCor)),
                                   by = list(methodNames,data_type)])

p_short_read <-ggplot(plotdata_final[(methodNames %in% c("sr_150bp", "sr_125bp","sr_100bp","sr_75bp","sr_50bp"))&(data_type != "all")], aes(x = data_type, y = spCor))+
    geom_boxplot(col = "black",width = 0.5)+
    geom_line(aes(group = runname), col = "grey", size = 0.4)+
    geom_point(aes(col = data_type), alpha = 0.5)+
    stat_compare_means(paired = TRUE, method = "t.test", comparisons = list(c("majorBoth","majorLongread"),c("majorBoth","majorShortRead"), c("majorLongread","majorShortRead")),label.y = c(0.99, 0.89, 0.79))+
    geom_text(data = labeldata, aes(y = 0.99, label = round(med_value,2)))+
    scale_color_brewer(type = "qual", palette = 2)+
    scale_fill_brewer(type = "qual", palette = 2)+
    facet_wrap(~factor(methodNames, c("sr_150bp", "sr_125bp","sr_100bp","sr_75bp","sr_50bp")), ncol = 5)+
    ylab("Spearman correlation")+
     xlab("Short reads: trimmed to single-end")+
    theme_classic()

```


## arrange supplementary figure
```{r}
pdf("suppl_figure5_draft_5Sep2023.pdf", width = 14, height = 20)
ggdraw() +
     draw_plot(p_tx_spikein_scatter_suppl+ theme(legend.position="none"),0,16/21, 1/3, 5/21) +
    draw_plot(p_transcript_spikein_abe_suppl+ theme(legend.position="none"),1/3,16/21, 1/3, 5/21) +
  draw_plot(p_tx_spikein_metrics_summarised + theme(legend.position="none"),2/3,18.5/21,1/3,2.5/21) +
    draw_plot( ggarrange(p_correlation_lr_sr_spike_in + theme(legend.position="none"), 
                          p_correlation_lr_sr_cellline + theme(legend.position="none"),
          nrow = 1, ncol = 2, align = "hv", labels = "auto") ,2/3,16/21, 1/3, 2.5/21) +
    draw_plot(p_tx_cellline_scatter_suppl_majorBoth + theme(legend.position="none"),0,11/21,1/4 , 5/21) +
    draw_plot(p_tx_cellline_scatter_suppl_majorLongReadOnly + theme(legend.position="none"),1/4,11/21,1/4, 5/21) +
    draw_plot(p_tx_cellline_scatter_suppl_majorShortReadOnly + theme(legend.position="none"),1/2,11/21,1/4 , 5/21) +
    draw_plot(p_tx_cellline_scatter_suppl_minor + theme(legend.position="none"),3/4,11/21,1/4, 5/21) +
    draw_plot(ggarrange(plotlist = p_metric_list_mod,nrow = 1, ncol = 4, align = "hv"),0,6/21,1/2 , 5/21) +
    draw_plot(ggarrange(p_sim_vs_lr+theme(legend.position="none"),
p_sr_vs_lr+theme(legend.position="none"),p_sr_vs_sim+theme(legend.position="none"),nrow = 3, ncol = 1, align = "hv"),1/2,6/21,1/2 , 5/21) +
    draw_plot(p_replicate_tx_complexity + theme(legend.position="none"),0,4/21,1/2 , 2/21) +
    draw_plot(p_combined_suppl + theme(legend.position="none"),0,1/21,1/2 , 3/21) +
     draw_plot(p_short_read + theme(legend.position="none"),1/2,1/21,1/2 , 5/21) +
    draw_plot(as_ggplot(get_legend(p_tx_spikein_scatter_suppl)),0,0,1/7,1/21)+
     draw_plot(protocol_annotation,1/7,0,1/7,1/21)+
    draw_plot(method_annotation,2/7,0,1/7,1/21)+
      draw_plot(as_ggplot(get_legend(p_correlation_lr_sr_spike_in)),3/7,0,1/7,1/21)+
    draw_plot(as_ggplot(get_legend(p_correlation_lr_sr_cellline)),4/7,0,1/7,1/21)+
     draw_plot(as_ggplot(get_legend(p_short_read)),5/7,0,1/7,1/21)+
    draw_plot(as_ggplot(get_legend(p_tx_cellline_scatter_suppl_majorBoth)),6/7,0,1/7,1/21)
dev.off()
```

Supplementary Text Fig. 20
```{r}
pdf("FigureR1.pdf", width = 14, height = 8)
ggdraw() +
     draw_plot(p_sim_vs_lr + xlim(0,15)+ylim(0,15)+theme(legend.position="none"),1/4,3/5, 1/2, 2/5) +
    draw_plot(p_sr_vs_lr + xlim(0,15)+ylim(0,15)+ theme(legend.position="none"),1/4,1/5, 1/2, 2/5)+
     draw_plot(p_sim_lr_mrd + theme(legend.position="none"),3/4,1/5, 1/4, 4/5)+
     draw_plot(as_ggplot(get_legend(p_sim_vs_lr)),0,0,1/2,1/5)+
    draw_plot(as_ggplot(get_legend(p_sr_vs_lr)),1/2,0,1/2,1/5)
dev.off()

```


# Supplementary Text Fig. 5
```{r, eval = FALSE}
library(BiocFileCache)
bfc1 <- BiocFileCache("RunBambu22Apr/rc", ask = FALSE)
info <- bfcinfo(bfc1)

dst_incompatible <- do.call("rbind",lapply(names(disttables), function(dst_name){
    dst <- data.table(as.data.frame(disttables[[dst_name]]))
    dst <- dst[which(!compatible)]
    rc.file <- info[info$rname == paste0(dst_name,"_readClassSe"),]$rpath
    rc <- readRDS(rc.file)
    rc <- data.table(as.data.frame(rowData(rc)), keep.rownames = TRUE)
    setnames(rc, "rn","readClassId")
    dst <- rc[dst, on = c("readClassId")]
    dst[, runname := dst_name]
    return(dst)
}))
saveRDS(dst_incompatible, file = "dst_incompatible.rds")
```

```{r}
dst_incompatible <- readRDS("dst_incompatible.rds")
ntx0 <- ntx_dt[ntx==0&(!is.na(runname))&(trimmed == FALSE)][,.(nperc,runname, protocol,ntotal)]

dst_incompatible <- ntx0[dst_incompatible, on = "runname"]
dst_incompatible[, readClassType := ifelse(i.readCount==1, "readCount1",
ifelse(grepl("unspliced",readClassId)&(novelGene),"unsplicedNovel",
ifelse(geneReadProp<0.05,"geneReadPropFilter",
ifelse(is.na(readCount)&is.na(i.readCount), "novelBySample","others")))), by = list(runname, nperc,ntotal, protocol, readClassId)]

data <- unique(dst_incompatible[,.(runname, readClassId, readClassType, i.readCount, nperc, ntotal, protocol)])
data_final <- unique(data[,list(nread = sum(i.readCount),
           percByType = sum(i.readCount)/ntotal), by = list(runname, readClassType, nperc, protocol, ntotal)])
data_final[, protocol := gsub("Stranded","",gsub("PromethionD","d",gsub("RandomPrimer","",protocol)))]
library(tidyr)
library(tidytext)
# protocols, celllines, spikein, reads
data_final[, readClassType_factor := factor(readClassType, rev(c("readCount1","unsplicedNovel","geneReadPropFilter","novelBySample","others")))]
data_final[, runname_reduced:=gsub("SGNex_|cDNA_|directcDNA_|cDNAStranded|directRNA_|licate","",runname)]
p_nperc <- ggplot(data_final, aes(x = reorder_within(runname_reduced,ntotal,protocol), y = percByType))+
    geom_bar(data = data_final,aes(fill = readClassType_factor),stat = "identity", position = position_stack())+
    geom_point(data = unique(data_final[,.(runname_reduced, ntotal, protocol, nperc)]),
               aes(x = reorder_within(runname_reduced,ntotal,protocol), y = nperc))+
    scale_x_reordered() +
    scale_fill_brewer(type = "qual",palette = 3, direction = -1)+
    facet_wrap(~protocol, scales = "free", ncol = 1)+
    xlab("Sample")+
    ylab("Percentage")+
    theme_minimal()


p_nperc_xlabel <- ggplot(data_final, aes(x = reorder_within(runname,ntotal,protocol), y = 1, fill = gsub("SGNex_|_(directRNA|cDNA|directcDNA).*|-(N082|N104|N122|EV)","",runname)))+
    geom_tile()+
     scale_x_reordered() +
     coord_flip()+
      facet_wrap(~protocol, scales = "free", ncol = 3)+
     scale_fill_brewer(type = "qual",palette = 3)+
     theme_minimal()
```


```{r}
pdf("r3.20_horizontal.pdf", width = 7, height = 7)
ggdraw() +
     draw_plot(p_nperc +theme(legend.position="none"),0,2/5, 1, 3/5) +
    draw_plot(p_number_of_transcripts_per_read_trimmed + theme(legend.position="none"),0, 1/5,1, 1/5)+
    draw_plot(as_ggplot(get_legend(p_nperc)),0,0,1/2,1/5)+
    draw_plot(as_ggplot(get_legend(p_number_of_transcripts_per_read_trimmed)),1/2,0,1/2,1/5)
dev.off()
```


# Supplementary Text Fig. 20
```{r}
pdf("r3.17_13Oct2023.pdf", width = 14, height = 10)
ggdraw() +
    draw_plot(p_sim_lr_vs_sr_cor_all +theme(legend.position="none"),0, 2/3, 1/5, 1/3) +
    draw_plot(p_sim_lr_vs_sr_cor + theme(legend.position="none"),1/5, 2/3,4/5, 1/3)+
     draw_plot(p_sim_lr_cor +theme(legend.position="none"),0, 1/3, 1/5, 1/3) +
     draw_plot(p_sim_vs_lr + theme(legend.position="none"),1/5, 1/2,4/5, 1/6)+
      draw_plot(p_sr_vs_lr + theme(legend.position="none"),1/5, 1/3,4/5, 1/6)+
     draw_plot(p_short_read + theme(legend.position="none"),0, 0,1, 1/3)
dev.off()
```

# Supplementary Text Fig. 10
10b-c
compare paired end sim vs long read, short read
```{r}
cat('Setting working directory')
wkdir <- '.'
general_list <- readRDS("general_list2023-04-27.rds")
dominant_typeData <- readRDS("ominant_typeData_25May2023.rds")

ensemblAnnotations.transcripts <- general_list$ensemblAnnotations.transcripts

plotdata <- comDataTranscriptTrimmed
plotdata[, method := gsub("_salmon","",gsub("trim_","",method))]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
setnames(ensemblAnnotations.transcripts, "gene_id","gene_name")
plotdata <- unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[plotdata, on = c("gene_name")]

samples <- general_list$samples
plotdata <- samples[plotdata, on = "runname"]
library(GGally)    
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
rep_dt <- unique(plotdata[,.(runname, method, cellLineRep)])
rep_dt[,nc := length(unique(method)), by = cellLineRep]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
repVec <- unique(rep_dt[nc==10]$cellLineRep)
test_runname <- unique(comDataTranscriptTrimmed[method == "lr_sim_pe"]$runname)
rv <- unique(rep_dt[runname == test_runname]$cellLineRep)

temp_data <- unique(plotdata[(runname == test_runname | (method == "salmon_sr"&cellLineRep == rv))&(gene_biotype %in% pro_types), 
    list(normEst = mean(normEst)), by = list(gene_name, tx_name, gene_biotype, method)])

temp_data_wide <- dcast(temp_data, gene_name + tx_name + gene_biotype  ~ method, value.var = "normEst")
    temp_data_wide <- dominant_typeData[cellLine == gsub("rep.*","", rv)][temp_data_wide, on = c("tx_name","gene_name")]
colIds <- c(21,24,23,22,26,25,19,17,18,20)

temp_data_wide[, majorType := ifelse(majorBoth, "majorBoth", 
    ifelse(majorEitherOnly&majorLongRead, "majorLongReadOnly",
           ifelse(majorEitherOnly&majorShortRead, "majorShortReadOnly","minor")))]

p_sim_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(lr_sim+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("fragmentation simulated sr vs lr")+theme_classic()
p_sim_pe_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(lr_sim_pe+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("fragmentation simulated sr vs lr")+theme_classic()    

p_sr_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(salmon_sr+1)))+
       geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("original sr vs lr")+theme_classic()

p_sr_vs_sim <- ggplot(temp_data_wide, aes(x = log2(lr_sim + 1), y = log2(salmon_sr+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("original sr vs fragmentation simulated")+theme_classic()

p_sr_vs_sim_pe <- ggplot(temp_data_wide, aes(x = log2(lr_sim_pe + 1), y = log2(salmon_sr+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("original sr vs fragmentation simulated")+theme_classic()

p_sim_vs_sim_pe <- ggplot(temp_data_wide, aes(x = log2(lr_sim_pe + 1), y = log2(lr_sim+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("original sr vs fragmentation simulated")+theme_classic()

pdf("comment_figure_simulation_paired_end.pdf", width = 16, height = 6)
ggarrange(p_sim_vs_lr+theme(legend.position="none"),
          p_sim_pe_vs_lr+theme(legend.position="none"),
p_sr_vs_sim+theme(legend.position="none"),
p_sr_vs_sim_pe+theme(legend.position="none"),nrow = 2, ncol = 2, align = "hv")
dev.off()
```



compare sim short read vs short read (using public short read)
```{r}
cat('Setting working directory')
wkdir <- 'wkdir'
general_list <- readRDS("general_list2023-04-27.rds")
dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")

ensemblAnnotations.transcripts <- general_list$ensemblAnnotations.transcripts

ensemblAnnotations.transcripts <- general_list$ensemblAnnotations.transcripts

plotdata <- comDataTranscriptTrimmed
plotdata[, method := gsub("_salmon","",gsub("trim_","",method))]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
setnames(ensemblAnnotations.transcripts, "gene_id","gene_name")
plotdata <- unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[plotdata, on = c("gene_name")]

samples <- general_list$samples
plotdata <- samples[plotdata, on = "runname"]
library(GGally)    
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
rep_dt <- unique(plotdata[,.(runname, method, cellLineRep)])
rep_dt[,nc := length(unique(method)), by = cellLineRep]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
repVec <- unique(rep_dt[nc==10]$cellLineRep)

rv <- repVec[1]

temp_data <- unique(plotdata[((method %in% c("kallisto_sr_public","rsem_sr_public")&i.cellLine == "A549")|(method %in% c("salmon_sr","salmon_lr","lr_sim")&(cellLineRep==rv)))&(gene_biotype %in% pro_types), 
    list(normEst = mean(normEst)), by = list(gene_name, tx_name, gene_biotype, method)])

temp_data_wide <- dcast(temp_data, gene_name + tx_name + gene_biotype  ~ method, value.var = "normEst")
    temp_data_wide <- unique(dominant_typeData[cellLine == gsub("rep.*","", rv),.(tx_name, gene_name,majorBoth, majorEitherOnly,majorLongRead,majorShortRead)])[temp_data_wide, on = c("tx_name","gene_name")]
colIds <- c(21,24,23,22,26,25,19,17,18,20)

temp_data_wide[, majorType := ifelse(majorBoth, "majorBoth", 
    ifelse(majorEitherOnly&majorLongRead, "majorLongReadOnly",
           ifelse(majorEitherOnly&majorShortRead, "majorShortReadOnly","minor")))]

temp_data_wide <- temp_data_wide[!is.na(majorType)]
p_sim_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(lr_sim+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("fragmentation simulated sr vs lr")+theme_classic()
  

p_sr_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(salmon_sr+1)))+
       geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("original sr vs lr")+theme_classic()

p_sr_encode_kallisto_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(kallisto_sr_public+1)))+
       geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("encode sr (kallisto) vs lr")+theme_classic()

p_sr_encode_rsem_vs_lr <- ggplot(temp_data_wide, aes(x = log2(salmon_lr+1), y = log2(rsem_sr_public+1)))+
       geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("encode sr (rsem) vs lr")+theme_classic()


p_sr_vs_sim <- ggplot(temp_data_wide, aes(x = log2(lr_sim + 1), y = log2(salmon_sr+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("original sr vs fragmentation simulated")+theme_classic()

p_sr_kallisto_public_vs_sim <- ggplot(temp_data_wide, aes(x = log2(lr_sim + 1), y = log2(kallisto_sr_public+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("encode kallisto vs fragmentation simulated")+theme_classic()

p_sr_rsem_public_vs_sim <- ggplot(temp_data_wide, aes(x = log2(lr_sim + 1), y = log2(rsem_sr_public+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~majorType, nrow = 1)+
        ggtitle("encode rsem vs fragmentation simulated")+theme_classic()

pdf("comment_figure_simulation_vs_public_a549.pdf", width = 16, height = 12)
ggarrange(
          #p_sim_vs_lr+theme(legend.position="none"),
          p_sr_vs_lr+theme(legend.position="none"),
           p_sr_vs_sim +theme(legend.position="none"),
          p_sr_encode_rsem_vs_lr+theme(legend.position="none"),
          p_sr_rsem_public_vs_sim+theme(legend.position="none"),
          p_sr_encode_kallisto_vs_lr+theme(legend.position="none"),
p_sr_kallisto_public_vs_sim+theme(legend.position="none"),
nrow = 3, ncol = 2, align = "hv")
dev.off()
```

compare pacbio vs short read, pacbio trimmed vs short read
```{r}
cat('Setting working directory')
wkdir <- 'wkdir'
general_list <- readRDS("general_list2023-04-27.rds")
dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")

ensemblAnnotations.transcripts <- general_list$ensemblAnnotations.transcripts

plotdata <- copy(comDataTranscriptTrimmed)
plotdata[, method := gsub("_salmon","",gsub("trim_","",method))]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
setnames(ensemblAnnotations.transcripts, "gene_id","gene_name")
plotdata <- unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[plotdata, on = c("gene_name")]

samples <- general_list$samples
plotdata <- samples[plotdata, on = "runname"]
library(GGally)    
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
rep_dt <- unique(plotdata[,.(runname, method, cellLineRep)])
rep_dt[,nc := length(unique(method)), by = cellLineRep]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
repVec <- unique(rep_dt[nc==10]$cellLineRep)
test_runname <- unique(plotdata[method == "lr_sim_pb"]$runname)

rc <- "SGNex_MCF7_Illumina_replicate2_run1" 
temp_data <- unique(plotdata[(runname == test_runname | runname == rc)&(gene_biotype %in% pro_types), 
    list(normEst = mean(normEst)), by = list(gene_name, tx_name, gene_biotype, method)])

temp_data_wide <- dcast(temp_data, gene_name + tx_name + gene_biotype  ~ method, value.var = "normEst")
    temp_data_wide <- dominant_typeData[cellLine == "MCF7"][temp_data_wide, on = c("tx_name","gene_name")]

temp_data_wide[, majorType := ifelse(majorBoth, "majorBoth", 
    ifelse(majorEitherOnly&majorLongRead, "majorLongReadOnly",
           ifelse(majorEitherOnly&majorShortRead, "majorShortReadOnly","minor")))]

p_sr_vs_sim_pb <- ggplot(temp_data_wide, aes(x = log2(lr_sim_pb + 1), y = log2(salmon_sr+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
        ggtitle("Illumina vs fragmented PacBio")+theme_classic()
p_sr_vs_pb <- ggplot(temp_data_wide, aes(x = log2(salmon_lr + 1), y = log2(salmon_sr+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
        ggtitle("Illumina vs PacBio")+theme_classic()

pdf("comment_figure_fragmentation_for_pacbio.pdf", width = 8, height = 6)
ggarrange(p_sr_vs_sim_pb+theme(legend.position="none"),
          p_sr_vs_pb +theme(legend.position="none"),nrow = 2, ncol = 1, align = "hv")
dev.off()
```


```{r}
plotdata <- copy(comDataGeneTrimmed)
plotdata[, method := gsub("_salmon","",gsub("trim_","",method))]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
setnames(ensemblAnnotations.transcripts, "gene_id","gene_name")
plotdata <- unique(ensemblAnnotations.transcripts[,.(gene_name, gene_biotype)])[plotdata, on = c("gene_name")]

samples <- general_list$samples
plotdata <- samples[plotdata, on = "runname"]
library(GGally)    
plotdata[, cellLineRep := paste0(cellLine, bioRep)]
rep_dt <- unique(plotdata[,.(runname, method, cellLineRep)])
rep_dt[,nc := length(unique(method)), by = cellLineRep]
pro_types <- c("protein_coding","antisense_RNA","lincRNA","non_coding","macro_lncRNA")
repVec <- unique(rep_dt[nc==10]$cellLineRep)
test_runname <- unique(plotdata[method == "lr_sim_pb"]$runname)

rc <- "SGNex_MCF7_Illumina_replicate2_run1" 
temp_data <- unique(plotdata[(runname == test_runname | runname == rc)&(gene_biotype %in% pro_types), 
    list(normEst = mean(normEst)), by = list(gene_name, gene_biotype, method)])

temp_data_wide <- dcast(temp_data, gene_name + gene_name + gene_biotype  ~ method, value.var = "normEst")
temp_data_wide[, gene_biotype_reclass:= ifelse(gene_biotype %in% c("protein_coding","lincRNA"), gene_biotype,"others")]
p_sr_vs_sim_pb_gene <- ggplot(temp_data_wide, aes(x = log2(lr_sim_pb + 1), y = log2(salmon_sr+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~gene_biotype_reclass, nrow = 1)+
        ggtitle("Illumina vs fragmented PacBio")+theme_classic()
p_sr_vs_pb_gene <- ggplot(temp_data_wide, aes(x = log2(salmon_lr + 1), y = log2(salmon_sr+1)))+
        geom_hex(binwidth = c(0.1, 0.1))+ 
        scale_fill_steps(
           breaks = c(10,100,1000,10000),
            low = "steelblue1", high = "steelblue4",trans = "log10")+
         ggpubr::stat_cor(aes(label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
    facet_wrap(~gene_biotype_reclass, nrow = 1)+
        ggtitle("Illumina vs PacBio")+theme_classic()

pdf("comment_figure_fragmentation_for_pacbio_gene.pdf", width = 8, height = 6)
ggarrange(p_sr_vs_sim_pb_gene+theme(legend.position="none"),
          p_sr_vs_pb_gene +theme(legend.position="none"),nrow = 2, ncol = 1, align = "hv")
dev.off()
```



10a short read insert size distribution for one sample 
```{r}
insertsizes_distribution <- fread("sr_stats.txt", skip = 49, fill = TRUE)
is <- insertsizes_distribution[V1 == "IS",1:5, with =FALSE]
isVec <- rep(as.numeric(is$V2), is$V3)

pdf("insert_size_distribution.pdf", width =6, height = 5)
ggplot(is[V3>0], aes(x = as.numeric(V2), y = V3/sum(V3)))+geom_bar(stat = "identity", col = "light grey", fill = "light grey", alpha = 0.5)+xlab("Estimated insert sizes")+ylab("Proportion of reads")+geom_vline(xintercept = c(151,201,221.3), col = "grey", linetype = 3)+theme_classic()
dev.off()
```


## # of isoforms in each category
```{r}
dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")
ddData <- unique(dominant_typeData[,
                .(cellLine, tx_name, gene_name,majorBoth, majorEither, majorEitherOnly, majorSecBoth, majorLongRead, majorShortRead)])
ddData[, major_status := ifelse(majorEitherOnly&majorLongRead, "majorLongReadOnly",ifelse(majorEitherOnly&majorShortRead, "majorShortReadOnly", ifelse(majorBoth, "majorBoth","others")))]
ddData <- ddData[,.(tx_name, gene_name, cellLine, major_status), with = TRUE]

ntxDt <- unique(ddData[,list(ntx = .N), by = list(cellLine, major_status)])

ntxDt <- ntxDt[cellLine %in% cellLines]
ntxDt[, meanNtx := mean(ntx), by = major_status]
ntxDt[, sdNtx := sd(ntx), by = major_status]


ntxDt_union <- unique(ddData[,list(ntx = length(unique(tx_name)),ngene = length(unique(gene_name))), by = list(major_status)])
```



## prepare data for number of junctions
```{r junctions-number}
data_lr <- readRDS(paste0(wkdir,"juncLR_May6.rds"))
data_lr_pacbio = readRDS(paste0(wkdir,"juncLR_pacbio_Apr25.rds"))
data_sr = readRDS(paste0(wkdir,"juncSR.rds"))
setnames(data_sr, "runName", "old_runname")
data_sr <- samples_wSpikein[,.(runname, old_runname)][data_sr, on = "old_runname"]
data_sr[, old_runname := NULL]
setnames(data_sr, "runname", "runName")
data_lr_spikein = readRDS(paste0(wkdir,"juncLR_spikein_Apr25.rds"))
data_sr_spikein = readRDS(paste0(wkdir,"juncSR_spikein.rds"))
data_sr_spikein[, runName := gsub("\\,","",runName)]

dt <- rbindlist(list(data_sr,data_lr, data_sr_spikein, data_lr_spikein, data_lr_pacbio), fill = TRUE)
dt[, runName := gsub("_sorted","",runName)]
dt <- dt[runName != "allSpikinReadsCombined_SIRV-1_PacBio"]
dt[,runName := gsub("sequinMixAv1.0_PacBio","sequinMixAV2_PacBio",runName)]
setnames(dt, "runName","runname")
dt[, cellLine:=gsub("-EV","",gsub('k562','K562',strsplit(runname, '\\_')[[1]][2])),by = runname]
dt[, protocol:=gsub("-SMRTcell","",strsplit(runname, '\\_')[[1]][3]), by = runname]
dt[, njunc:=as.numeric(njunc)]
dt[, nread_sum := sum(nread), by = runname]
dt[, nperc := nread/nread_sum, by = list(runname, cellLine, protocol)]

dt[, njunc_cat := cut(njunc, breaks = c(-0.01,0,1,6,10,(2:15)*10,187))]

dt <- rbind(samples[,.(runname, protocol_type, Platform, demultiplexed, cellLine)],
                    samples_wSpikein[grepl("Illumina",runname),.(runname, protocol_type, Platform, demultiplexed, cellLine)])[dt, on = "runname"]
dt[, protocol_type := gsub("-SMRTcell","",protocol_type)]
saveRDS(dt, file = "junc_dt_26June2023.rds")
```


```{r}
plotData <- unique(dt[, list(count=sum(nread)), by = list(njunc_cat,runname,protocol_type,cellLine)])
plotData[,total_count:= sum(count), by = list(protocol_type,runname)]
plotData[, frac_count:=count/total_count]
plotData[, protocol_type_factor:=factor(protocol_type, levels = protocolVec )]

p_njunc <- ggplot(plotData[!grepl("allSpikein",runname)], 
                  aes(x = njunc_cat,
                           y = count/total_count,
                           fill = protocol_type_factor,
                           col = protocol_type_factor))+
     geom_boxplot(aes(fill = protocol_type_factor,col = protocol_type_factor), width = 0.5, position=position_dodge(0.75), outlier.shape = NA)+
      scale_fill_manual(values = protocolCol,
                      name = "Protocol",
                      limits = protocolVec,
                      labels = protocolLabel)+
    scale_color_manual(values = protocolCol,
                       name = "Protocol",
                       limits = protocolVec,
                       labels = protocolLabel)+
    stat_summary(
        fun = median,
        geom = 'line',
        aes(group = protocol_type_factor, colour = protocol_type_factor),
        position = position_dodge(width = 0.9) #this has to be added
    )+
    labs(title = 'All cell lines',
         x = 'Number of junctions',
         y = 'Percentage over all reads')+
    theme_classic()

p_njunc
```




## number of transcripts compatible per read 
```{r number-of-transcripts-per-read, eval = FALSE}
seOutput <- readRDS("bambuOutput_May25.rds")
disttables <- metadata(seOutput)$distTables
dst_ntx <- do.call("rbind",lapply(names(disttables), function(dst_name){
    dst <- data.table(as.data.frame(disttables[[dst_name]]))
    dst_ntx <- unique(dst[, list(ntx =length(unique(.SD[which(compatible)]$annotationTxId))), by = list(readClassId, readCount)])
    dst_ntx[, runname := dst_name]
    return(dst_ntx)
}))
saveRDS(dst_ntx, file = "number_of_transcripts_per_read_lr.rds")


seOutput <- readRDS("bambuOutput_PacBio_May22.rds")
disttables <- metadata(seOutput)$distTables
dst_ntx <- do.call("rbind",lapply(names(disttables), function(dst_name){
    dst <- data.table(as.data.frame(disttables[[dst_name]]))
    dst_ntx <- unique(dst[, list(ntx =length(unique(.SD[which(compatible)]$annotationTxId))), by = list(readClassId, readCount)])
    dst_ntx[, runname := dst_name]
    return(dst_ntx)
}))
saveRDS(dst_ntx, file = "number_of_transcripts_per_read_lr_pacbio.rds")



seSpikein_lr <- readRDS("bambuOutput_spikein_bam_ont_May22.rds")
disttables <- metadata(seSpikein_lr)$distTables
dst_ntx <- do.call("rbind",lapply(names(disttables), function(dst_name){
    dst <- data.table(as.data.frame(disttables[[dst_name]]))
   dst_ntx <- unique(dst[, list(ntx =length(unique(.SD[which(compatible)]$annotationTxId))), by = list(readClassId, readCount)])
    dst_ntx[, runname := dst_name]
    return(dst_ntx)
}))
saveRDS(dst_ntx, file = "number_of_transcripts_per_read_spikein_lr_ont.rds")

seSpikein_lr <- readRDS("bambuOutput_spikein_bam_pacbio_May22.rds")
disttables <- metadata(seSpikein_lr)$distTables
dst_ntx <- do.call("rbind",lapply(names(disttables), function(dst_name){
    dst <- data.table(as.data.frame(disttables[[dst_name]]))
   dst_ntx <- unique(dst[, list(ntx =length(unique(.SD[which(compatible)]$annotationTxId))), by = list(readClassId, readCount)])
    dst_ntx[, runname := dst_name]
    return(dst_ntx)
}))
saveRDS(dst_ntx, file = "number_of_transcripts_per_read_spikein_lr_pacbio.rds")


seSpikein_sr <- readRDS("bambuOutput_spikein_bam_sr.rds")
disttables <- metadata(seSpikein_sr)$distTables
dst_ntx <- do.call("rbind",lapply(names(disttables), function(dst_name){
    dst <- data.table(as.data.frame(disttables[[dst_name]]))
    dst_ntx <- unique(dst[, list(ntx =length(unique(.SD[which(compatible)]$annotationTxId))), by = list(readClassId, readCount)])
    dst_ntx[, runname := dst_name]
    return(dst_ntx)
}))
saveRDS(dst_ntx, file = "number_of_transcripts_per_read_spikein_sr.rds")


sr_rds <- dir("srSe_revised", pattern = "genome.rds", full.names = TRUE)
dst_ntx_sr <- do.call("rbind",lapply(sr_rds, function(k){
    srSe <- readRDS(k)
    disttable <- metadata(srSe)$distTables[[1]]
    dst <- data.table(as.data.frame(disttable))
   dst_ntx <- unique(dst[, list(ntx =length(unique(.SD[which(compatible)]$annotationTxId))), by = list(readClassId, readCount)])
    dst_ntx[, runname := gsub("_genome.rds","",basename(k))]
    return(dst_ntx)
}))
saveRDS(dst_ntx_sr, file = "number_of_transcripts_per_read_sr.rds")


# trim_reads--- need to check again 
sr_rds <- dir("trim_reads/lr_150bpSingleEnd_1ts_incompatible/03_Bambu/se/", pattern = "seOutput.rds", full.names = TRUE)
dst_ntx_sr <- do.call("rbind",lapply(sr_rds, function(k){
    srSe <- readRDS(k)
    disttable <- metadata(srSe)$distTables[[1]]
    dst <- data.table(as.data.frame(disttable))
   dst_ntx <- unique(dst[, list(ntx =length(unique(.SD[which(compatible)]$annotationTxId))), by = list(readClassId, readCount)])
    dst_ntx[, runname := gsub("_seOutput.rds","",basename(k))]
    return(dst_ntx)
}))
saveRDS(dst_ntx_sr, file = "number_of_transcripts_per_read_lr_trimmed_reads_updated.rds")

```


```{r}
dst_ntx_lr <- readRDS("number_of_transcripts_per_read_lr.rds")
dst_ntx_lr_trimmed <- readRDS("number_of_transcripts_per_read_lr_trimmed_reads_updated.rds")
dst_ntx_lr_trimmed[, trimmed := TRUE]
dst_ntx_lr_pacbio <- readRDS("number_of_transcripts_per_read_lr_pacbio.rds")
dst_ntx_sr <- readRDS("number_of_transcripts_per_read_sr.rds")
dst_ntx_spikein_lr_ont <- readRDS("number_of_transcripts_per_read_spikein_lr_ont.rds")
dst_ntx_spikein_lr_pacbio <- readRDS("number_of_transcripts_per_read_spikein_lr_pacbio.rds")
dst_ntx_spikein_sr <- readRDS("number_of_transcripts_per_read_spikein_sr.rds")
dst_ntx <- rbindlist(list(dst_ntx_lr, dst_ntx_lr_trimmed, dst_ntx_lr_pacbio, dst_ntx_sr, dst_ntx_spikein_lr_ont, dst_ntx_spikein_lr_pacbio, dst_ntx_spikein_sr), fill = TRUE)
dst_ntx[is.na(trimmed), trimmed := FALSE]
ntx_dt <- unique(dst_ntx[, list(nread=sum(readCount)), by = list(ntx,runname, trimmed)])

ntx_dt[, ntotal:=sum(nread), by = list(runname, trimmed)]
ntx_dt[, runname := gsub("GIS_Hct116_PacBio-SMRTcell_Rep5_Run1","GIS_HCT116_PacBio-SMRTcell_Rep7_Run1",gsub("HCT116","Hct116",gsub("_genome.rds|STAR_alignment|_sorted","", runname)))]
ntx_dt[, old_runname := runname]
temp <- samplesRC_combined[,.(old_runname, runname)]
setnames(temp, "runname", "new_name")
ntx_dt <- temp[ntx_dt, on = "old_runname"]
ntx_dt[!is.na(new_name), runname := new_name]
ntx_dt[, `:=`(old_runname = NULL, new_name = NULL)]
ntx_dt <- samplesRC_combined[ntx_dt, on = "runname"]

ntx_dt[, ntx_cat:=ifelse(ntx>10, ">10",ntx)]
ntx_dt[, nperc := nread/ntotal]

ntx_dt[, alpha_value := ifelse(trimmed, 0.5, 1)]
ntx_dt_sum <- unique(ntx_dt[, list(nperc = sum(nperc)), by = list(ntx_cat, protocol_type_factor, runname, trimmed)])

saveRDS(ntx_dt_sum, file = "ntx_dt_sum_26June2023.rds")
p_number_of_transcripts_per_read_trimmed <- ggplot(ntx_dt_sum[!grepl("all",runname)], aes(x = ntx_cat, y = nperc,  fill = paste0(protocol_type_factor,trimmed), col =  paste0(protocol_type_factor,trimmed)))+ 
    geom_boxplot(aes(alpha =1-trimmed/2), width = 0.5, position=position_dodge(0.75), outlier.shape = NA)+ # outliers removed already, otherwise, the outlier points for directRNA will be very high up to 30/40%
     stat_summary(
        fun = median,
        geom = 'line',
        aes(group = paste0(protocol_type_factor, trimmed), colour = paste0(protocol_type_factor,trimmed)),
        position = position_dodge(width = 0.75) #this has to be added
    )+
    scale_fill_brewer(type = "qual", palette = 3)+
    scale_color_brewer(type = "qual", palette = 3)+
    scale_x_discrete(limits = c(0:10,">10"))+
    ylab("Relative proportion of reads")+
    xlab("Number of transcripts per read aligned to")+
    theme_classic()

```


 
# spike in split by CPM < 2.5 and above 2.5
```{r}
plotdata_tx <- si_data_transcript[method %in% c("salmon_lr","salmon_sr")&!(protocol %in% c("directRNA","PacBio"))&(cpm_norm_conc<2.5)]
plotdata_tx[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_tx_spikein_scatter_cpm0_2.5 <- ggplot(plotdata_tx, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = spike_in_version), size = 0.5, se = FALSE )+
                geom_point(aes(col = spike_in_version),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = spike_in_version,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(spike_in_general_name_revised~protocol)+
                theme_classic()

plotdata_tx <- si_data_transcript[method %in% c("salmon_lr","salmon_sr")&!(protocol %in% c("directRNA","PacBio"))&(cpm_norm_conc>=2.5)]
plotdata_tx[, spike_in_general_name_revised := factor(spike_in_general_name_revised, 
            levels = c("Sequin","ERCC","SIRV"))]
p_tx_spikein_scatter_cpm2.5 <- ggplot(plotdata_tx, aes(x=log2(cpm_norm_conc+1), y=log2(normEst+1))) +
                 geom_smooth(aes(col = spike_in_version), size = 0.5, se = FALSE )+
                geom_point(aes(col = spike_in_version),
                           alpha = 0.5,size = 2) +
                geom_abline(intercept = 0, slope = 1, col = "grey")+
                xlab('Expected CPM (log2)')+
                ylab('Observed CPM (log2)')+
                scale_color_brewer(type = "qual", palette = 3)+
                ggpubr::stat_cor(aes(col = spike_in_version,
                                     label = ..r.label..),method = "spearman", cor.coef.name = "Sp.R")+ 
                facet_grid(spike_in_general_name_revised~protocol)+
                theme_classic()
```

## check samples for figure 8 used
```{r}
figure8_samples <- readxl::read_xlsx("Supplementary_Table11.xlsx")
```


## Prepare candidates for dPCR and qPCR experiments
find gene candidates with dominant isoform switch between long and short reads 
```{r}
dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")
ddData <- unique(dominant_typeData[,
                .(cellLine, tx_name, gene_name,majorBoth, majorEither, majorEitherOnly, majorSecBoth, majorLongRead, majorShortRead)])
tt <- dcast(ddData[which(majorEitherOnly)], gene_name + cellLine ~ majorLongRead, value.var = "tx_name")
setnames(tt, 3:4, c("sr","lr"))
devtools::load_all("bambu")
seOutput <- readRDS("bambuOutput_May25.rds")
annotations <- rowRanges(seOutput)

comDataTranscript <- readRDS(paste0(wkdir, "combinedExpressionDataTranscript_19June2023.rds"))
comDataGene <- readRDS(paste0(wkdir, "combinedExpressionDataGene_19June2023.rds"))

comDataGene[, old_runname := runname]
comDataGene <- temp[comDataGene, on = "old_runname"]
comDataGene[!is.na(new_name), runname := new_name]
comDataGene[, `:=`(old_runname = NULL, new_name = NULL)]
comDataGene[runname == "GIS_HCT116_PacBio-SMRTcell_Rep5_Run1", runname := "SGNex_Hct116_PacBio-SMRTcell_replicate7_run1"]
comDataGene[, cellLine := gsub("HCT116","Hct116",cellLine)]

comDataTranscript[, old_runname := runname]
comDataTranscript <- temp[comDataTranscript, on = "old_runname"]
comDataTranscript[!is.na(new_name), runname := new_name]
comDataTranscript[, `:=`(old_runname = NULL, new_name = NULL)]
comDataTranscript[runname == "GIS_HCT116_PacBio-SMRTcell_Rep5_Run1", runname := "SGNex_Hct116_PacBio-SMRTcell_replicate7_run1"]
comDataTranscript[, cellLine := gsub("HCT116","Hct116",cellLine)]
```


```{r}
# use full-length read support to filter 
fullLengthReads <- data.table(as.data.frame(assays(seOutput)$fullLengthCounts), keep.rownames = TRUE)
uniqueReads <- data.table(as.data.frame(assays(seOutput)$uniqueCounts), keep.rownames = TRUE)
```


```{r}
estDataMean <- unique(comDataTranscript[,list(meanEst = mean(normEst)), by = list(tx_name,cellLine,protocol_general,method)], by = NULL)

tt_mcf7 <- melt(tt[cellLine == "MCF7"], id.vars = c("gene_name"), measure.vars = c("lr","sr"))
setnames(tt_mcf7,"value","tx_name")

tt_mcf7 <- estDataMean[method == "salmon_lr"&(protocol_general == "cDNA")&(cellLine == "MCF7"),.(tx_name,meanEst)][tt_mcf7, on = "tx_name"]
setnames(tt_mcf7, "meanEst","lrEst")
tt_mcf7 <- estDataMean[method == "salmon_sr"&(cellLine == "MCF7"),.(tx_name,meanEst)][tt_mcf7, on = "tx_name"]
setnames(tt_mcf7, "meanEst","srEst")
```

```{r}
tt_mcf7[order(lrEst)]
```


```{r}
candidateList_srbased <- tt_mcf7[order(srEst, decreasing = TRUE)][1:20]$gene_name
for(gene in candidateList_srbased){
    print(tt_mcf7[gene_name == gene])
}
```
```{r}
intersect(candidateList, candidateList_srbased) # 11 overlap 
setdiff(candidateList_srbased, candidateList) # sr only
setdiff(candidateList, candidateList_srbased) # lr only
```


```{r}
candidateList <- tt_mcf7[order(lrEst, decreasing = TRUE)][1:20]$gene_name
candidateList <- c(candidateList,setdiff(candidateList_srbased, candidateList))
for(gene in candidateList){
    print(tt_mcf7[gene_name == gene])
}
```
```{r}
library(ggbio)
for(gene in candidateList[-c(2,3,11)]){
    major_tx <- tt_mcf7[gene_name == gene]$tx_name
    p_major <- autoplot(rowRanges(seOutput)[major_tx], group.selfish = TRUE)
    other_tx <- setdiff(rowData(seOutput)[rowData(seOutput)$GENEID == gene,]$TXNAME,major_tx)
    if(length(other_tx)>0){
        p_other <- autoplot(rowRanges(seOutput)[other_tx], group.selfish = TRUE)
        print(tracks(p_major, p_other))
    }else{
        print(tracks(p_major))
    }
    
}

```

```{r}
for(gene in candidateList[-c(2,3,11)]){
    major_tx <- tt_mcf7[gene_name == gene]$tx_name
print(estDataMean[protocol_general == "cDNA"][tx_name %in% major_tx][cellLine == "MCF7"])
}
```
```{r}
for(gene in candidateList[-c(2,3,11,15,16)]){
    major_tx <- tt_mcf7[gene_name == gene]$tx_name
    other_tx <- setdiff(rowData(seOutput)[rowData(seOutput)$GENEID == gene,]$TXNAME,major_tx)
print(estDataMean[protocol_general == "cDNA"][tx_name %in%other_tx][cellLine == "MCF7"])
}
```


```{r}
library(Biostrings)
txSeqFile <- "hg38_sequins_SIRV_ERCCs_longSIRVs_cdna.fa"
txSeq <- readDNAStringSet(file=txSeqFile)
txSeqNames <- names(txSeq)
listNames <- unlist(lapply(strsplit(txSeqNames," "),'[[',1))

gcDt <- data.table(annotationTxId = listNames, gc_perc=letterFrequency(txSeq, letters = "CG", as.prob = TRUE))
```

```{r}
for(gene in candidateList[-c(2,3,11)]){
     major_tx <- tt_mcf7[gene_name == gene]$tx_name
     other_tx <- setdiff(rowData(seOutput)[rowData(seOutput)$GENEID == gene,]$TXNAME,major_tx)
     print(gcDt[annotationTxId %in% major_tx])
}
```
```{r}
for(gene in candidateList[-c(2,3,11)]){
    major_tx <- tt_mcf7[gene_name == gene]$tx_name
print_data <- fullLengthReads[rn %in% major_tx, c(1,which(grepl("MCF7",colnames(fullLengthReads))&grepl("_cDNA",colnames(fullLengthReads)))), with = FALSE]
 setnames(print_data, colnames(print_data), gsub("SGNex_MCF7|licate|_","",colnames(print_data)))
    print(print_data)
}
```
```{r}
for(gene in candidateList[-c(2,3,11)]){
    major_tx <- tt_mcf7[gene_name == gene]$tx_name
    print_data <- uniqueReads[rn %in% major_tx, c(1,which(grepl("MCF7",colnames(uniqueReads))&grepl("_cDNA",colnames(uniqueReads)))), with = FALSE]
    setnames(print_data, colnames(print_data), gsub("SGNex_MCF7|licate|_","",colnames(print_data)))
    print(print_data)
}
```

get the unique sequences for each of the isoforms in the major isoform pair
```{r}
candidateGene <- c("ENSG00000186468",
                   "ENSG00000144713",
                   "ENSG00000071082",
                   "ENSG00000149273",
                   "ENSG00000196262",
                   "ENSG00000111640",
                   "ENSG00000197756",
                   "ENSG00000167526",
                   "ENSG00000164587",
                   "ENSG00000133112",
                   "ENSG00000137154",
                   "ENSG00000167996")

for(gene in candidateGene){
    print(rowRanges(seOutput)[tt_mcf7[gene_name == gene]$tx_name])
}
```
rep ratio 
```{r}
path.repeatMasker <- "repeatMasker_Grch38_ucsc.txt.gz" 
df <- fread(path.repeatMasker)
ervRanges <- makeGRangesFromDataFrame(df, keep.extra.columns=TRUE,
                                      seqnames.field="genoName",
                                      start.field="genoStart",
                                      end.field="genoEnd",
                                      strand.field="strand",
                                      starts.in.df.are.0based=TRUE)
names(ervRanges) <- paste0('rep',1:length(ervRanges))
ignore.strand <- FALSE
repResOnt <- repeat_analysis_function(seOutput, ervRanges)
```


unique sequences 
```{r}
library(Biostrings)
geneSeqFile <- "hg38_sequins_SIRV_ERCCs_longSIRVs.fa"
geneSeq <- readDNAStringSet(file=geneSeqFile)
geneSeqNames <- names(geneSeq)
listNames <- unlist(lapply(strsplit(geneSeqNames," "),'[[',1))

candidateGene <- c("ENSG00000186468",
                   "ENSG00000144713",
                   "ENSG00000071082",
                   "ENSG00000149273",
                   "ENSG00000196262",
                   "ENSG00000111640",
                   "ENSG00000197756",
                   "ENSG00000167526",
                   "ENSG00000164587",
                   "ENSG00000137154",
                   "ENSG00000167996",
                   "ENSG00000147604",
                   "ENSG00000145592")

setnames(gcDt, c(1,2), c("tx_name","gc_perc"))
source("temp.R")
    print(geneidd)
    gene <- candidateGene[geneidd]
    print(rowRanges(seOutput)[tt_mcf7[gene_name == gene]$tx_name])
    lrgr <- rowRanges(seOutput)[[tt_mcf7[gene_name == gene&(variable == "lr")]$tx_name]]
    srgr <- rowRanges(seOutput)[[tt_mcf7[gene_name == gene&(variable == "sr")]$tx_name]]
    autoplot(rowRanges(seOutput)[tt_mcf7[gene_name == gene]$tx_name])
    chrname <- as.character(unlist(unique(seqnames(lrgr))))
    strandtemp <- as.character(unique(strand(lrgr)))
    seqid <- match(as.character(unique(seqnames(lrgr))),listNames)
    tt_mcf7[gene_name == gene]
    if(geneidd==1){
        # enlonged second exon of sr isoform 
        lrposStart <- 82276061
        lrposEnd <-end(srgr)[2]#82276177
        lrUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][lrposStart:lrposEnd])))
       
        
        # first and second exon of sr isoform
        srposStart <- start(srgr)[1]
        srposEnd <- end(srgr)[2] 
        srUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]])))
         uniquePos <- c(79,195)
        substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
         uniquePos <- c(79,80)
         substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
         
         lrExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),as.character(geneSeq[[seqid]][start(lrgr)[4]:82278354])))
         srExtendSeq <- lrExtendSeq
             
         lrCompleteSeq <- paste0(lrExtendSeq, lrUniqueSeq)
         srCompleteSeq <- paste0(srExtendSeq, srUniqueSeq)
    }
    
     if(geneidd==2){
     lrUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]])))
     srUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]]),as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]])))
     lrposStart <- start(lrgr)[3]
     lrposEnd <- end(lrgr)[4]
     srposStart <- start(srgr)[3]
     srposEnd <- end(srgr)[4]
     
     # find unique sequences
     uniquePos <- sort(nchar(lrUniqueSeq)-c(101,102)+1)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     uniquePos <- sort(nchar(srUniqueSeq)-c(102,212)+1)
      substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
      
     
         srExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]])))
           lrExtendSeq <- srExtendSeq
             
         lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
     
    }
     if(geneidd==3){
     lrposStart <- 101002289
     lrposEnd <- end(lrgr)[2]
     lrUniqueSeq <- paste0(as.character(geneSeq[[seqid]][lrposStart:end(lrgr)[1]]),as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]))
     srUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]))
     srposStart <- start(lrgr)[2]
     srposEnd <- end(lrgr)[4]
     uniquePos <- c(1,28)
      substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     lrExtendSeq <-paste0(as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]),
                          as.character(geneSeq[[seqid]][start(lrgr)[5]:101006423]))
           srExtendSeq <- ""
         lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
    }
     if(geneidd==4){
     lrposStart <- start(lrgr)[6]
     lrposEnd <- 75405692
     lrUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(lrgr)[6]:end(lrgr)[6]]),
                      as.character(geneSeq[[seqid]][start(lrgr)[7]:lrposEnd]))
     srUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]),as.character(geneSeq[[seqid]][start(lrgr)[5]:end(lrgr)[5]]),as.character(geneSeq[[seqid]][start(lrgr)[6]:end(lrgr)[6]]))
    
     srposStart <- start(lrgr)[4]
     srposEnd <- end(lrgr)[6]
     uniquePos <- c(197,198)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     lrExtendSeq <-paste0(as.character(geneSeq[[seqid]][75399518:end(lrgr)[1]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[5]:end(lrgr)[5]]))
      srExtendSeq <- paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),
                           as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]]),
                           as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]]))
         lrCompleteSeq <- paste0(lrExtendSeq,lrUniqueSeq)
         srCompleteSeq <- paste0(srExtendSeq,srUniqueSeq)
    }
     if(geneidd==5){
         lrposStart <- start(lrgr)[4]
     lrposEnd  <- 44801630
     lrUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[5]:lrposEnd]))
     srUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]))
    
     srposStart <- start(lrgr)[2]
     srposEnd <- end(lrgr)[4]
     
     uniquePos <- c(173,517)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     lrExtendSeq <-paste0(as.character(geneSeq[[seqid]][75399518:end(lrgr)[1]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]))
      srExtendSeq <- paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]))
         lrCompleteSeq <- paste0(lrExtendSeq,lrUniqueSeq)
         srCompleteSeq <- paste0(srExtendSeq,srUniqueSeq)
    }
     if(geneidd==6){
     lrposStart <- 6534517
     lrposEnd <- end(lrgr)[2]
     lrUniqueSeq <- paste0(as.character(geneSeq[[seqid]][lrposStart:end(lrgr)[1]]),
                      as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]))
    srUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]))
    
     srposStart <- start(lrgr)[2]
     srposEnd <- end(lrgr)[4]
     
     uniquePos <- c(1,54)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     lrExtendSeq <-paste0(as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[5]:end(lrgr)[5]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[6]:end(lrgr)[6]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[7]:end(lrgr)[7]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[8]:end(lrgr)[8]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[9]:end(srgr)[8]]))
      srExtendSeq <- paste0(as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]]),
                           as.character(geneSeq[[seqid]][start(srgr)[5]:end(srgr)[5]]),
                           as.character(geneSeq[[seqid]][start(srgr)[6]:end(srgr)[6]]),
                           as.character(geneSeq[[seqid]][start(srgr)[7]:end(srgr)[7]]),
                           as.character(geneSeq[[seqid]][start(srgr)[8]:end(srgr)[8]]))
         lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq,srExtendSeq)
     }
     if(geneidd==7){
     lrposStart <- 216498844
     lrUniqueSeq <- paste0(as.character(geneSeq[[seqid]][lrposStart:end(lrgr)[1]]),
                      as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]))
     srUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),
                      as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]]))
     
     lrposEnd <- end(lrgr)[2]
     srposStart <- start(srgr)[1]
     srposEnd <- end(srgr)[2]
     
     uniquePos <- c(1,35)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     uniquePos <- c(1,67)
     substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
      srExtendSeq <- paste0(as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]]),as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]]))
           lrExtendSeq <- srExtendSeq
         lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
     }
    if(geneidd==8){
     lrUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(lrgr)[1]:end(lrgr)[1]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),
                           as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]))
    srUniqueSeq <- paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),
                           as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]]))

     lrposStart <- start(lrgr)[1]
     lrposEnd <- end(lrgr)[3]
     srposStart <- start(srgr)[1]
     srposEnd <- end(srgr)[2]
     
     uniquePos <- c(56,181)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     uniquePos <- c(32,33)
     substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
      srExtendSeq <- paste0(as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]]),as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]]),as.character(geneSeq[[seqid]][start(srgr)[5]:end(srgr)[5]]))
      lrExtendSeq <-paste0(as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]),as.character(geneSeq[[seqid]][start(lrgr)[5]:end(lrgr)[5]]),as.character(geneSeq[[seqid]][start(lrgr)[6]:89564542]))
      lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
      srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
     
    }
     if(geneidd==9){
     lrUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]),as.character(geneSeq[[seqid]][start(lrgr)[5]:end(lrgr)[5]])))
     srUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]]),as.character(geneSeq[[seqid]][start(srgr)[5]:end(srgr)[5]])))
     lrposStart <- start(lrgr)[4]
     lrposEnd <- end(lrgr)[5]
     srposStart <- start(srgr)[4]
     srposEnd <- end(srgr)[5]
     
     uniquePos <- c(1,46)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     uniquePos <- c(1,82)
     substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
     
    srExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]]),as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]])))
           lrExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][150444225:end(lrgr)[1]]),as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]])))
             
         lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
     }
     if(geneidd==10){
          lrposEnd <- 19380236
     lrUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(lrgr)[5]:end(lrgr)[5]]),as.character(geneSeq[[seqid]][start(lrgr)[6]:lrposEnd])))
     srUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[5]:end(srgr)[5]]),as.character(geneSeq[[seqid]][start(srgr)[6]:end(srgr)[6]])))
     lrposStart <- start(lrgr)[5]
     srposStart <- start(srgr)[5]
     srposEnd <- end(srgr)[6]
     
     uniquePos <- c(1,48)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     uniquePos <- c(1,112)
     substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
      srExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]]),as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]]),as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]])))
           lrExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][19376255:end(lrgr)[1]]),as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]])))
             
         lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
      }
    if(geneidd==11){
     lrposEnd <- 61967634
     lrUniqueSeq <- microseq::reverseComplement(as.character(geneSeq[[seqid]][start(lrgr)[4]:lrposEnd]))
     srUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]]),as.character(geneSeq[[seqid]][start(srgr)[5]:end(srgr)[5]])))
     lrposStart <- start(lrgr)[4]
     srposStart <- start(srgr)[4]
     srposEnd <- end(srgr)[5]
     
     uniquePos <- c(87,182)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     uniquePos <- c(84,85)
     substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
     
      srExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]]),as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]])))
           lrExtendSeq <- srExtendSeq
             
         lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
    }
    
     if(geneidd==12){
     lrposEnd <- 73293633 
     lrUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(lrgr)[6]:end(lrgr)[6]]), as.character(geneSeq[[seqid]][start(lrgr)[7]:lrposEnd])))
     srUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[5]:end(srgr)[5]]),as.character(geneSeq[[seqid]][start(srgr)[6]:end(srgr)[6]])))
     lrposStart <- start(lrgr)[6]
     srposStart <- start(srgr)[5]
     srposEnd <- end(srgr)[6]
     
     uniquePos <- c(1,36)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     uniquePos <- c(1,26)
     substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
     
     srExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]]),as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]]),as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]])))
           lrExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(lrgr)[1]:end(lrgr)[1]]),as.character(geneSeq[[seqid]][start(lrgr)[2]:end(lrgr)[2]]),as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]),as.character(geneSeq[[seqid]][start(lrgr)[4]:end(lrgr)[4]]),as.character(geneSeq[[seqid]][start(lrgr)[5]:end(lrgr)[5]])))
           
            lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
    }
     if(geneidd==13){
     lrposEnd <- 40835222
     lrUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(lrgr)[3]:end(lrgr)[3]]), as.character(geneSeq[[seqid]][start(lrgr)[4]:lrposEnd])))
     srUniqueSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[3]:end(srgr)[3]]),as.character(geneSeq[[seqid]][start(srgr)[4]:end(srgr)[4]])))
     lrposStart <- start(lrgr)[3]
     srposStart <- start(srgr)[3]
     srposEnd <- end(srgr)[4]
     
     uniquePos <- c(1,40)
     substr(lrUniqueSeq, uniquePos[1], uniquePos[2]) <-  tolower(substr(lrUniqueSeq, uniquePos[1], uniquePos[2]))
     uniquePos <- c(1,169)
     substr(srUniqueSeq, uniquePos[1], uniquePos[2]) <- tolower(substr(srUniqueSeq, uniquePos[1], uniquePos[2]))
     
     srExtendSeq <- microseq::reverseComplement(paste0(as.character(geneSeq[[seqid]][start(srgr)[1]:end(srgr)[1]]),as.character(geneSeq[[seqid]][start(srgr)[2]:end(srgr)[2]])))
           lrExtendSeq <- srExtendSeq
             
         lrCompleteSeq <- paste0(lrUniqueSeq,lrExtendSeq)
         srCompleteSeq <- paste0(srUniqueSeq, srExtendSeq)
    }
    dt <- tt_mcf7[gene_name == gene]
    dtLrList <- lapply(seq_along(lrUniqueSeq), function(x){
        dt_temp <- dt[variable=="lr"]
        dt_temp[,`:=`(uniqueSeq = lrUniqueSeq[x], 
                      nChar = nchar(lrUniqueSeq[x]),
                      completeSeq = lrCompleteSeq[x],
                      nCharComplete = nchar(lrCompleteSeq[x]),
                      start = lrposStart[x],
                      end = lrposEnd[x],
                      strand = strandtemp)]
        return(dt_temp)
    })
    dtSrList <- lapply(seq_along(srUniqueSeq), function(x){
        dt_temp <- dt[variable=="sr"]
         dt_temp[,`:=`(uniqueSeq = srUniqueSeq[x],
                      nChar = nchar(srUniqueSeq[x]),
                      completeSeq = srCompleteSeq[x],
                      nCharComplete = nchar(srCompleteSeq[x]),
                      start = srposStart[x],
                      end = srposEnd[x],
                      strand = strandtemp)]
        return(dt_temp)
    })
    dt <- rbindlist(do.call("c",list(dtSrList, dtLrList)))
    dt$chrname <- chrname
    return(dt)
}))
candidateTable <- gcDt[candidateTable, on = "tx_name"]
candidateTable <- unique(repResOnt[[2]][,.(tx_name, repRatio_byisoform_all)])[candidateTable, on = "tx_name"]
write.table(candidateTable, file = "candidateTable_dPCR_24Jun2024.csv", row.names = FALSE, col.names = TRUE, sep = ",")
```

```{r}
sink("digital_PCR/dpcr_seq.txt")
for( gene in candidateGene){
    cat("####",gene, "  \n")
    cat("#### LR isoform ",candidateTable[gene_name == gene&(variable == "lr")]$tx_name, "  \n")
    cat("#### SR isoform ",candidateTable[gene_name == gene&(variable == "sr")]$tx_name, "  \n")
    cat(">LRseq_rev \n")
    cat(candidateTable[gene_name == gene&(variable == "lr")]$uniqueSeq, "  \n")
    cat(">SRseq_rev \n")
    cat(candidateTable[gene_name == gene&(variable == "sr")]$uniqueSeq, "  \n")
    cat(">LRFullseq_rev \n")
    cat(candidateTable[gene_name == gene&(variable == "lr")]$completeSeq, "  \n")
    cat(">SRFullseq_rev \n")
    cat(candidateTable[gene_name == gene&(variable == "sr")]$completeSeq, "  \n")
    cat("\n")
    cat("\n")
}
sink()
```



create customised annotations
```{r}
# convert gtf file to ucsc format 
se.filtered <- seOutput[tt_mcf7[gene_name %in% candidateGene]$tx_name]
annotations.UCSC = rowRanges(se.filtered)
seqlevelsStyle(annotations.UCSC) <- "UCSC" #this reformats the chromosome names
annotations.UCSC = keepStandardChromosomes(annotations.UCSC, pruning.mode="coarse") #removes chromosomes UCSC doesn't have
writeToGTF(annotations.UCSC, "dpcr_candidateTx_ucsc.gtf")
```


## check in simulation data 
```{r}
simEst <- unique(comDataTranscriptTrimmed[cellLine == "MCF7"&(protocol_general == "cDNA"),list(meanEst=mean(normEst)), by = list(tx_name,method)],by=NULL)

candidateTable <- simEst[method == "trim_lr_sim",.(tx_name, meanEst)][candidateTable, on = "tx_name"]
setnames(candidateTable, "meanEst","lrTrimSimEst")
write.table(candidateTable, file = "candidateTable.csv", row.names = FALSE, col.names = TRUE, sep = ",")
```

```{r}

for(gene in candidateList){
    major_tx <- tt_mcf7[gene_name == gene]$tx_name
    print(simEst[method == "trim_lr_sim",.(tx_name, meanEst)][tx_name %in% major_tx])
}
```

# Supplementary Text Fig 21 GOseq enrichment analysis 
```{r}
# For genes that differ in major isoform in lr vs sr vs genes that not share major isoform in lr vs sr 
library(data.table)

dominant_typeData <- readRDS("dominant_typeData_25May2023.rds")
dominant_typeData[, gene_expressed := all(geneExpression>0), by = list(gene_name, cellLine, short_read)]
dominant_typeData[, gene_expressed_all := any(gene_expressed), by = list(gene_name, cellLine)]

geneExpressed <- unique(dominant_typeData[,.(gene_name, cellLine, gene_expressed_all)], by = NULL)
ddData <- unique(dominant_typeData[,
                .(cellLine, tx_name, gene_name,majorBoth, majorEither, majorEitherOnly, majorSecBoth, majorLongRead, majorShortRead,gene_expressed_all)])
ddData[, major_status := ifelse(majorEitherOnly&majorLongRead, "majorLongReadOnly",ifelse(majorEitherOnly&majorShortRead, "majorShortReadOnly", ifelse(majorBoth, "majorBoth","others")))]

geneList <- unique(ddData[, list(share_status = any(major_status == "majorBoth")), by = list(gene_name, cellLine,gene_expressed_all)], by = NULL)

# by cell line
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggplot2)
library(ggpubr)
cellLines <- unique(geneList$cellLine)[c(1,2,3,5,7,9,10)]
for(ont_type in c("MF","CC","BP","ALL")){
    print(ont_type)
    goObjectList <- lapply(cellLines, function(x){
    ego <- enrichGO(gene         = geneList[cellLine == x &(share_status == FALSE)&(gene_expressed_all==TRUE)]$gene_name,
                    universe = geneExpressed[cellLine == x &(gene_expressed_all==TRUE)]$gene_name, 
                OrgDb         = org.Hs.eg.db,
                keyType       = 'ENSEMBL',
                ont           = ont_type,
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                pool = FALSE)
    return(ego)
    
})

goObjectListShared <- lapply(cellLines, function(x){
    ego <- enrichGO(gene         = geneList[cellLine == x &(share_status == TRUE)&(gene_expressed_all==TRUE)]$gene_name,
                    universe = geneExpressed[cellLine == x &(gene_expressed_all==TRUE)]$gene_name, 
                OrgDb         = org.Hs.eg.db,
                keyType       = 'ENSEMBL',
                ont           = ont_type,
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                pool = FALSE)
    return(ego)
    
})
saveRDS(c(goObjectList, goObjectListShared), file = paste0("goObject_",ont_type,"_nolevel_bgexpressed.rds"))
}

for(ont_type in c("MF","CC","BP")){
    ddd <- readRDS(paste0("goObject_",ont_type,"_nolevel_bgexpressed.rds"))
    goObjectList <- ddd[1:7]
    goObjectListShared <- ddd[8:14]
    pListDiffGene <- lapply(1:7, function(k){
     p <- dotplot(goObjectList[[k]], showCategory=30)  + ggtitle(paste0("Diff (",cellLines[k],")"))
     return(p)
})
pListSharedGene <- lapply(1:7, function(k){
     p <- dotplot(goObjectListShared[[k]], showCategory=30)  + ggtitle(paste0("Same (",cellLines[k],")"))
     return(p)
})
pdf(paste0("GOEnrichmentAnalysis_LRvsSRDiffMajorIsoform_Genes_",ont_type,"_bgexpressed.pdf"), width = 24, height = 16)
ggarrange(plotlist = pListDiffGene, nrow = 2, ncol = 4)
dev.off()
pdf(paste0("GOEnrichmentAnalysis_LRvsSRShareMajorIsoform_Genes_",ont_type,"_bgexpressed.pdf"), width = 24, height = 16)
ggarrange(plotlist = pListSharedGene, nrow = 2, ncol = 4)
dev.off()
}



```

```{r}
mfList <- readRDS("goObject_MF_nolevel_bgexpressed.rds")
ccList <- readRDS("goObject_CC_nolevel_bgexpressed.rds")
bpList <- readRDS("goObject_BP_nolevel_bgexpressed.rds")
```


```{r}
st <- CJ(type = c(1,2,3), # mf, cc, bp
         shared = c(1,2),
         cellLine = 1:7) # diff, shared

dtCombined <- do.call("rbind",lapply(seq_len(nrow(st)), function(x){
    dt <- list(mfList, ccList, bpList)[[st[x]$type]][[(st[x]$shared-1)*7+st[x]$cellLine]]@result
    dt$geneID <- NULL
    dt <- data.table(dt)
    dt[, type := c("mf","cc","bp")[st[x]$type]]
    dt[, shared := c("diff","shared")[st[x]$shared]]
    dt[, cellLine := cellLines[st[x]$cellLine]]
    return(dt)
}))

dtCombined[, generatio := Count/as.integer(unlist(strsplit(GeneRatio,"\\/"))[2]), by = list(GeneRatio, Count)]
# compute odds ratio 
dtCombined[, count2 := (as.integer(unlist(strsplit(BgRatio,"\\/"))[1])-Count), by = list(BgRatio, Count)]
dtCombined[, total2 := (as.integer(unlist(strsplit(BgRatio,"\\/"))[2])-as.integer(unlist(strsplit(GeneRatio,"\\/"))[2])), by = list(BgRatio, GeneRatio)]

dtCombined[,oddRatio := generatio/(1-generatio)/(count2/total2/(1-count2/total2))]
dtCombined[, signif_status := p.adjust<0.05]
dtCombined[, rankNO := order(-generatio), by = list(type, shared, cellLine)]



ggplot(dtCombined, aes(x = cellLine, y = reorder(ID,generatio)))+
    geom_tile(aes(size = Count, fill = generatio))+
    facet_grid(type ~ shared, scales = "free")+
    theme(legend.position = "none")
    
library(ggrepel)
ggplot(dtCombined, aes(x = generatio, y = -log10(p.adjust)))+geom_point(aes(col = cellLine), alpha = 0.5, shape = 1)+facet_wrap(type ~ shared, scales = "free", nrow = 3, ncol = 2)+geom_text_repel(data = dtCombined[generatio> 0.03],aes(label = Description), max.overlaps = 30)+theme_classic()

dtCombined[signif_status==TRUE&(generatio>0.02)]

library(ComplexHeatmap)

countDt <- dcast(dtCombined, ID + Description + type ~ paste0(shared,cellLine), value.var = "Count")
padjDt <- dcast(dtCombined, ID + Description + type ~ paste0(shared,cellLine), value.var = "p.adjust")

geneRatioDt <- dcast(dtCombined[ID %in% unique(dtCombined[(p.adjust<0.05)&(pvalue < 0.05)&(qvalue < 0.05)&(generatio>0.01)]$ID)], ID + Description + type ~ paste0(shared,cellLine), value.var = "generatio")
geneRatioDt[is.na(geneRatioDt)] <- 0
diff_mf <- Heatmap(as.matrix(geneRatioDt[type == "mf",colnames(geneRatioDt)[grep("diff",colnames(geneRatioDt))],with = FALSE]), 
                   row_labels = geneRatioDt[type == "mf"]$Description,
                   name = "diff_mf")
shared_mf <- Heatmap(as.matrix(geneRatioDt[type == "mf",colnames(geneRatioDt)[grep("shared",colnames(geneRatioDt))],with = FALSE]), 
                     row_labels = geneRatioDt[type == "mf"]$Description,
                     name = "shared_mf")
mf <- diff_mf + shared_mf


diff_cc <- Heatmap(as.matrix(geneRatioDt[type == "cc",colnames(geneRatioDt)[grep("diff",colnames(geneRatioDt))],with = FALSE]), 
                   row_labels = geneRatioDt[type == "cc"]$Description,
                   name = "diff_cc")
shared_cc <- Heatmap(as.matrix(geneRatioDt[type == "cc",colnames(geneRatioDt)[grep("shared",colnames(geneRatioDt))],with = FALSE]), 
                     row_labels = geneRatioDt[type == "cc"]$Description,
                     name = "shared_cc")
cc <- diff_cc + shared_cc

geneRatioDt <- dcast(dtCombined[ID %in% unique(dtCombined[(p.adjust<0.05)&(pvalue < 0.05)&(qvalue < 0.05)&(generatio>0.025)]$ID)], ID + Description + type ~ paste0(shared,cellLine), value.var = "generatio")
geneRatioDt[is.na(geneRatioDt)] <- 0
diff_bp <- Heatmap(as.matrix(geneRatioDt[type == "bp",colnames(geneRatioDt)[grep("diff",colnames(geneRatioDt))],with = FALSE]), 
                   row_labels = geneRatioDt[type == "bp"]$Description,
                   name = "diff_bp")
shared_bp <- Heatmap(as.matrix(geneRatioDt[type == "bp",colnames(geneRatioDt)[grep("shared",colnames(geneRatioDt))],with = FALSE]), 
                     row_labels = geneRatioDt[type == "bp"]$Description,
                     name = "shared_bp")
bp <- diff_bp + shared_bp

pdf("lrsr_goenrichment_0.01_0.025_geneexpressed.pdf", width = 10, height = 9)
draw(bp, column_title = "BP",heatmap_legend_side = "left")
draw(cc, column_title = "CC", heatmap_legend_side = "left")
draw(mf, column_title = "MF", heatmap_legend_side = "left")
dev.off()
```



```{r}
hist(dtCombined[signif_status == TRUE&(type == "cc")]$generatio)
```



